
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Additional plots and customizations &#8212; Mosaic v2.1 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Mosaic analysis walk-through" href="analysis-walkthrough.html" />
    <link rel="prev" title="Multisample analysis" href="multi-sample-analysis.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Mosaic v2.1 documentation</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../1_introduction.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../2_install.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../3_getting_started.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../4_data_structure.html">
   Data structure
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="index.html">
   Code Snippets
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="basics.html">
   Basic usage of mosaic
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cnv.html">
   Manual CNV analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="filtering.html">
   Filtering barcodes and ids
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="customizations.html">
   Plot customizations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="multi-sample-analysis.html">
   Multisample analysis
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Additional plots and customizations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysis-walkthrough.html">
   Mosaic analysis walk-through
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modules
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.io.html">
   io
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.utils.html">
   utils
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pages/missionbio.mosaic.constants.html">
   constants
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.COLORS.html">
     missionbio.mosaic.constants.COLORS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.PCA_LABEL.html">
     missionbio.mosaic.constants.PCA_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.UMAP_LABEL.html">
     missionbio.mosaic.constants.UMAP_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.SCALED_LABEL.html">
     missionbio.mosaic.constants.SCALED_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.LABEL.html">
     missionbio.mosaic.constants.LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.READS.html">
     missionbio.mosaic.constants.READS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.PALETTE.html">
     missionbio.mosaic.constants.PALETTE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.NORMALIZED_READS.html">
     missionbio.mosaic.constants.NORMALIZED_READS
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Classes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.assay._Assay.html">
   _Assay
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.dna.Dna.html">
   Dna
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.cnv.Cnv.html">
   Cnv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.protein.Protein.html">
   Protein
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.sample.Sample.html">
   Sample
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.samplegroup.SampleGroup.html">
   SampleGroup
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/examples/additional-plots-and-customizations.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-in-dependencies">
   Load in dependencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-in-data">
   Load in data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#gq-violin-plots">
   GQ violin plots
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#stacked-bar-graph-for-multi-sample-analysis">
   Stacked bar graph for multi-sample analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#reads-per-amplicon">
   Reads per amplicon
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#count-barcodes-with-at-least-one-mutation">
   Count barcodes with at least one mutation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#custom-cluster-order">
   Custom cluster order
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#custom-heatmap-legend">
   Custom heatmap legend
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#perform-cnv-analysis-on-your-data">
   Perform CNV analysis on your data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#p-value-chart-per-gene">
   P-value chart per gene
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculate-an-average-p-value-per-gene">
   Calculate an average P-value per gene
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#protein-read-distribution-tool">
   Protein read distribution tool
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plot-a-joint-scatterplot-and-histogram">
   Plot a joint scatterplot and histogram
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cnv-signature-per-chromosome">
   CNV signature per chromosome
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#save-output-for-one-clone">
   Save output for one clone
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Additional plots and customizations</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-in-dependencies">
   Load in dependencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-in-data">
   Load in data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#gq-violin-plots">
   GQ violin plots
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#stacked-bar-graph-for-multi-sample-analysis">
   Stacked bar graph for multi-sample analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#reads-per-amplicon">
   Reads per amplicon
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#count-barcodes-with-at-least-one-mutation">
   Count barcodes with at least one mutation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#custom-cluster-order">
   Custom cluster order
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#custom-heatmap-legend">
   Custom heatmap legend
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#perform-cnv-analysis-on-your-data">
   Perform CNV analysis on your data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#p-value-chart-per-gene">
   P-value chart per gene
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculate-an-average-p-value-per-gene">
   Calculate an average P-value per gene
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#protein-read-distribution-tool">
   Protein read distribution tool
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plot-a-joint-scatterplot-and-histogram">
   Plot a joint scatterplot and histogram
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cnv-signature-per-chromosome">
   CNV signature per chromosome
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#save-output-for-one-clone">
   Save output for one clone
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="additional-plots-and-customizations">
<h1>Additional plots and customizations<a class="headerlink" href="#additional-plots-and-customizations" title="Permalink to this headline">#</a></h1>
<p><b>Objective</b></p>
<p>This vignette contains a wide variety of code snippets for additional plots and plot customizations that are not currently included in our latest version of the jupyter notebook.</p>
<section id="load-in-dependencies">
<h2>Load in dependencies<a class="headerlink" href="#load-in-dependencies" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import mosaic libraries</span>
<span class="kn">import</span> <span class="nn">missionbio.mosaic</span> <span class="k">as</span> <span class="nn">ms</span>

<span class="c1"># Import these to display entire dataframes</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>

<span class="c1"># Import graph_objects from the plotly package to display figures when saving the notebook as an HTML</span>
<span class="c1"># Import numpy for statistics</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># Import additional packages for specific visuals</span>
<span class="kn">import</span> <span class="nn">missionbio.mosaic.utils</span> <span class="k">as</span> <span class="nn">mutils</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># Import the colors</span>
<span class="kn">from</span> <span class="nn">missionbio.mosaic.constants</span> <span class="kn">import</span> <span class="n">COLORS</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="load-in-data">
<h2>Load in data<a class="headerlink" href="#load-in-data" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">load_example_dataset</span><span class="p">(</span><span class="s2">&quot;Multisample PBMC&quot;</span><span class="p">,</span> <span class="n">single</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading, &lt;_io.BytesIO object at 0x7fbc83cabf40&gt;
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loaded in 0.6s.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/casp/Documents/code/mosaic/src/missionbio/mosaic/assay.py:321: UserWarning: Multiple samples detected. Appending the sample name to the barcodes to avoid duplicate barcodes. The original barcodes are stored in the original_barcode row attribute
  warnings.warn(msg)
</pre></div>
</div>
</div>
</div>
<p>Subset the DNA data to look for high quality variants of interest.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Adjust filters if needed by overwriting dna_vars</span>
<span class="n">dna_vars</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">(</span>
    <span class="n">min_dp</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">min_gq</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">vaf_ref</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">vaf_hom</span><span class="o">=</span><span class="mi">95</span><span class="p">,</span>
    <span class="n">vaf_het</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
    <span class="n">min_prct_cells</span><span class="o">=</span><span class="mi">70</span><span class="p">,</span>
    <span class="n">min_mut_prct_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">[</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(),</span> <span class="n">dna_vars</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="gq-violin-plots">
<h2>GQ violin plots<a class="headerlink" href="#gq-violin-plots" title="Permalink to this headline">#</a></h2>
<p>Plot GQ violin plots for all variants of interest to assess genotype quality. It is recommended that this is added before the GQ heatmap, so poor quality variants seen in the violin plots can be subsequently de-selected in the heatmap.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">GQ_flipped</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;GQ&#39;</span><span class="p">])</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">())</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">())):</span>
    <span class="n">this_col</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">df</span><span class="p">[</span><span class="n">this_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">GQ_flipped</span><span class="p">[</span><span class="n">i</span><span class="p">,:]</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;Variant&#39;</span><span class="p">,</span> <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;GQ&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s1">&#39;whitegrid&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Variant&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;GQ&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;Variant&#39;</span><span class="p">,</span> <span class="n">inner</span><span class="o">=</span><span class="s1">&#39;quartile&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.02</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">borderaxespad</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/efbe1037c886ab1e0a0eef335e1776b579d68f2e94ebbae2a7119bae59ea4c2c.png" src="../_images/efbe1037c886ab1e0a0eef335e1776b579d68f2e94ebbae2a7119bae59ea4c2c.png" />
</div>
</div>
<p>Identify the subclones present in your sample.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Plot a heatmap of the variant allele frequenices being called across your variants of interest</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;AF_MISSING&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "da594db4a52748af83ce5fea8f78897c"}</script></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Annotate the DNA variants</span>
<span class="n">annotation</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_annotations</span><span class="p">()</span>  

<span class="c1"># Store the annotations in the dna assay as a new column attribute</span>
<span class="k">for</span> <span class="n">col</span><span class="p">,</span> <span class="n">content</span> <span class="ow">in</span> <span class="n">annotation</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_col_attr</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">content</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    
<span class="c1"># Add annotation to the id names</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">set_ids_from_cols</span><span class="p">([</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;ALT&quot;</span><span class="p">])</span>

<span class="c1"># Annotations are now added to the variants</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([&#39;SF3B1:2:198266943:C:T&#39;, &#39;NPM1:5:170837457:A:G&#39;,
       &#39;EZH2:7:148504854:A:AGACTT&#39;, &#39;EZH2:7:148508833:A:G&#39;,
       &#39;WT1:11:32417945:T:C&#39;, &#39;TP53:17:7578115:T:C&#39;,
       &#39;PHF6:X:133547814:T:C&#39;], dtype=object)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Cluster the data to identify the subclones present</span>

<span class="n">clone_data</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">group_by_genotype</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">(),</span> <span class="n">group_missing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">min_clone_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;NGT_FILTERED&quot;</span><span class="p">,</span> <span class="n">show_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">clone_data</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/casp/Documents/code/mosaic/src/missionbio/mosaic/dna.py:960: FutureWarning:

The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
</pre></div>
</div>
<div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>clone</th>
      <th>1</th>
      <th>2</th>
      <th>Missing GT clones (n=409)</th>
      <th>Small subclones (n=22)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>SF3B1:2:198266943:C:T</th>
      <td>Hom (99.54%)</td>
      <td>Hom (99.65%)</td>
      <td>Missing in 13.54%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>NPM1:5:170837457:A:G</th>
      <td>WT (0.47%)</td>
      <td>Hom (99.58%)</td>
      <td>Missing in 20.29%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>EZH2:7:148504854:A:AGACTT</th>
      <td>WT (0.28%)</td>
      <td>Het (62.06%)</td>
      <td>Missing in 28.74%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>EZH2:7:148508833:A:G</th>
      <td>WT (0.72%)</td>
      <td>Het (60.0%)</td>
      <td>Missing in 23.03%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>WT1:11:32417945:T:C</th>
      <td>Het (58.35%)</td>
      <td>WT (0.49%)</td>
      <td>Missing in 22.34%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>TP53:17:7578115:T:C</th>
      <td>Hom (99.8%)</td>
      <td>Het (58.84%)</td>
      <td>Missing in 26.93%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>PHF6:X:133547814:T:C</th>
      <td>Hom (99.61%)</td>
      <td>WT (0.41%)</td>
      <td>Missing in 18.40%% of cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Total Cell Number</th>
      <td>490 (13.06%)</td>
      <td>189 (5.04%)</td>
      <td>2975 (79.31%)</td>
      <td>97 (2.59%)</td>
    </tr>
    <tr>
      <th>Sample 1 Cell Number</th>
      <td>8 (0.41%)</td>
      <td>183 (9.42%)</td>
      <td>1690 (86.98%)</td>
      <td>62 (3.19%)</td>
    </tr>
    <tr>
      <th>Sample 2 Cell Number</th>
      <td>482 (26.66%)</td>
      <td>6 (0.33%)</td>
      <td>1285 (71.07%)</td>
      <td>35 (1.94%)</td>
    </tr>
    <tr>
      <th>Parents</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Sisters</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ADO score</th>
      <td>0</td>
      <td>0</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table></div><img alt="../_images/94532f5ae526fa1be67692c76b7da1b4f08a8fe899ede156ea37a5fa1b2f162d.png" src="../_images/94532f5ae526fa1be67692c76b7da1b4f08a8fe899ede156ea37a5fa1b2f162d.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Rename the subclones and drop the clones that are not of interest by labelling them &#39;FP&#39;</span>

<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">rename_labels</span><span class="p">(</span>
  <span class="p">{</span>
    <span class="s1">&#39;1&#39;</span><span class="p">:</span> <span class="s1">&#39;Clone 1&#39;</span><span class="p">,</span> 
    <span class="s1">&#39;2&#39;</span><span class="p">:</span> <span class="s1">&#39;Clone 2&#39;</span><span class="p">,</span>    
    <span class="s1">&#39;missing&#39;</span><span class="p">:</span> <span class="s1">&#39;FP&#39;</span><span class="p">,</span> 
    <span class="s1">&#39;small&#39;</span><span class="p">:</span> <span class="s1">&#39;FP&#39;</span>
  <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove barcodes (clones) from data based on renamed labels</span>
<span class="c1"># The reduced data set will now be called &#39;sample.dna2&#39;</span>
<span class="n">fp_barcodes</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">({</span><span class="s2">&quot;FP&quot;</span><span class="p">})</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">fp_barcodes</span><span class="p">)</span> 
<span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;Clone 1&#39;, &#39;Clone 2&#39;}
</pre></div>
</div>
</div>
</div>
</section>
<section id="stacked-bar-graph-for-multi-sample-analysis">
<h2>Stacked bar graph for multi-sample analysis<a class="headerlink" href="#stacked-bar-graph-for-multi-sample-analysis" title="Permalink to this headline">#</a></h2>
<p>Plot a stacked bar graph for clone frequency within each sample, where each sample is an individual bar. Note colors can be changed, as can labels. Graph is highly customizable, but large amounts of code may deter users.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;sample_name&#39;</span><span class="p">]</span>
<span class="n">a_list</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="n">b</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
<span class="n">b_list</span> <span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="n">ab</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">a_list</span><span class="p">,</span> <span class="n">b_list</span><span class="p">))</span>

<span class="n">counts_dic</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">sample_tots</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">ab</span><span class="p">:</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">s</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sample_tots</span><span class="p">:</span>
        <span class="n">sample_tots</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sample_tots</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">s</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">counts_dic</span><span class="p">:</span>
        <span class="n">counts_dic</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">clone</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">clone</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">counts_dic</span><span class="p">[</span><span class="n">s</span><span class="p">]:</span>
        <span class="n">counts_dic</span><span class="p">[</span><span class="n">s</span><span class="p">][</span><span class="n">clone</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">counts_dic</span><span class="p">[</span><span class="n">s</span><span class="p">][</span><span class="n">clone</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="n">freq_dic</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">sample_names</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">counts_dic</span><span class="p">:</span>
    <span class="n">sample_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
    <span class="n">s_tot</span> <span class="o">=</span> <span class="n">sample_tots</span><span class="p">[</span><span class="n">s</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts_dic</span><span class="p">[</span><span class="n">s</span><span class="p">]:</span>
        <span class="n">freq</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">counts_dic</span><span class="p">[</span><span class="n">s</span><span class="p">][</span><span class="n">c</span><span class="p">])</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">s_tot</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">freq_dic</span><span class="p">:</span>
            <span class="n">freq_dic</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">freq</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">freq_dic</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">freq</span><span class="p">)</span>       

<span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_names</span><span class="p">))]</span>

<span class="c1">#IMPORTANT NOTE: DO NOT ALTER ANYTHING ABOVE THIS COMMENT</span>

<span class="n">width</span> <span class="o">=</span> <span class="mf">.5</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s1">&#39;ticks&#39;</span><span class="p">)</span>
<span class="c1">#One line item per clone</span>
<span class="c1">#After first line, add all previous clone data to the &#39;bottom&#39; of the next bar</span>
<span class="c1">#Insert your clone names</span>
<span class="c1">#Colors can be custom, insert preferred hex-codes</span>
<span class="n">clone_1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">freq_dic</span><span class="p">[</span><span class="s1">&#39;Clone 1&#39;</span><span class="p">],</span> <span class="n">width</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#269FCD&#39;</span><span class="p">)</span>
<span class="n">clone_2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">freq_dic</span><span class="p">[</span><span class="s1">&#39;Clone 2&#39;</span><span class="p">],</span> <span class="n">width</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="n">freq_dic</span><span class="p">[</span><span class="s1">&#39;Clone 1&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#9BBDF9&#39;</span><span class="p">)</span>
<span class="c1">#clone_3 = plt.bar(x, freq_dic[&#39;Clone 3&#39;], width, bottom=np.array(freq_dic[&#39;Clone 1&#39;])+np.array(freq_dic[&#39;Clone 2&#39;]), color=&#39;#67AAf9&#39;)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">sample_names</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Clone frequency per sample&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">((</span><span class="n">clone_1</span><span class="p">,</span> <span class="n">clone_2</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;Clone 1&#39;</span><span class="p">,</span> <span class="s1">&#39;Clone 2&#39;</span><span class="p">),</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.02</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">borderaxespad</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;matplotlib.legend.Legend at 0x7fbc716f09d0&gt;
</pre></div>
</div>
<img alt="../_images/c21070ae66ac272bae17f0d829d2e62eb8eae056248801bb11cea89cd3265b09.png" src="../_images/c21070ae66ac272bae17f0d829d2e62eb8eae056248801bb11cea89cd3265b09.png" />
</div>
</div>
</section>
<section id="reads-per-amplicon">
<h2>Reads per amplicon<a class="headerlink" href="#reads-per-amplicon" title="Permalink to this headline">#</a></h2>
<p>Plot a histogram of the reads per cell across any given amplicon, and print out the mean reads per cell of that amplicon.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reads</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="s1">&#39;DP&#39;</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row+col&#39;</span><span class="p">)</span>
<span class="n">amplicons</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;amplicon&#39;</span><span class="p">]</span>
<span class="n">amp_average</span> <span class="o">=</span> <span class="n">reads</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">amplicons</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="c1">#Insert your amplicon of interest below</span>
<span class="n">DP</span> <span class="o">=</span> <span class="n">amp_average</span><span class="p">[</span><span class="s1">&#39;MYE_SF3B1_198266730&#39;</span><span class="p">]</span> 
<span class="c1">#Print the mean</span>
<span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">DP</span><span class="p">))</span>
<span class="c1">#Plot histogram</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">DP</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>39.1149026926153
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(array([  37.,  118.,  255.,  305.,  340.,  444.,  393., 1321.,  514.,
          24.]),
 array([ 0. ,  6.4, 12.8, 19.2, 25.6, 32. , 38.4, 44.8, 51.2, 57.6, 64. ]),
 &lt;a list of 10 Patch objects&gt;)
</pre></div>
</div>
<img alt="../_images/de50f70c1f273651a9c1b5246fef14be0f19469958acffcccb45fe5078db7ee2.png" src="../_images/de50f70c1f273651a9c1b5246fef14be0f19469958acffcccb45fe5078db7ee2.png" />
</div>
</div>
</section>
<section id="count-barcodes-with-at-least-one-mutation">
<h2>Count barcodes with at least one mutation<a class="headerlink" href="#count-barcodes-with-at-least-one-mutation" title="Permalink to this headline">#</a></h2>
<p>Define a region of interest, and then call cell-barcodes that have a least one mutation in that region of interest. This code may be useful for gene-editing experiments where you are expecting mutations on a particular amplicon, or within a small region of the amplicon surrounding a cut-site.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Define an amplicon of interest</span>
<span class="n">amp</span> <span class="o">=</span> <span class="s1">&#39;MYE_SF3B1_198266730&#39;</span>
<span class="nb">print</span> <span class="p">(</span><span class="n">amp</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;amplicon&#39;</span><span class="p">]))</span> <span class="c1">#sanity check</span>
<span class="c1">#Filter the data to only look at variants called from this amplicon</span>
<span class="n">filt</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s2">&quot;amplicon&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">amp</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()[</span><span class="n">filt</span><span class="p">]</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">[:,</span> <span class="n">variants</span><span class="p">]</span>
 
<span class="c1">#THE FOLLOWING SECTION IS OPTIONAL</span>

<span class="c1">#filtering variants by cut site</span>
<span class="c1">#e.g. cut site at chr2:60722401, range 60722481 - 60722421</span>
<span class="c1">#filt2 = sample.dna.col_attrs[&quot;POS&quot;] &gt; 60722381</span>
<span class="c1">#variants2 = sample.dna.ids()[filt2]</span>
<span class="c1">#sample.dna = sample.dna[:, variants2]</span>
<span class="c1">#filt3 = sample.dna.col_attrs[&quot;POS&quot;] &lt; 60722421</span>
<span class="c1">#variants3 = sample.dna.ids()[filt3]</span>
<span class="c1">#sample.dna = sample.dna[:, variants3]</span>

<span class="c1">#counting barcodes with at least one mutation in region of interest(Het or Hom call)</span>
<span class="c1">#creating a dataframe with numeric genotypes</span>
<span class="n">genotypes</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row+col&#39;</span><span class="p">)</span>
<span class="c1">#set WT to True</span>
<span class="n">zero_bool</span> <span class="o">=</span> <span class="p">(</span><span class="n">genotypes</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
<span class="c1">#number of WT calls per each cell</span>
<span class="n">zero_bool</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1">#set missing to True</span>
<span class="n">missing_bool</span> <span class="o">=</span> <span class="p">(</span><span class="n">genotypes</span> <span class="o">==</span> <span class="mi">3</span><span class="p">)</span>
<span class="c1">#number of Missing calls per each cell</span>
<span class="n">missing_bool</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1">#number of Het and Hom calls per barcode</span>
<span class="c1">#len(genotypes.columns) is the total number of variants</span>
<span class="n">barcode_muts</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genotypes</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">missing_bool</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">zero_bool</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

<span class="c1">#number of barcodes with one or more mutations</span>
<span class="n">barcode_muts_bool</span> <span class="o">=</span> <span class="p">(</span><span class="n">barcode_muts</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">barcode_muts_bool</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
3243
</pre></div>
</div>
</div>
</div>
</section>
<section id="custom-cluster-order">
<h2>Custom cluster order<a class="headerlink" href="#custom-cluster-order" title="Permalink to this headline">#</a></h2>
<p>For sample.dna or sample.cnv grouped heatmaps in Mosaic notebook, this code will help you order the clones along the Y-axis in whatever order you prefer.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1">#get the &#39;barcodes&#39; and the labels information from the sample.dna object</span>
<span class="c1">#label is the attribute that will be used for ordering cell groups</span>
<span class="c1">#labels are stored in the &#39;clusters&#39; variable</span>
<span class="n">barcodes</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">]</span>
<span class="n">clusters</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>

<span class="c1">#combining barcodes and clusters into one single numpy array</span>
<span class="n">combined</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">clusters</span><span class="p">))</span><span class="o">.</span><span class="n">T</span>

<span class="c1">#creating a pandas data frame based on the numpy array</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">combined</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cell&#39;</span><span class="p">,</span><span class="s1">&#39;cluster&#39;</span><span class="p">])</span>

<span class="c1">#create a list &#39;cell_clusters&#39; with the desired order of the clusters</span>
<span class="c1">#In this example, clusters are labeled numerically, i.e. 1, 2, 3,...</span>
<span class="c1">#Use the labels you have previously created for the clones</span>
<span class="n">cell_clusters</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Clone 1&#39;</span><span class="p">,</span> <span class="s1">&#39;Clone 2&#39;</span><span class="p">]</span>

<span class="c1">#recreate the &#39;cluster&#39; column in the pandas data frame, but as a categorical column</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;cluster&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;cluster&quot;</span><span class="p">],</span> <span class="n">categories</span> <span class="o">=</span> <span class="n">cell_clusters</span><span class="p">)</span>
<span class="c1">#group the barcodes by cluster</span>
<span class="n">df_sorted</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
<span class="c1">#change data type (back to numpy)</span>
<span class="n">ordered_cells</span> <span class="o">=</span> <span class="n">df_sorted</span><span class="o">.</span><span class="n">cell</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ordered_cells</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ordered_cells</span><span class="p">)</span>
<span class="c1">#call the plot again, use &#39;ordered_cells&#39; to specify the order</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;AF_MISSING&#39;</span><span class="p">,</span> <span class="n">bars_order</span> <span class="o">=</span> <span class="n">ordered_cells</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "b9123449ce114f48b32e57c90271c08b"}</script></div>
</div>
</section>
<section id="custom-heatmap-legend">
<h2>Custom heatmap legend<a class="headerlink" href="#custom-heatmap-legend" title="Permalink to this headline">#</a></h2>
<p>For grouped sample.dna heatmaps, after clones are identified and labeled, this helps to remove unnecessary categories from the legend, for example the Missing category.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Customize legend to show only three categories</span>
<span class="c1">#e.g. WT, HET, HOM</span>
<span class="c1">#if the data are filtered, to remove the missing genotypes, the &#39;Missing&#39; category will still appear in the legend</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span> <span class="o">=</span> <span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="c1">#Define the categories you want to appear</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorbar</span><span class="o">.</span><span class="n">ticktext</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;WT&#39;</span><span class="p">,</span> <span class="s1">&#39;HET&#39;</span><span class="p">,</span> <span class="s1">&#39;HOM&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorbar</span><span class="o">.</span><span class="n">tickvals</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.375</span><span class="p">,</span> <span class="mf">1.025</span><span class="p">,</span> <span class="mf">1.675</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">cmax</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="p">[[</span><span class="mf">0.0</span><span class="p">,</span> <span class="s1">&#39;#3b4d73&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.33</span><span class="p">,</span> <span class="s1">&#39;#3b4d73&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.33</span><span class="p">,</span> <span class="s1">&#39;#78a3bc&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.66</span><span class="p">,</span> <span class="s1">&#39;#78a3bc&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.66</span><span class="p">,</span> <span class="s1">&#39;#d7ecee&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">1.0</span><span class="p">,</span> <span class="s1">&#39;#d7ecee&#39;</span><span class="p">]]</span>
<span class="n">fig</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "b2fe9ac9a01f4803bca857d562a011ef"}</script></div>
</div>
</section>
<section id="perform-cnv-analysis-on-your-data">
<h2>Perform CNV analysis on your data<a class="headerlink" href="#perform-cnv-analysis-on-your-data" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reads</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="s1">&#39;read_counts&#39;</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row+col&#39;</span><span class="p">)</span>
<span class="n">reads</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>MYE_ASXL1_31022366</th>
      <th>MYE_ASXL1_31022711</th>
      <th>MYE_ASXL1_31022977</th>
      <th>MYE_ASXL1_31023229</th>
      <th>MYE_ASXL1_31023478</th>
      <th>MYE_ASXL1_31023678</th>
      <th>MYE_ASXL1_31023938</th>
      <th>MYE_ASXL1_31024173</th>
      <th>MYE_ASXL1_31024623</th>
      <th>MYE_ASXL1_31024877</th>
      <th>...</th>
      <th>MYE_ZRSR2_15817959</th>
      <th>MYE_ZRSR2_15821773</th>
      <th>MYE_ZRSR2_15822168</th>
      <th>MYE_ZRSR2_15826277</th>
      <th>MYE_ZRSR2_15827302</th>
      <th>MYE_ZRSR2_15833836</th>
      <th>MYE_ZRSR2_15838246</th>
      <th>MYE_ZRSR2_15840789</th>
      <th>MYE_ZRSR2_15841002</th>
      <th>MYE_ZRSR2_15841178</th>
    </tr>
    <tr>
      <th>read_counts</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AACAACCTACAATCATAC-Sample 1</th>
      <td>0</td>
      <td>5</td>
      <td>32</td>
      <td>12</td>
      <td>7</td>
      <td>21</td>
      <td>22</td>
      <td>9</td>
      <td>13</td>
      <td>12</td>
      <td>...</td>
      <td>12</td>
      <td>4</td>
      <td>7</td>
      <td>54</td>
      <td>3</td>
      <td>17</td>
      <td>3</td>
      <td>6</td>
      <td>0</td>
      <td>15</td>
    </tr>
    <tr>
      <th>AACAACCTACATCTCACG-Sample 2</th>
      <td>0</td>
      <td>8</td>
      <td>7</td>
      <td>11</td>
      <td>4</td>
      <td>11</td>
      <td>27</td>
      <td>5</td>
      <td>3</td>
      <td>12</td>
      <td>...</td>
      <td>9</td>
      <td>5</td>
      <td>4</td>
      <td>37</td>
      <td>4</td>
      <td>20</td>
      <td>25</td>
      <td>13</td>
      <td>0</td>
      <td>12</td>
    </tr>
    <tr>
      <th>AACAACCTAGACGTCAAT-Sample 2</th>
      <td>0</td>
      <td>7</td>
      <td>25</td>
      <td>11</td>
      <td>10</td>
      <td>9</td>
      <td>34</td>
      <td>3</td>
      <td>3</td>
      <td>19</td>
      <td>...</td>
      <td>47</td>
      <td>8</td>
      <td>23</td>
      <td>47</td>
      <td>4</td>
      <td>33</td>
      <td>31</td>
      <td>12</td>
      <td>0</td>
      <td>7</td>
    </tr>
    <tr>
      <th>AACAACCTAGCCTTATCT-Sample 2</th>
      <td>0</td>
      <td>14</td>
      <td>18</td>
      <td>24</td>
      <td>33</td>
      <td>14</td>
      <td>43</td>
      <td>13</td>
      <td>14</td>
      <td>19</td>
      <td>...</td>
      <td>74</td>
      <td>2</td>
      <td>19</td>
      <td>42</td>
      <td>43</td>
      <td>21</td>
      <td>7</td>
      <td>8</td>
      <td>0</td>
      <td>14</td>
    </tr>
    <tr>
      <th>AACAACCTATACCGATTG-Sample 2</th>
      <td>1</td>
      <td>3</td>
      <td>10</td>
      <td>10</td>
      <td>9</td>
      <td>4</td>
      <td>27</td>
      <td>15</td>
      <td>13</td>
      <td>13</td>
      <td>...</td>
      <td>6</td>
      <td>10</td>
      <td>26</td>
      <td>43</td>
      <td>17</td>
      <td>40</td>
      <td>30</td>
      <td>11</td>
      <td>0</td>
      <td>35</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTGTATCACGAGGTGAGC-Sample 2</th>
      <td>2</td>
      <td>6</td>
      <td>48</td>
      <td>10</td>
      <td>16</td>
      <td>13</td>
      <td>27</td>
      <td>20</td>
      <td>22</td>
      <td>26</td>
      <td>...</td>
      <td>1</td>
      <td>12</td>
      <td>17</td>
      <td>52</td>
      <td>7</td>
      <td>15</td>
      <td>4</td>
      <td>7</td>
      <td>0</td>
      <td>68</td>
    </tr>
    <tr>
      <th>TTGTATCACTTATCGCAC-Sample 1</th>
      <td>0</td>
      <td>32</td>
      <td>59</td>
      <td>41</td>
      <td>27</td>
      <td>32</td>
      <td>55</td>
      <td>39</td>
      <td>37</td>
      <td>30</td>
      <td>...</td>
      <td>0</td>
      <td>9</td>
      <td>37</td>
      <td>107</td>
      <td>37</td>
      <td>73</td>
      <td>49</td>
      <td>22</td>
      <td>0</td>
      <td>82</td>
    </tr>
    <tr>
      <th>TTGTCAACCGATGCTCCT-Sample 1</th>
      <td>1</td>
      <td>8</td>
      <td>7</td>
      <td>13</td>
      <td>14</td>
      <td>17</td>
      <td>35</td>
      <td>4</td>
      <td>12</td>
      <td>16</td>
      <td>...</td>
      <td>28</td>
      <td>4</td>
      <td>3</td>
      <td>39</td>
      <td>10</td>
      <td>40</td>
      <td>8</td>
      <td>5</td>
      <td>0</td>
      <td>9</td>
    </tr>
    <tr>
      <th>TTGTTAGAGTATACCACT-Sample 1</th>
      <td>0</td>
      <td>0</td>
      <td>19</td>
      <td>6</td>
      <td>8</td>
      <td>19</td>
      <td>10</td>
      <td>7</td>
      <td>14</td>
      <td>20</td>
      <td>...</td>
      <td>20</td>
      <td>14</td>
      <td>13</td>
      <td>44</td>
      <td>12</td>
      <td>25</td>
      <td>13</td>
      <td>13</td>
      <td>0</td>
      <td>12</td>
    </tr>
    <tr>
      <th>TTGTTAGAGTTCGCATCC-Sample 1</th>
      <td>0</td>
      <td>16</td>
      <td>10</td>
      <td>8</td>
      <td>6</td>
      <td>18</td>
      <td>14</td>
      <td>3</td>
      <td>14</td>
      <td>11</td>
      <td>...</td>
      <td>44</td>
      <td>15</td>
      <td>18</td>
      <td>26</td>
      <td>2</td>
      <td>7</td>
      <td>2</td>
      <td>12</td>
      <td>0</td>
      <td>39</td>
    </tr>
  </tbody>
</table>
<p>3751 rows  312 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">working_amplicons</span> <span class="o">=</span> <span class="p">(</span><span class="n">reads</span><span class="o">.</span><span class="n">median</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[:,</span> <span class="n">working_amplicons</span><span class="p">]</span>

<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(),:]</span>

<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Normalize your amplicons to a WT/known diploid population</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">compute_ploidy</span><span class="p">(</span><span class="n">diploid_cells</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(</span><span class="s1">&#39;Clone 1&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">set_labels</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">set_palette</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">get_palette</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="p-value-chart-per-gene">
<h2>P-value chart per gene<a class="headerlink" href="#p-value-chart-per-gene" title="Permalink to this headline">#</a></h2>
<p>For CNV statistics, this code will allow you to show the pval_cnv chart with gene name labels, instead of amplicon names. Note: if there are multiple amplicons per gene, the gene name will appear for each amplicon, and this is not an aggregate p-val across the entire gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">get_gene_names</span><span class="p">()</span>

<span class="n">pval_cnv</span><span class="p">,</span> <span class="n">tstat_cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">test_signature</span><span class="p">(</span><span class="s2">&quot;normalized_counts&quot;</span><span class="p">)</span>
<span class="n">pval_cnv</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span>
<span class="n">pval_cnv</span> <span class="o">=</span> <span class="n">pval_cnv</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">**</span> <span class="o">-</span><span class="mi">50</span> <span class="o">+</span> <span class="n">pval_cnv</span>
<span class="n">tstat_cnv</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span>
<span class="n">pvals_cnv</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pval_cnv</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">tstat_cnv</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pvals_cnv</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fetching annotation for amplicons.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Added gene names to the &quot;gene_name&quot; column attribute
</pre></div>
</div>
<img alt="../_images/132963fd950b6dc77e76bdd7d82c937cb3b74a9c4aa171d9d67d97e96b416472.png" src="../_images/132963fd950b6dc77e76bdd7d82c937cb3b74a9c4aa171d9d67d97e96b416472.png" />
</div>
</div>
</section>
<section id="calculate-an-average-p-value-per-gene">
<h2>Calculate an average P-value per gene<a class="headerlink" href="#calculate-an-average-p-value-per-gene" title="Permalink to this headline">#</a></h2>
<p>For CNV statistics, this code will allow you to show the average p-value per gene in a heatmap format, instead of at an amplicon level.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#statistics</span>
<span class="n">pval_cnv</span><span class="p">,</span> <span class="n">tstat_cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">test_signature</span><span class="p">(</span><span class="s2">&quot;normalized_counts&quot;</span><span class="p">)</span>

<span class="c1">#replace amplicon IDs by gene names</span>
<span class="n">pval_cnv</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span>

<span class="c1">#transpose the data frame</span>
<span class="n">tpval_cnv</span> <span class="o">=</span> <span class="n">pval_cnv</span><span class="o">.</span><span class="n">T</span>

<span class="c1">#add column header to the index</span>
<span class="n">tpval_cnv</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1">#average the p-values by gene_id and transpose</span>
<span class="n">ttpval_cnv</span> <span class="o">=</span> <span class="n">tpval_cnv</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;gene_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
<span class="n">ttpval_cnv</span> <span class="o">=</span> <span class="n">ttpval_cnv</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">**</span> <span class="o">-</span><span class="mi">50</span> <span class="o">+</span> <span class="n">ttpval_cnv</span>

<span class="c1">#plot it</span>
<span class="n">pvals_cnv</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">ttpval_cnv</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pvals_cnv</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/27bfdfbff447eaf30df25995272932c2c4cb29ba7981ae53735415ad72ac8a7c.png" src="../_images/27bfdfbff447eaf30df25995272932c2c4cb29ba7981ae53735415ad72ac8a7c.png" />
</div>
</div>
</section>
<section id="protein-read-distribution-tool">
<h2>Protein read distribution tool<a class="headerlink" href="#protein-read-distribution-tool" title="Permalink to this headline">#</a></h2>
<p>This code will read in your h5 file, and spit out a table of your antibodies, the raw read counts and what percentage of total counts the antibody accounts for, as well as a pie graph of antibody percentages, or a bar chart.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">read_counts</span><span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;read_counts&quot;</span><span class="p">]</span>
<span class="n">ids</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">raw</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">read_counts</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">ids</span><span class="p">)</span>

<span class="n">total_ab</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">raw</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">skipna</span> <span class="o">=</span> <span class="kc">True</span><span class="p">))</span>
<span class="n">total_ab</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;Raw Count&#39;</span><span class="p">]</span>
<span class="n">total_ab</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Antibody&#39;</span>
<span class="n">total_ab</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">total_ab</span><span class="p">[</span><span class="s1">&#39;percent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">total_ab</span><span class="p">[</span><span class="s1">&#39;Raw Count&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">total_ab</span><span class="p">[</span><span class="s1">&#39;Raw Count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">*</span> <span class="mi">100</span>
<span class="nb">print</span><span class="p">(</span><span class="n">total_ab</span><span class="p">)</span>

<span class="c1">#Pie chart</span>
<span class="c1">#fig = px.pie(total_ab, values=&#39;Raw Count&#39;,names=&#39;Antibody&#39;, title=&#39;Sum of Raw&#39;)</span>

<span class="c1">#Bar chart</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">total_ab</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Antibody&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;percent&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   Antibody  Raw Count   percent
0      CD10     216964  0.378954
1     CD117     343388  0.599770
2     CD11b    1976532  3.452258
3     CD11c    1448022  2.529150
4     CD123    1263044  2.206063
5      CD13     983932  1.718559
6     CD138     590295  1.031023
7      CD14    1280941  2.237322
8     CD141     692292  1.209174
9      CD16    2276106  3.975501
10    CD163     582052  1.016626
11     CD19     456095  0.796626
12     CD1c     332980  0.581591
13      CD2    1588429  2.774388
14     CD22     244551  0.427139
15     CD25     252888  0.441700
16      CD3    4431120  7.739499
17     CD30     494197  0.863176
18    CD303     207141  0.361797
19    CD304     125132  0.218559
20     CD33    1460364  2.550706
21     CD34     184545  0.322331
22     CD38    2048252  3.577526
23      CD4    3998551  6.983964
24     CD44    3832460  6.693866
25     CD45    4025459  7.030962
26   CD45RA    4077514  7.121883
27   CD45RO     643098  1.123250
28    CD49d    2033111  3.551080
29      CD5    1890299  3.301641
30     CD56     801676  1.400226
31    CD62L    1942536  3.392879
32    CD62P     147567  0.257744
33     CD64     502127  0.877027
34     CD69     325517  0.568556
35      CD7    2520817  4.402919
36     CD71     979444  1.710720
37      CD8    3539863  6.182809
38     CD83     721215  1.259691
39     CD90     137773  0.240638
40   FcRI     527347  0.921077
41   HLA-DR     828210  1.446571
42     IgG1     116941  0.204252
43    IgG2a      90132  0.157427
44    IgG2b      92397  0.161383
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">NameError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="nn">Input In [21],</span> in <span class="ni">&lt;cell line: 18&gt;</span><span class="nt">()</span>
<span class="g g-Whitespace">     </span><span class="mi">12</span> <span class="nb">print</span><span class="p">(</span><span class="n">total_ab</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">14</span> <span class="c1">#Pie chart</span>
<span class="g g-Whitespace">     </span><span class="mi">15</span> <span class="c1">#fig = px.pie(total_ab, values=&#39;Raw Count&#39;,names=&#39;Antibody&#39;, title=&#39;Sum of Raw&#39;)</span>
<span class="g g-Whitespace">     </span><span class="mi">16</span> 
<span class="g g-Whitespace">     </span><span class="mi">17</span> <span class="c1">#Bar chart</span>
<span class="ne">---&gt; </span><span class="mi">18</span> <span class="n">fig</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">total_ab</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Antibody&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;percent&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">19</span> <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="ne">NameError</span>: name &#39;px&#39; is not defined
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Normalize protein counts</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">(</span><span class="s1">&#39;CLR&#39;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="plot-a-joint-scatterplot-and-histogram">
<h2>Plot a joint scatterplot and histogram<a class="headerlink" href="#plot-a-joint-scatterplot-and-histogram" title="Permalink to this headline">#</a></h2>
<p>This will create a scatterplot of any two proteins, where additional histograms of the protein expression profile will be shown on the sides of the scatterplots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Plot a joint scatterplot with histograms of two proteins of interest</span>
<span class="c1">#for more: https://seaborn.pydata.org/generated/seaborn.jointplot.html</span>

<span class="c1">#create data frame with normalised protein counts (&quot;protein_df&quot;)</span>
<span class="n">protein_df</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row+col&#39;</span><span class="p">)</span>

<span class="c1">#specify the proteins of choice using the &quot;x&quot; and &quot;y&quot; arguments</span>
<span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;CD3&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;CD19&quot;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">protein_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Create a new sample object for multi-omics visualizations</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">barcodes</span><span class="p">()</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">barcodes</span><span class="p">()</span>

<span class="n">z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>

<span class="c1"># Subset protein data to include only barcodes that were included in both analytes and store in sample.protein.3 variable</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein3</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="p">[</span><span class="n">z</span><span class="p">,</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">()]</span>

<span class="c1"># Subset dna data to include only barcodes that were included in both analytes and store in sample.dna.3 variable</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna3</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="p">[</span><span class="n">z</span><span class="p">,</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">ids</span><span class="p">()]</span> 

<span class="c1">#Update each sub-class (e.g., DNA, CNV, Protein) in order to visualize &quot;clean&quot; data only with relevant features and cells.</span>
<span class="n">sample2</span> <span class="o">=</span> <span class="n">sample</span>
<span class="c1">#sub-assays</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna3</span> 
<span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein3</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="cnv-signature-per-chromosome">
<h2>CNV signature per chromosome<a class="headerlink" href="#cnv-signature-per-chromosome" title="Permalink to this headline">#</a></h2>
<p>This code is for the first heatmap plot in the multi-assay section of the Mosaic notebook (DNA cluster signature vs Analyte and Barcode). It allows you to show average ploidy per chromosome in the CNV section of the heatmap, instead of by amplicon or gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate mean ploidy for each chromosome</span>
<span class="n">amp_ploidy</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="s2">&quot;ploidy&quot;</span><span class="p">,</span> <span class="s2">&quot;rc&quot;</span><span class="p">)</span>
<span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s2">&quot;CHROM&quot;</span><span class="p">]</span>
<span class="n">chrom_ploidy</span> <span class="o">=</span> <span class="n">amp_ploidy</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

<span class="c1"># Create an assay with the ploidy values</span>
<span class="n">chrom_sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[:]</span>
<span class="n">chrom_cnv</span> <span class="o">=</span> <span class="n">chrom_sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[:,</span> <span class="p">:]</span>
<span class="n">chrom_cnv</span><span class="o">.</span><span class="n">select_columns</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">chrom_ploidy</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">chrom_cnv</span><span class="o">.</span><span class="n">col_attrs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">chrom_ploidy</span><span class="o">.</span><span class="n">columns</span><span class="p">)}</span>
<span class="n">chrom_cnv</span><span class="o">.</span><span class="n">layers</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;read_counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">chrom_ploidy</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span>  <span class="c1"># Empty data needed for the heatmap function</span>
    <span class="s2">&quot;normalized_counts&quot;</span><span class="p">:</span> <span class="n">chrom_ploidy</span><span class="o">.</span><span class="n">values</span>  <span class="c1"># The multi sample heatmap plots this layer</span>
<span class="p">}</span>
<span class="n">chrom_sample</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">chrom_cnv</span>

<span class="c1"># Draw normal heatmap</span>
<span class="n">chrom_sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s2">&quot;normalized_counts&quot;</span><span class="p">)</span>

<span class="c1"># Draw multi-assay heatmap</span>
<span class="n">chrom_sample</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">flatten</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="save-output-for-one-clone">
<h2>Save output for one clone<a class="headerlink" href="#save-output-for-one-clone" title="Permalink to this headline">#</a></h2>
<p>This code will output a filtered h5 file, containing only cells of the clone of interest, which can be used for further analysis or filtering.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#save one clone (or one sample) as a separate .h5 file</span>
<span class="c1">#row attribute &#39;label&#39; stores the group (e.g. clone) information</span>
<span class="c1">#set the sampleID to match the clone that you would like to extract from the others</span>
<span class="n">sampleID</span> <span class="o">=</span> <span class="s1">&#39;Clone 2&#39;</span>

<span class="c1">#create a filter based on sampleID</span>
<span class="n">filt</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">sampleID</span>

<span class="c1">#create a new dna layer under the same sample</span>
<span class="c1">#sample.dna2 contains only the barcodes from the selected group</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">[</span><span class="n">filt</span><span class="p">,</span> <span class="p">:]</span>

<span class="c1">#create a new sample</span>
<span class="n">sample2</span> <span class="o">=</span> <span class="n">sample</span>

<span class="c1">#transfer the layer sample.dna2 to the new sample</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span>

<span class="c1">#save sample2 to disk</span>
<span class="n">ms</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">sample2</span><span class="p">,</span> <span class="s2">&quot;Clone2_FilteredOnly.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>


              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="multi-sample-analysis.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Multisample analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="analysis-walkthrough.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Mosaic analysis walk-through</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Saurabh Parikh<br/>
  
      &copy; Copyright 2021, Mission Bio.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>