

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Changelog &#8212; Mosaic v3.12.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'manual/changelog';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="_Assay" href="../pages/missionbio.mosaic.assay._Assay.html" />
    <link rel="prev" title="Additional resources" href="help.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Mosaic v3.12.2 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="data_structure.html">Data structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="cohort.html">Cohort Analysis Pipeline</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="vignettes.html">Vignettes</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/overview.html">Overview of mosaic</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/filtering.html">Filtering barcodes and ids</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/cnv.html">Manual CNV analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/plot-customizations.html">Plot customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/multi-sample-analysis.html">Multisample analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/additional-plots-and-customizations.html">Additional plots and customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/code-snippets.html">Code Snippets</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="curated_jupyter_notebooks.html">Curated Jupyter Notebooks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna.html">Single Sample DNA-only</a></li>






<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-multisample.html">Multisample DNA-only</a></li>






<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-protein.html">Single Sample DNA + Protein</a></li>







<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-protein-multisample.html">Multisample DNA + Protein</a></li>







</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Additional resources</a></li>

<li class="toctree-l1 current active"><a class="current reference internal" href="#">Changelog</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Classes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.assay._Assay.html">_Assay</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.dna.Dna.html">Dna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.cnv.Cnv.html">Cnv</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.protein.Protein.html">Protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.sample.Sample.html">Sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.samplegroup.SampleGroup.html">SampleGroup</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interactive Workflows</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.workflows.copy_number.CopyNumber.html">CopyNumber</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html">VariantSubcloneTable</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Algorithms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.nsp.NSP.html">NSP</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.algorithms.group_by_genotype.GroupByGenotype.html">GroupByGenotype</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.algorithms.compass.COMPASS.html">COMPASS</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Functional Modules</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.io.html">io</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.utils.html">utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.constants.html">constants</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Custom Plots</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.heatmap.Heatmap.html">Heatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.lineplot.LinePlot.html">LinePlot</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.bargraph.BarGraph.html">BarGraph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.multimap.MultiMap.html">MultiMap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.phylotree.PhyloTree.html">PhyloTree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.fishplot.Fishplot.html">Fishplot</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">DNA assignment</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.dna.truth.Truth.html">Truth</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.dna.likelihood.LikelihoodMethod.html">LikelihoodMethod</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.algorithms.dna_assignment.DnaAssignment.html">DnaAssignment</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Protein assignment</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.truth.Truth.html">Truth</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.pace.likelihood.LikelihoodMethod.html">LikelihoodMethod</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.pace.probability.ProbabilityMethod.html">ProbabilityMethod</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/manual/changelog.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Changelog</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-12-0">v3.12.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#added">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fixed">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-7-0">v3.7.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#changed">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-4-0">v3.4.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-1-1">v3.1.1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-1-0">v3.1.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id9">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id11">Fixed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#updated">Updated</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-0-1">v3.0.1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id14">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id15">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-0-0">v3.0.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id16">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id17">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id18">Fixed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#removed">Removed</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
    <!--Add a white cover to hide the .bd-container grey on the right side of the page.-->
    <div class="background-cover" id="background-cover"></div>
    <section id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this heading">#</a></h1>
<section id="v3-12-0">
<h2>v3.12.0<a class="headerlink" href="#v3-12-0" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2025-21-04</p>
<section id="added">
<h3>Added<a class="headerlink" href="#added" title="Permalink to this heading">#</a></h3>
<p>A new attribute for all assays: <code class="docutils literal notranslate"><span class="pre">info</span></code>. This can be used to store arbitrary information of various types including pandas dataframes and dictionaries.
Unlike row attributes, column attributes, and layers, it is not confined to any shape. For example, the signature of the clones generated for only
the somatic variants would have fewer columns than ids and fewer rows than cells in the assay. Compass stores a lot information in it’s variables,
until now there was no easy way of saving it in the h5. These are dataframes with shapes that do not align with that of the assay. These values can
now be stored in the assay <code class="docutils literal notranslate"><span class="pre">info</span></code> as follows:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_info</span><span class="p">(</span><span class="s2">&quot;somatic_signature&quot;</span><span class="p">,</span> <span class="n">compass</span><span class="o">.</span><span class="n">node_genotypes_</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_info</span><span class="p">(</span><span class="s2">&quot;assignment_probability&quot;</span><span class="p">,</span> <span class="n">compass</span><span class="o">.</span><span class="n">probability_</span><span class="p">)</span>
</pre></div>
</div>
<p>The assays also now store their palette in the info as a dictionary. <a class="reference internal" href="../pages/attributes/missionbio.mosaic.assay._Assay.sampleinfo.html#missionbio.mosaic.assay._Assay.sampleinfo" title="missionbio.mosaic.assay._Assay.sampleinfo"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sampleinfo()</span></code></a> is a shortcut for accessing info for single sample assays.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">sampleinfo</span><span class="p">[</span><span class="s2">&quot;palette&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>The functions for handling info are <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.has_info.html#missionbio.mosaic.assay._Assay.has_info" title="missionbio.mosaic.assay._Assay.has_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">has_info()</span></code></a>, <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.add_info.html#missionbio.mosaic.assay._Assay.add_info" title="missionbio.mosaic.assay._Assay.add_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_info()</span></code></a>, and <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.del_info.html#missionbio.mosaic.assay._Assay.del_info" title="missionbio.mosaic.assay._Assay.del_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">del_info()</span></code></a></p>
<p><strong>Updates to COMPASS</strong></p>
<ul class="simple">
<li><p>Update to the latest version of <a class="reference external" href="https://github.com/MissionBio/compass/tree/e3e0169bbe52555b751752699cb22399c1b70763">COMPASS</a></p></li>
<li><p>Add option to rename compass clones with a new probability value using <a class="reference internal" href="../pages/methods/missionbio.mosaic.algorithms.compass.COMPASS.relabel.html#missionbio.mosaic.algorithms.compass.COMPASS.relabel" title="missionbio.mosaic.algorithms.compass.COMPASS.relabel"><code class="xref py py-meth docutils literal notranslate"><span class="pre">relabel()</span></code></a>.</p></li>
<li><p>Allow passing prefixes for file names to <a class="reference internal" href="../pages/methods/missionbio.mosaic.algorithms.compass.COMPASS.run.html#missionbio.mosaic.algorithms.compass.COMPASS.run" title="missionbio.mosaic.algorithms.compass.COMPASS.run"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run()</span></code></a></p></li>
<li><p>Added option to get fractions of each clone using <a class="reference internal" href="../pages/methods/missionbio.mosaic.algorithms.compass.COMPASS.node_fractions.html#missionbio.mosaic.algorithms.compass.COMPASS.node_fractions" title="missionbio.mosaic.algorithms.compass.COMPASS.node_fractions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">node_fractions()</span></code></a></p></li>
<li><p>Allow passing directories to <a class="reference internal" href="../pages/methods/missionbio.mosaic.algorithms.compass.COMPASS.run.html#missionbio.mosaic.algorithms.compass.COMPASS.run" title="missionbio.mosaic.algorithms.compass.COMPASS.run"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run()</span></code></a></p></li>
<li><p>COMPASS can call LOH clones without calling CNV using the <code class="docutils literal notranslate"><span class="pre">cnv</span></code> parameter in <a class="reference internal" href="../pages/methods/missionbio.mosaic.algorithms.compass.COMPASS.run.html#missionbio.mosaic.algorithms.compass.COMPASS.run" title="missionbio.mosaic.algorithms.compass.COMPASS.run"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run()</span></code></a>.</p></li>
</ul>
<p><strong>Plotting updates</strong></p>
<ul class="simple">
<li><p>Cytoband information added to the CNV ploidy <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.signaturemap.html#missionbio.mosaic.cnv.Cnv.signaturemap" title="missionbio.mosaic.cnv.Cnv.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">signaturemap()</span></code></a> and <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.plot_ploidy.html#missionbio.mosaic.cnv.Cnv.plot_ploidy" title="missionbio.mosaic.cnv.Cnv.plot_ploidy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">plot_ploidy()</span></code></a> figures. These are also shown in the CopyNumber workflow.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ticks</span></code> option to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">heatmap()</span></code></a> to disable plotting of ticks for assay with large number of ids.</p></li>
<li><p>The scatterplot size modifies depending on the length of the labels to ensure that the plotting area is constant.</p></li>
</ul>
<p><strong>Updates to CNV</strong></p>
<ul class="simple">
<li><p>Locally stored gene name annotations are used instead of pulling them from Ensembl when running <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.get_gene_names.html#missionbio.mosaic.cnv.Cnv.get_gene_names" title="missionbio.mosaic.cnv.Cnv.get_gene_names"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_gene_names()</span></code></a>.</p></li>
<li><p>When using <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.get_gene_names.html#missionbio.mosaic.cnv.Cnv.get_gene_names" title="missionbio.mosaic.cnv.Cnv.get_gene_names"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_gene_names()</span></code></a>, the gene name will always be the best match for the amplicon instead of mutliple values separated by a <code class="docutils literal notranslate"><span class="pre">/</span></code></p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.get_gene_names.html#missionbio.mosaic.cnv.Cnv.get_gene_names" title="missionbio.mosaic.cnv.Cnv.get_gene_names"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_gene_names()</span></code></a> returns the gene names while also adding it to the column attributes.</p></li>
</ul>
<p><strong>Updates to DNA</strong></p>
<ul class="simple">
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.get_annotated_ids.html#missionbio.mosaic.dna.Dna.get_annotated_ids" title="missionbio.mosaic.dna.Dna.get_annotated_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_annotated_ids()</span></code></a> returns a list of human-readable names for all the variants in the assay.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.set_annotated_ids.html#missionbio.mosaic.dna.Dna.set_annotated_ids" title="missionbio.mosaic.dna.Dna.set_annotated_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_annotated_ids()</span></code></a> updates the ids to be human-readable names.</p></li>
<li><p><a class="reference internal" href="../pages/attributes/missionbio.mosaic.dna.Dna.genome.html#missionbio.mosaic.dna.Dna.genome" title="missionbio.mosaic.dna.Dna.genome"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.genome()</span></code></a> to easily access the <cite>genome_version</cite> metadata value.</p></li>
<li><p>Store annotations in the DNA assay <cite>info</cite> instead of the column attributes to ease their access.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.snps.html#missionbio.mosaic.dna.Dna.snps" title="missionbio.mosaic.dna.Dna.snps"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.snps()</span></code></a> to quickly get the ids that are SNPs.</p></li>
</ul>
<p><strong>Functionality updates</strong></p>
<ul class="simple">
<li><p>Option to disable matching of ids in <a class="reference internal" href="../pages/missionbio.mosaic.samplegroup.SampleGroup.html#missionbio.mosaic.samplegroup.SampleGroup" title="missionbio.mosaic.samplegroup.SampleGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleGroup</span></code></a> using the <code class="docutils literal notranslate"><span class="pre">match_ids</span></code> parameter.</p></li>
<li><p>Ability to load h5 files with only raw counts.</p></li>
<li><p><cite>whitelist=[]</cite> is treated just like <cite>whitelist=None</cite> in <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">load()</span></code></a></p></li>
<li><p>Deduplication of barcodes is done using integers instead of sample names and no warning is raised when deduplication of barcodes is performed.
It can be manually performed using <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.deduplicate_barcodes.html#missionbio.mosaic.assay._Assay.deduplicate_barcodes" title="missionbio.mosaic.assay._Assay.deduplicate_barcodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">deduplicate_barcodes()</span></code></a> and inverted using <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.normalize_barcodes.html#missionbio.mosaic.assay._Assay.normalize_barcodes" title="missionbio.mosaic.assay._Assay.normalize_barcodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">normalize_barcodes()</span></code></a>.</p></li>
<li><p>Store the palette in the <cite>info</cite> instead of the metadata of the assay, making saved h5 files valid in format even with the stored palette.</p></li>
<li><p><code class="xref py py-attr docutils literal notranslate"><span class="pre">sample.Sample.vdj</span></code> to easily access the VDJ assay in the h5 file.</p></li>
</ul>
</section>
<section id="fixed">
<h3>Fixed<a class="headerlink" href="#fixed" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>ADO score shown in the figure is the same as the ADO score shown in the table in the <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSubcloneTable</span></code></a> workflow</p></li>
<li><p>Use the correct genome version from the dna assay in <a class="reference internal" href="../pages/missionbio.mosaic.algorithms.compass.COMPASS.html#missionbio.mosaic.algorithms.compass.COMPASS" title="missionbio.mosaic.algorithms.compass.COMPASS"><code class="xref py py-class docutils literal notranslate"><span class="pre">COMPASS</span></code></a></p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.get_annotations.html#missionbio.mosaic.dna.Dna.get_annotations" title="missionbio.mosaic.dna.Dna.get_annotations"><code class="xref py py-class docutils literal notranslate"><span class="pre">get_annotations</span></code></a> raises an appropriate error when the genome version is not supported.</p></li>
</ul>
</section>
</section>
<section id="v3-7-0">
<h2>v3.7.0<a class="headerlink" href="#v3-7-0" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2024-08-05</p>
<section id="id1">
<h3>Added<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.filter_somatic_variants.html#missionbio.mosaic.dna.Dna.filter_somatic_variants" title="missionbio.mosaic.dna.Dna.filter_somatic_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_somatic_variants()</span></code></a> for automatic filtering of pathogenic somatic variants.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.assign_from_truth.html#missionbio.mosaic.dna.Dna.assign_from_truth" title="missionbio.mosaic.dna.Dna.assign_from_truth"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.assign_from_truth()</span></code></a> to label the cells for a known set of clones.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.cluster_and_label.html#missionbio.mosaic.protein.Protein.cluster_and_label" title="missionbio.mosaic.protein.Protein.cluster_and_label"><code class="xref py py-meth docutils literal notranslate"><span class="pre">protein.Protein.cluster_and_label()</span></code></a> to find all protein clusters and label them based on the provided truth. This function can be used to novel cell types.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.label_sticky_cells.html#missionbio.mosaic.protein.Protein.label_sticky_cells" title="missionbio.mosaic.protein.Protein.label_sticky_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">protein.Protein.label_sticky_cells()</span></code></a> to mark cells which are likely to be sticky.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.assign_from_truth.html#missionbio.mosaic.protein.Protein.assign_from_truth" title="missionbio.mosaic.protein.Protein.assign_from_truth"><code class="xref py py-meth docutils literal notranslate"><span class="pre">protein.Protein.assign_from_truth()</span></code></a> label the cells for a known set of cell types. By default, it labels the PBMC subtypes.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.truth.html#missionbio.mosaic.protein.Protein.truth" title="missionbio.mosaic.protein.Protein.truth"><code class="xref py py-meth docutils literal notranslate"><span class="pre">protein.Protein.truth()</span></code></a> to convert cluster signatures to a truth that can be used for <a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.assign_from_truth.html#missionbio.mosaic.protein.Protein.assign_from_truth" title="missionbio.mosaic.protein.Protein.assign_from_truth"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assign_from_truth()</span></code></a></p></li>
<li><p>Ability to pass an external control to <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.compute_ploidy.html#missionbio.mosaic.cnv.Cnv.compute_ploidy" title="missionbio.mosaic.cnv.Cnv.compute_ploidy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compute_ploidy()</span></code></a></p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.read_depth_dependence.html#missionbio.mosaic.protein.Protein.read_depth_dependence" title="missionbio.mosaic.protein.Protein.read_depth_dependence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read_depth_dependence()</span></code></a> - a plot to quickly visualize the need and effectiveness of NSP normalization.</p></li>
<li><p>Option to load a subset of the assays in the h5 file.</p></li>
<li><p>No error is raised when h5 files with unknown assays are loaded.</p></li>
<li><p>Raise an error when the number of variants to annotate is more than 1,000. This is a safeguard to prevent incorrect API calls.</p></li>
<li><p>Varsome annotations are stored locally and will not be fetched again unless the local file is deleted.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">copy</span></code> parameter to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.get_attribute.html#missionbio.mosaic.assay._Assay.get_attribute" title="missionbio.mosaic.assay._Assay.get_attribute"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_attribute()</span></code></a> to return a view of the data instead of a copy.</p></li>
<li><p>Option to pass <code class="docutils literal notranslate"><span class="pre">order</span></code> of labels to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.ridgeplot.html#missionbio.mosaic.assay._Assay.ridgeplot" title="missionbio.mosaic.assay._Assay.ridgeplot"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ridgeplot()</span></code></a>.</p></li>
<li><p>ADO score is formatted more conveniently in the <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">workflows.variant_subclone_table.VariantSubcloneTable</span></code></a> workflow. If it’s 0, then it’s shown as “-” and if &lt;0.05 then it’s shown as “~0.0”.</p></li>
<li><p>Ability to rename a sample using <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.rename.html#missionbio.mosaic.sample.Sample.rename" title="missionbio.mosaic.sample.Sample.rename"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rename()</span></code></a>.</p></li>
</ul>
</section>
<section id="changed">
<h3>Changed<a class="headerlink" href="#changed" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="../pages/attributes/missionbio.mosaic.sample.Sample.name.html#missionbio.mosaic.sample.Sample.name" title="missionbio.mosaic.sample.Sample.name"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sample.Sample.name</span></code></a> is a now a property, and cannot be set. It returns a value according to the current <code class="docutils literal notranslate"><span class="pre">sample_name</span></code> metadata.</p></li>
<li><p><a class="reference internal" href="../pages/attributes/missionbio.mosaic.assay._Assay.title.html#missionbio.mosaic.assay._Assay.title" title="missionbio.mosaic.assay._Assay.title"><code class="xref py py-attr docutils literal notranslate"><span class="pre">assay._Assay.title</span></code></a> is a now a property, and cannot be set. It returns a value according to the current <cite>sample_name</cite> metadata.</p></li>
<li><p>Behavior of <code class="docutils literal notranslate"><span class="pre">default_label</span></code> in <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.set_labels.html#missionbio.mosaic.assay._Assay.set_labels" title="missionbio.mosaic.assay._Assay.set_labels"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.set_labels()</span></code></a>. When <code class="docutils literal notranslate"><span class="pre">default_label</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>, only the labels of the provided barcodes are updated.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">normalized_counts</span></code> in <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.compute_ploidy.html#missionbio.mosaic.cnv.Cnv.compute_ploidy" title="missionbio.mosaic.cnv.Cnv.compute_ploidy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compute_ploidy()</span></code></a> is no longer used. The <code class="docutils literal notranslate"><span class="pre">read_counts</span></code> layer is used directly.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ANNOTATION_COLUMNS</span></code> constant was moved to  <code class="docutils literal notranslate"><span class="pre">missionbio.annotation.constants</span></code></p></li>
<li><p>Use <code class="docutils literal notranslate"><span class="pre">pynndescent</span></code> instead of <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code> to speed up nearest neighbors calculation during graph-community clustering. Results will not be backwards compatible.</p></li>
</ul>
</section>
<section id="id2">
<h3>Fixed<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Ordering of the barcodes in the heatmap when a subset of the variants are used.</p></li>
<li><p>Fetching of CNV amplicon gene names for regions where ensembl returns an incomplete response.</p></li>
<li><p>Allow custom grouping of amplicons for <code class="xref py py-meth docutils literal notranslate"><span class="pre">cnv.CNV.heatmap()</span></code> by passing amplicons to <code class="docutils literal notranslate"><span class="pre">features</span></code> and <code class="docutils literal notranslate"><span class="pre">x_groups</span></code> values.</p></li>
</ul>
</section>
</section>
<section id="v3-4-0">
<h2>v3.4.0<a class="headerlink" href="#v3-4-0" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2024-04-01</p>
<section id="id3">
<h3>Added<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Support to pass <code class="docutils literal notranslate"><span class="pre">x_groups</span></code> to <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.signaturemap.html#missionbio.mosaic.sample.Sample.signaturemap" title="missionbio.mosaic.sample.Sample.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">signaturemap()</span></code></a> and <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.heatmap.html#missionbio.mosaic.sample.Sample.heatmap" title="missionbio.mosaic.sample.Sample.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">heatmap()</span></code></a>.</p></li>
<li><p>Support to pass variant filters to <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">load()</span></code></a>.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.positions.html#missionbio.mosaic.cnv.Cnv.positions" title="missionbio.mosaic.cnv.Cnv.positions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">positions()</span></code></a>, <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.amplicon_performance.html#missionbio.mosaic.cnv.Cnv.amplicon_performance" title="missionbio.mosaic.cnv.Cnv.amplicon_performance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">amplicon_performance()</span></code></a>, &amp; <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.panel_uniformity.html#missionbio.mosaic.cnv.Cnv.panel_uniformity" title="missionbio.mosaic.cnv.Cnv.panel_uniformity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">panel_uniformity()</span></code></a> to quickly get amplicon positions, performance and panel uniformity.</p></li>
<li><p>Option to hide columns in the variants table of the <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSubcloneTable</span></code></a> workflow.</p></li>
<li><p>Ability to filter variants through the GUI in the <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSubcloneTable</span></code></a> workflow.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">override</span></code> parameter for the <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">heatmap()</span></code></a> function which is simply passed to <code class="docutils literal notranslate"><span class="pre">clustered_ids</span></code> and <code class="docutils literal notranslate"><span class="pre">clustered_barcodes</span></code>.</p></li>
<li><p>The first column of the subclone table is frozen.</p></li>
<li><p>Mandate <code class="docutils literal notranslate"><span class="pre">features</span></code> when <code class="docutils literal notranslate"><span class="pre">x_groups</span></code> is provided in <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">heatmap()</span></code></a>.</p></li>
<li><p>An appropriate error is raised when any cell has 0 total reads when running <code class="docutils literal notranslate"><span class="pre">NSP</span></code>.</p></li>
<li><p>An appropriate error is raised when the annotation API is not available.</p></li>
</ul>
</section>
<section id="id4">
<h3>Changed<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Increased the vertical spacing between the graph and the fishplot from 0 to 0.1.</p></li>
<li><p>The plotting functions in <code class="docutils literal notranslate"><span class="pre">missionbio.mosaic.plotting</span></code> were moved to <code class="docutils literal notranslate"><span class="pre">missionbio.plotting</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">missionbio.algorithms.nsp</span></code> was moved to <code class="docutils literal notranslate"><span class="pre">missionbio.demultiplex.protein.nsp</span></code></p></li>
<li><p>Unpinned <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code> and <code class="docutils literal notranslate"><span class="pre">hdbscan</span></code> as their latest versions are compatible with each other.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">scikit-learn&gt;1.3.1</span></code> is installed by default which results in slightly different NSP calls due to changes to its Gaussian mixture model.</p></li>
</ul>
</section>
<section id="id5">
<h3>Fixed<a class="headerlink" href="#id5" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Load the whitelist variants correctly when <code class="docutils literal notranslate"><span class="pre">filter_variants=True</span></code> is passed to <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">load()</span></code></a>.</p></li>
<li><p>Nill values of DANN score are shown as empty cells instead of <code class="docutils literal notranslate"><span class="pre">º</span></code>.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.name_id_by_pos.html#missionbio.mosaic.cnv.Cnv.name_id_by_pos" title="missionbio.mosaic.cnv.Cnv.name_id_by_pos"><code class="xref py py-meth docutils literal notranslate"><span class="pre">name_id_by_pos()</span></code></a> does not filter the amplicons.</p></li>
<li><p>Lineplot in <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.plot_ploidy.html#missionbio.mosaic.cnv.Cnv.plot_ploidy" title="missionbio.mosaic.cnv.Cnv.plot_ploidy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">plot_ploidy()</span></code></a> does not connect the medians with a line when using <code class="docutils literal notranslate"><span class="pre">genes+amplicons</span></code> or <code class="docutils literal notranslate"><span class="pre">positions+amplicons</span></code>.</p></li>
<li><p>The violin plot range is fixed to (0, 100) for the <code class="docutils literal notranslate"><span class="pre">AF</span></code> and <code class="docutils literal notranslate"><span class="pre">GQ</span></code> layers in <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSubcloneTable</span></code></a>.</p></li>
<li><p>Violin plots generated using <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.violinplot.html#missionbio.mosaic.assay._Assay.violinplot" title="missionbio.mosaic.assay._Assay.violinplot"><code class="xref py py-meth docutils literal notranslate"><span class="pre">violinplot()</span></code></a> are equally spaced when split by labels.</p></li>
<li><p>Fix resetting of <code class="docutils literal notranslate"><span class="pre">selected_bars</span></code> when scatterplots are created.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.rename_labels.html#missionbio.mosaic.assay._Assay.rename_labels" title="missionbio.mosaic.assay._Assay.rename_labels"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rename_labels()</span></code></a> allows swapping of labels.</p></li>
<li><p>Fishplot does not disappear when a clone and its parent both have 0 cells at some timepoint.</p></li>
</ul>
</section>
</section>
<section id="v3-1-1">
<h2>v3.1.1<a class="headerlink" href="#v3-1-1" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2023-09-25</p>
<section id="id6">
<h3>Added<a class="headerlink" href="#id6" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Relaxed missionbio.h5 requirement to &gt;=4.13.0,&lt;6</p></li>
</ul>
</section>
<section id="id7">
<h3>Changed<a class="headerlink" href="#id7" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Disable autouploading of tagged packages to anaconda.</p></li>
<li><p>Removed check for h5 file compatibility with H5Reader.</p></li>
</ul>
</section>
<section id="id8">
<h3>Fixed<a class="headerlink" href="#id8" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">whitelist</span></code> option in <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">load()</span></code></a> correctly loads exact matches of variants.</p></li>
</ul>
</section>
</section>
<section id="v3-1-0">
<h2>v3.1.0<a class="headerlink" href="#v3-1-0" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2023-09-13</p>
<section id="id9">
<h3>Added<a class="headerlink" href="#id9" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>The order of the names in the legend matches the order of the traces in the ridgeplot.</p></li>
<li><p>Option to pass any sequence type to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.get_attribute.html#missionbio.mosaic.assay._Assay.get_attribute" title="missionbio.mosaic.assay._Assay.get_attribute"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_attribute()</span></code></a> besides np.ndarray. This includes list, tuple, and range.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">features</span></code> parameter to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.signature.html#missionbio.mosaic.assay._Assay.signature" title="missionbio.mosaic.assay._Assay.signature"><code class="xref py py-meth docutils literal notranslate"><span class="pre">signature()</span></code></a> which allows grouping across ids, just like <code class="docutils literal notranslate"><span class="pre">splitby</span></code> allows grouping across cells.
* The <code class="docutils literal notranslate"><span class="pre">feautures</span></code> option in <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.signaturemap.html#missionbio.mosaic.assay._Assay.signaturemap" title="missionbio.mosaic.assay._Assay.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">signaturemap()</span></code></a> allows plotting using grouped data from <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.signature.html#missionbio.mosaic.assay._Assay.signature" title="missionbio.mosaic.assay._Assay.signature"><code class="xref py py-meth docutils literal notranslate"><span class="pre">signature()</span></code></a></p></li>
<li><p>Support for hg38 along with all species available through Ensembl in <code class="xref py py-meth docutils literal notranslate"><span class="pre">get_annotations()</span></code></p></li>
<li><p>Support for hg38 in <a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.get_annotations.html#missionbio.mosaic.dna.Dna.get_annotations" title="missionbio.mosaic.dna.Dna.get_annotations"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_annotations()</span></code></a>.</p></li>
<li><p>Sped up NSP by 2x by using <code class="docutils literal notranslate"><span class="pre">statsmodels</span></code> for the KDE and using spherical covariance with kmeans++ initialization for the GMM parameters.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ANSP</span></code> - Approximate NSP to protein normalization. It runs in constant time for large datasets.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.get_attribute.html#missionbio.mosaic.assay._Assay.get_attribute" title="missionbio.mosaic.assay._Assay.get_attribute"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_attribute()</span></code></a> also accepts dataframes.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">heatmap()</span></code></a> can plot arbitrary dataframes as long as it has the expected number of cells.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TreeGraph</span></code> now supports html tags like <code class="docutils literal notranslate"><span class="pre">&lt;br&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;b&gt;</span></code>, and <code class="docutils literal notranslate"><span class="pre">&lt;span&gt;</span></code> in the descriptions.</p></li>
</ul>
</section>
<section id="id10">
<h3>Changed<a class="headerlink" href="#id10" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Use latest python 3.8 in installer instead of 3.8.0</p></li>
</ul>
</section>
<section id="id11">
<h3>Fixed<a class="headerlink" href="#id11" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>The title of <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.clone_vs_analyte.html#missionbio.mosaic.sample.Sample.clone_vs_analyte" title="missionbio.mosaic.sample.Sample.clone_vs_analyte"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clone_vs_analyte()</span></code></a> plot does not overlap with the DNA heatmap.</p></li>
<li><p>The x-axis label order for CNV in the <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.clone_vs_analyte.html#missionbio.mosaic.sample.Sample.clone_vs_analyte" title="missionbio.mosaic.sample.Sample.clone_vs_analyte"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clone_vs_analyte()</span></code></a> plot matches the order of the points in the data shown.</p></li>
<li><p>NGT layer not modified after running <a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.filter_variants.html#missionbio.mosaic.dna.Dna.filter_variants" title="missionbio.mosaic.dna.Dna.filter_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_variants()</span></code></a></p></li>
<li><p>“Last modified” timestamp does not change when loading an H5 file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">jitter</span></code> parameter in <code class="docutils literal notranslate"><span class="pre">NSP</span></code> works</p></li>
<li><p>Failure of <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSubcloneTable</span></code></a> when all the variant calls are filtered.</p></li>
<li><p>Pinned hdbscan to v0.8.29. Higher versions (&gt;=0.8.30,&lt;=0.8.33) have runtime issues.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.heatmap.html#missionbio.mosaic.sample.Sample.heatmap" title="missionbio.mosaic.sample.Sample.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">heatmap()</span></code></a> and <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.signaturemap.html#missionbio.mosaic.sample.Sample.signaturemap" title="missionbio.mosaic.sample.Sample.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">signaturemap()</span></code></a> execute successfully when “cnv” is passed before “dna”.</p></li>
<li><p>Fix y-compression of <code class="docutils literal notranslate"><span class="pre">TreeGraph</span></code> by checking the upwards and downwards movement of only the highest and lowest nodes respectively.</p></li>
</ul>
</section>
<section id="updated">
<h3>Updated<a class="headerlink" href="#updated" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Switched from using the depracated JupyterDash to the builtin jupyter dash in Dash v2.11. <a class="reference external" href="https://dash.plotly.com/dash-in-jupyter">Documentation</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">jupyter_client</span></code> from &lt;8 to &gt;=8.1.0 as the ThreadedZMQStream error is fixed in it. <a class="reference external" href="https://jupyter-client.readthedocs.io/en/stable/changelog.html#id6">Changelog</a></p></li>
</ul>
</section>
</section>
<section id="v3-0-1">
<h2>v3.0.1<a class="headerlink" href="#v3-0-1" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2023-06-20</p>
<section id="id13">
<h3>Added<a class="headerlink" href="#id13" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.crosstab.html#missionbio.mosaic.assay._Assay.crosstab" title="missionbio.mosaic.assay._Assay.crosstab"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.crosstab()</span></code></a> to wrap <code class="docutils literal notranslate"><span class="pre">pandas.crosstab</span></code> for ease of use with mosaic.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.crosstabmap.html#missionbio.mosaic.assay._Assay.crosstabmap" title="missionbio.mosaic.assay._Assay.crosstabmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.crosstabmap()</span></code></a> to create heatmaps of the output of <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.crosstab.html#missionbio.mosaic.assay._Assay.crosstab" title="missionbio.mosaic.assay._Assay.crosstab"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.crosstab()</span></code></a>.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.hierarchical_cluster.html#missionbio.mosaic.assay._Assay.hierarchical_cluster" title="missionbio.mosaic.assay._Assay.hierarchical_cluster"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.hierarchical_cluster()</span></code></a> to get the hierarchical clustering order of the rows of a DataFrame.</p></li>
</ul>
</section>
<section id="id14">
<h3>Changed<a class="headerlink" href="#id14" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated matplotlib dependency from <code class="docutils literal notranslate"><span class="pre">&lt;=3.2.2</span></code> to <code class="docutils literal notranslate"><span class="pre">&gt;=3.4.0</span></code></p></li>
</ul>
</section>
<section id="id15">
<h3>Fixed<a class="headerlink" href="#id15" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.heatmap()</span></code></a> subclustering performed when <cite>convolve=0</cite>. It was disabled by default.</p></li>
<li><p>Custom <code class="docutils literal notranslate"><span class="pre">typography.css</span></code> used in workflows is included in the package data</p></li>
<li><p>Setting labels using dictionaries in <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.set_labels.html#missionbio.mosaic.assay._Assay.set_labels" title="missionbio.mosaic.assay._Assay.set_labels"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.set_labels()</span></code></a>.</p></li>
</ul>
</section>
</section>
<section id="v3-0-0">
<h2>v3.0.0<a class="headerlink" href="#v3-0-0" title="Permalink to this heading">#</a></h2>
<p><strong>Release date</strong>: 2023-06-16</p>
<section id="id16">
<h3>Added<a class="headerlink" href="#id16" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>A wrapper for COMPASS.</p></li>
<li><p>New variant filters that account for missing data.</p></li>
<li><p>Recipe and instructions for building installers.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">plot_kind</span></code> parameter to <a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.group_by_genotype.html#missionbio.mosaic.dna.Dna.group_by_genotype" title="missionbio.mosaic.dna.Dna.group_by_genotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.group_by_genotype()</span></code></a> to change the type of plot shown.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">filter_cells</span></code> to <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">io.load()</span></code></a> which loads only the intersection algorithm cells.</p></li>
<li><p>Progress bar to <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">io.load()</span></code></a></p></li>
<li><p><code class="xref py py-class docutils literal notranslate"><span class="pre">algorithms.nsp.NSP</span></code> and <code class="xref py py-class docutils literal notranslate"><span class="pre">algorithms.nsp.ExpressionProfile</span></code> to modularize the NSP code.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">x_groups</span></code> to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.heatmap()</span></code></a> to group the x-axis by a given list of ids.</p></li>
<li><p>Simplify and speedup <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.heatmap()</span></code></a> by removing duplicate data. (By using <code class="xref py py-class docutils literal notranslate"><span class="pre">plots.heatmap.Heatmap</span></code>)</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.convolve.html#missionbio.mosaic.assay._Assay.convolve" title="missionbio.mosaic.assay._Assay.convolve"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.convolve()</span></code></a> to convolve the data that was earlier performed in the Heatmap.</p></li>
<li><p>Configuration options accessible via <code class="xref py py-class docutils literal notranslate"><span class="pre">Config</span></code>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">ms.Config.Colorscale.Dna</span></code> to change the default color palette for all DNA plots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ms.Config.Colorscale.Cnv</span></code> to change the default color palette for all CNV plots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ms.Config.Colorscale.Protein</span></code> to change the default color palette for all Protein plots.</p></li>
</ul>
</li>
<li><p>Custom divirgent colorscale for Cnv Ploidy heatmaps</p></li>
<li><p>Option to return indices instead of barcodes in <code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Asasy.clustered_barcodes()</span></code>.</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.common_barcodes.html#missionbio.mosaic.sample.Sample.common_barcodes" title="missionbio.mosaic.sample.Sample.common_barcodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.common_barcodes()</span></code></a> to get the common barcodes across assays.</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">subcluster</span></code> paramter to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.clustered_barcodes.html#missionbio.mosaic.assay._Assay.clustered_barcodes" title="missionbio.mosaic.assay._Assay.clustered_barcodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.clustered_barcodes()</span></code></a> to prevent clustering within the labels</p></li>
<li><p>Option to pass n-dimensional arrays as splitby in <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.clustered_barcodes.html#missionbio.mosaic.assay._Assay.clustered_barcodes" title="missionbio.mosaic.assay._Assay.clustered_barcodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.clustered_barcodes()</span></code></a></p></li>
<li><p>Option to fetch a subset of the assays in <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.assays.html#missionbio.mosaic.sample.Sample.assays" title="missionbio.mosaic.sample.Sample.assays"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.assays()</span></code></a> using the <code class="docutils literal notranslate"><span class="pre">names</span></code> parameter</p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.clustered_barcodes.html#missionbio.mosaic.sample.Sample.clustered_barcodes" title="missionbio.mosaic.sample.Sample.clustered_barcodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.clustered_barcodes()</span></code></a> to hierarchically cluster using multiple assays</p></li>
<li><p>Multiple options added to <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.heatmap.html#missionbio.mosaic.sample.Sample.heatmap" title="missionbio.mosaic.sample.Sample.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.heatmap()</span></code></a> to sort the assays, barcodes, and the features</p></li>
<li><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.signature`()</span></code> accepts a <code class="docutils literal notranslate"><span class="pre">splitby</span></code> parameter to get the signature for each unique label in <code class="docutils literal notranslate"><span class="pre">splitby</span></code>.</p></li>
<li><p>Improvements to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.signaturemap.html#missionbio.mosaic.assay._Assay.signaturemap" title="missionbio.mosaic.assay._Assay.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.signaturemap()</span></code></a>:</p>
<ul>
<li><p>labels and ids are clustered by default.</p></li>
<li><p>Option to pass a list of labels to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.signaturemap.html#missionbio.mosaic.assay._Assay.signaturemap" title="missionbio.mosaic.assay._Assay.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.signaturemap()</span></code></a> to order the labels.</p></li>
<li><p>The default <code class="docutils literal notranslate"><span class="pre">features</span></code> option for <a class="reference internal" href="../pages/methods/missionbio.mosaic.cnv.Cnv.signaturemap.html#missionbio.mosaic.cnv.Cnv.signaturemap" title="missionbio.mosaic.cnv.Cnv.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">cnv.Cnv.signaturemap()</span></code></a> is set to <code class="docutils literal notranslate"><span class="pre">positions</span></code>.</p></li>
</ul>
</li>
<li><p>Option to copy the labels and palette together by passing an <a class="reference internal" href="../pages/missionbio.mosaic.assay._Assay.html#missionbio.mosaic.assay._Assay" title="missionbio.mosaic.assay._Assay"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay()</span></code></a> to <a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.set_labels.html#missionbio.mosaic.assay._Assay.set_labels" title="missionbio.mosaic.assay._Assay.set_labels"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.set_labels()</span></code></a></p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.assay._Assay.heatmap.html#missionbio.mosaic.assay._Assay.heatmap" title="missionbio.mosaic.assay._Assay.heatmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assay._Assay.heatmap()</span></code></a> sets <code class="docutils literal notranslate"><span class="pre">subcluster=False</span></code> when calculating the barcode order when convolve is provided.</p></li>
<li><p>Varsome URLs as hyperlinks on the variant name in the <a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html#missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable" title="missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSubcloneTable</span></code></a></p></li>
<li><p>Add percentage of cells and amplicons present to the <code class="xref py py-class docutils literal notranslate"><span class="pre">CopyNumberWorkflow</span></code></p></li>
<li><p><a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.mutated_cells.html#missionbio.mosaic.dna.Dna.mutated_cells" title="missionbio.mosaic.dna.Dna.mutated_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.mutated_cells()</span></code></a> to get the number of cells with at least 1 mutation in each given clone. This is used in <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.signaturemap.html#missionbio.mosaic.sample.Sample.signaturemap" title="missionbio.mosaic.sample.Sample.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.signaturemap()</span></code></a>.</p></li>
</ul>
</section>
<section id="id17">
<h3>Changed<a class="headerlink" href="#id17" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">apply_filter</span></code> changed to <code class="docutils literal notranslate"><span class="pre">filter_variants</span></code> in <a class="reference internal" href="../pages/functions/missionbio.mosaic.io.load.html#missionbio.mosaic.io.load" title="missionbio.mosaic.io.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">io.load()</span></code></a></p></li>
<li><p>SubcloneTree and SubcloneTreeGraph classes are renamed to Tree and TreeGraph respectively.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">show_plot</span></code> to <code class="docutils literal notranslate"><span class="pre">return_plot</span></code> in <a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.group_by_genotype.html#missionbio.mosaic.dna.Dna.group_by_genotype" title="missionbio.mosaic.dna.Dna.group_by_genotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.group_by_genotype()</span></code></a></p></li>
<li><p><code class="xref py py-class docutils literal notranslate"><span class="pre">plots.heatmap.Heatmap</span></code> splits the vertical and horizontal lines on the main heatmap into two traces.</p></li>
<li><p>The default value of <code class="docutils literal notranslate"><span class="pre">vaf_het</span></code> in <a class="reference internal" href="../pages/methods/missionbio.mosaic.dna.Dna.filter_variants.html#missionbio.mosaic.dna.Dna.filter_variants" title="missionbio.mosaic.dna.Dna.filter_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dna.Dna.filter_variants()</span></code></a> changed from 35 to 30.</p></li>
<li><p>Flattened <code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.heatmap`()</span></code> option has been removed. A more customizable version is available under the <a class="reference internal" href="../pages/methods/missionbio.mosaic.sample.Sample.signaturemap.html#missionbio.mosaic.sample.Sample.signaturemap" title="missionbio.mosaic.sample.Sample.signaturemap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample.Sample.signaturemap()</span></code></a> function.</p></li>
<li><p>The constant - <code class="xref py py-attr docutils literal notranslate"><span class="pre">constants.COLORS</span></code> to have unique values.</p>
<ul>
<li><p>The grey values at the 10th, 20th, 30th.. positions were modified to be unique</p></li>
<li><p>The black (<code class="docutils literal notranslate"><span class="pre">#000000</span></code>) value was moved from the 20th position to the last position</p></li>
</ul>
</li>
</ul>
</section>
<section id="id18">
<h3>Fixed<a class="headerlink" href="#id18" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Get indexes maintains the order as per <code class="docutils literal notranslate"><span class="pre">find_list</span></code> when there are duplicates in the <code class="docutils literal notranslate"><span class="pre">find_list</span></code> and <code class="docutils literal notranslate"><span class="pre">order_using_find_list</span></code> is True.</p></li>
<li><p>DANN score in the variants subclone table is shown correctly for saved h5 files.</p></li>
<li><p>Overlapping of text in phylogeny trees.</p></li>
<li><p>Error in multiprocessing when fetching gene_names for CNV by adding a <code class="docutils literal notranslate"><span class="pre">max_workers</span></code> parameter and using threads instead of processes.</p></li>
<li><p>Missing clone is ignored when finding ADO sisters.</p></li>
</ul>
</section>
<section id="removed">
<h3>Removed<a class="headerlink" href="#removed" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Functions to convert legacy loom files to h5 files - <code class="docutils literal notranslate"><span class="pre">io._loom_to_h5</span></code>, <code class="docutils literal notranslate"><span class="pre">io._update_file</span></code></p></li>
<li><p>Functions to read data from csv files - <code class="docutils literal notranslate"><span class="pre">io._merge_files</span></code>, <code class="docutils literal notranslate"><span class="pre">io._cnv_raw_counts</span></code>, <code class="docutils literal notranslate"><span class="pre">io._protein_raw_counts</span></code></p></li>
<li><p>Function to merge h5 files - <code class="docutils literal notranslate"><span class="pre">io._merge</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">show_plot</span></code> from <a class="reference internal" href="../pages/methods/missionbio.mosaic.protein.Protein.normalize_reads.html#missionbio.mosaic.protein.Protein.normalize_reads" title="missionbio.mosaic.protein.Protein.normalize_reads"><code class="xref py py-meth docutils literal notranslate"><span class="pre">protein.Protein.normalize_reads()</span></code></a>. The same plot can be created in plotly using <code class="xref py py-meth docutils literal notranslate"><span class="pre">algorithms.nsp.NSP.plot()</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">show_plot</span></code> from <code class="xref py py-meth docutils literal notranslate"><span class="pre">protein.Protein.get_signal_profile()</span></code>. The same plot can be created in plotly using <code class="xref py py-meth docutils literal notranslate"><span class="pre">algorithms.nsp.ExpressionProfile.plot()</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">protein.Protein.get_signal_profile</span></code> function. It can be executed using <code class="xref py py-meth docutils literal notranslate"><span class="pre">algorithms.nsp.ExpressionProfile.fit()</span></code> if needed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">protein.Protein.get_scaling_factor</span></code> function. It can be executed using <code class="xref py py-meth docutils literal notranslate"><span class="pre">algorithms.nsp.NSP.scaling_factor()</span></code> if needed</p></li>
</ul>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="help.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Additional resources</p>
      </div>
    </a>
    <a class="right-next"
       href="../pages/missionbio.mosaic.assay._Assay.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">_Assay</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-12-0">v3.12.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#added">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fixed">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-7-0">v3.7.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#changed">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-4-0">v3.4.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-1-1">v3.1.1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-1-0">v3.1.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id9">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id11">Fixed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#updated">Updated</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-0-1">v3.0.1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id14">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id15">Fixed</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v3-0-0">v3.0.0</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id16">Added</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id17">Changed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id18">Fixed</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#removed">Removed</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Saurabh Parikh
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021, Mission Bio.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>