

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Cohort Analysis Pipeline &#8212; Mosaic v3.12.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'manual/cohort';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Vignettes" href="vignettes.html" />
    <link rel="prev" title="Data structure" href="data_structure.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Mosaic v3.12.2 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="data_structure.html">Data structure</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Cohort Analysis Pipeline</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="vignettes.html">Vignettes</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/overview.html">Overview of mosaic</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/filtering.html">Filtering barcodes and ids</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/cnv.html">Manual CNV analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/plot-customizations.html">Plot customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/multi-sample-analysis.html">Multisample analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/additional-plots-and-customizations.html">Additional plots and customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/code-snippets.html">Code Snippets</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="curated_jupyter_notebooks.html">Curated Jupyter Notebooks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna.html">Single Sample DNA-only</a></li>






<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-multisample.html">Multisample DNA-only</a></li>






<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-protein.html">Single Sample DNA + Protein</a></li>







<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-protein-multisample.html">Multisample DNA + Protein</a></li>







</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Additional resources</a></li>

<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Classes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.assay._Assay.html">_Assay</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.dna.Dna.html">Dna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.cnv.Cnv.html">Cnv</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.protein.Protein.html">Protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.sample.Sample.html">Sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.samplegroup.SampleGroup.html">SampleGroup</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interactive Workflows</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.workflows.copy_number.CopyNumber.html">CopyNumber</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html">VariantSubcloneTable</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Algorithms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.nsp.NSP.html">NSP</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.algorithms.group_by_genotype.GroupByGenotype.html">GroupByGenotype</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.algorithms.compass.COMPASS.html">COMPASS</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Functional Modules</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.io.html">io</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.utils.html">utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.constants.html">constants</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Custom Plots</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.heatmap.Heatmap.html">Heatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.lineplot.LinePlot.html">LinePlot</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.bargraph.BarGraph.html">BarGraph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.multimap.MultiMap.html">MultiMap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.phylotree.PhyloTree.html">PhyloTree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.plotting.fishplot.Fishplot.html">Fishplot</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">DNA assignment</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.dna.truth.Truth.html">Truth</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.dna.likelihood.LikelihoodMethod.html">LikelihoodMethod</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.mosaic.algorithms.dna_assignment.DnaAssignment.html">DnaAssignment</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Protein assignment</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.truth.Truth.html">Truth</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.pace.likelihood.LikelihoodMethod.html">LikelihoodMethod</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pages/missionbio.demultiplex.protein.pace.probability.ProbabilityMethod.html">ProbabilityMethod</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/manual/cohort.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Cohort Analysis Pipeline</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prerequisites">Prerequisites</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pipeline-steps">Pipeline Steps</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-line-usage">Command-Line Usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#h5-files">H5 Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#demultiplex-csv-file-optional">Demultiplex CSV File (Optional)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#spike-in-csv-file-optional">Spike-in CSV File (Optional)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#somatic-variants-csv-file-optional">Somatic Variants CSV File (Optional)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#yaml-configuration-file">YAML Configuration File</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#error-handling">Error Handling</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
    <!--Add a white cover to hide the .bd-container grey on the right side of the page.-->
    <div class="background-cover" id="background-cover"></div>
    <section id="cohort-analysis-pipeline">
<h1>Cohort Analysis Pipeline<a class="headerlink" href="#cohort-analysis-pipeline" title="Permalink to this heading">#</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">#</a></h2>
<p>The Tertiary Cohort Analysis Pipeline is designed to process and analyze cohorts of <cite>.h5</cite> files
generated from Mission Bio’s Tapestri platform. It supports variant detection, whole genome CNV
analysis, phylogeny generation, and report creation for individual samples and the entire cohort.</p>
<p>This guide explains how to configure and run the pipeline.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Download the <a class="reference external" href="https://github.com/MissionBio/mosaic/raw/refs/heads/master/dataset/cohort/cohort_dataset.zip">example dataset</a> to test the pipeline locally.</p>
</div>
</section>
<section id="prerequisites">
<h2>Prerequisites<a class="headerlink" href="#prerequisites" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><strong>Python Environment</strong>: Ensure you have mosaic 3.12.0 or above installed and activated in your terminal.</p></li>
<li><p><strong>Input Files</strong>: Organize the relevant <code class="docutils literal notranslate"><span class="pre">.h5</span></code> files in a folder.</p></li>
<li><p><strong>Configuration File</strong>: A YAML configuration file specifying pipeline parameters. (Explained below)</p></li>
</ol>
</section>
<section id="pipeline-steps">
<h2>Pipeline Steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this heading">#</a></h2>
<p>The pipeline consists of the following steps:</p>
<ol class="arabic simple">
<li><p><strong>Demultiplexing</strong>: Splits multiplexed samples per patient <code class="docutils literal notranslate"><span class="pre">.h5</span></code> files.</p></li>
<li><p><strong>Genome-wide CNV Analysis (Optional)</strong>: Detects copy number variations and CNV subclones.</p></li>
<li><p><strong>Variant Detection</strong>: Identifies somatic variants for each patient.</p></li>
<li><p><strong>Phylogeny</strong>: Identifies clones using the detected variants and estimates the phylogeny.</p></li>
<li><p><strong>Postprocessing</strong>: Normalizes protein data and assigns final clone labels.</p></li>
<li><p><strong>Report Generation</strong>: Creates HTML reports for individual patients and the cohort.</p></li>
</ol>
</section>
<section id="command-line-usage">
<h2>Command-Line Usage<a class="headerlink" href="#command-line-usage" title="Permalink to this heading">#</a></h2>
<p>The pipeline can be executed using the <cite>tapestri tertiary cohort</cite> command from the CLI.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tapestri<span class="w"> </span>tertiary<span class="w"> </span>cohort<span class="w"> </span>process<span class="w"> </span>--config<span class="w"> </span>&lt;config_file&gt;<span class="w"> </span>--output-folder<span class="w"> </span>&lt;output_folder&gt;<span class="w"> </span><span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<ul class="simple">
<li><p><cite>–config</cite>: Path to the YAML configuration file.</p></li>
<li><p><cite>–output-folder</cite>: Directory where output files will be saved.</p></li>
<li><p><cite>–overwrite</cite>: Optional flag to overwrite the output folder if it already exists.</p></li>
</ul>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<section id="h5-files">
<h3>H5 Files<a class="headerlink" href="#h5-files" title="Permalink to this heading">#</a></h3>
<p>Any h5 file can be used as input. Most commonly, the h5 files generated by the Tapestri DNA or
DNA+Protein pipeline are used for the cohort pipeline.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All h5 files used in the analysis must be of the same type. DNA only h5 files cannot be
combined with DNA+Protein h5 files. Also files from different DNA or protein panels cannot be
combined in one analysis.</p>
</div>
</section>
<section id="demultiplex-csv-file-optional">
<h3>Demultiplex CSV File (Optional)<a class="headerlink" href="#demultiplex-csv-file-optional" title="Permalink to this heading">#</a></h3>
<p>Only required when processing multiplexed samples.</p>
<p>The demultiplexing csv file contains variant information to identify each patient sample in a
multiplexed run. Only one file is needed for the entire cohort. It must contain information for all
patients across all the runs. The genotype values must be 0, 1, or 2 which correspond to homozygous
reference (wildtype), heterozygous, and homozygous alternate respectively. An example of the
demultiplexing CSV file is shown below:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>variant_id,sample_id,type,genotype
chr1:36933096:T/C,Patient1,germline,0
chr1:36933096:T/C,Patient2,germline,1
chr1:36933096:T/C,Patient3,germline,0
chr1:43805240:A/G,Patient1,germline,2
chr1:43805240:A/G,Patient2,germline,0
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only variants that overlap with the panel will be used for demultiplexing. At least 10 variants
per patient are required by the pipeline.</p>
</div>
</section>
<section id="spike-in-csv-file-optional">
<h3>Spike-in CSV File (Optional)<a class="headerlink" href="#spike-in-csv-file-optional" title="Permalink to this heading">#</a></h3>
<p>Only required for genome-wide CNV samples.</p>
<p>This file is similar to the demultiplexing CSV file but is used to identify the genotypes for the
diploid spike-in cell line. It must contain only one <code class="docutils literal notranslate"><span class="pre">sample_id</span></code>. An example of the spike-in CSV
file is shown below:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>variant_id,sample_id,type,genotype
chr1:36933096:T/C,SpikeIn,germline,1
chr1:43805240:A/G,SpikeIn,germline,2
</pre></div>
</div>
</section>
<section id="somatic-variants-csv-file-optional">
<h3>Somatic Variants CSV File (Optional)<a class="headerlink" href="#somatic-variants-csv-file-optional" title="Permalink to this heading">#</a></h3>
<p>Needed if the variant selection has to be modified using a whitelist/blacklist</p>
<p>This file lists variants that are whitelisted or blacklisted for the phylogeny creation. Only a
single file containing variants for all samples in the cohort is needed. The <code class="docutils literal notranslate"><span class="pre">genotype</span></code> column is
optional for this file. It does not have any effect on the output. An example of the somatic
variants CSV file is shown below:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>variant_id,sample_id,type
chr5:170837543:C/CTCTG,Patient1,blacklist
chr4:55599436:T/C,Patient2,whitelist
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To ensure quick processing, the pipeline only analyses variants that pass the default Tapestri
pipeline filters. These are variants with mutation rates over 1% in the pre-demultiplexed h5 file.
But for timecourse samples, if any one time point has such a variant, then it will be used for other
timepoints irrespective of the mutation rate. Any variant with a lower mutation rate must be given
through the whitelist.</p>
</div>
</section>
<section id="yaml-configuration-file">
<h3>YAML Configuration File<a class="headerlink" href="#yaml-configuration-file" title="Permalink to this heading">#</a></h3>
<p>The YAML configuration file specifies pipeline parameters and paths to input files. An example of a
complete configuration file is shown below. Some of the parameters are optional and dependent on the
type of analysis being performed.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">spike_in</span><span class="p">:</span>
<span class="w">    </span><span class="nt">variants</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/spike_in_variants.csv</span><span class="w">  </span><span class="c1"># Optional - required for genome-wide CNV</span>
<span class="nt">demultiplex</span><span class="p">:</span>
<span class="w">   </span><span class="nt">variants</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/demultiplexing.csv</span><span class="w">  </span><span class="c1"># Optional - required for multiplexed samples</span>
<span class="nt">tertiary</span><span class="p">:</span>
<span class="w">   </span><span class="nt">variants</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/somatic_variants.csv</span><span class="w">  </span><span class="c1"># Optional - if whitelist/blacklist is needed</span>
<span class="w">   </span><span class="nt">report</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">hem-onc</span>
<span class="w">   </span><span class="nt">input</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">h5</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/sample1.dna+protein.h5</span>
<span class="w">        </span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Sample1</span>
<span class="w">        </span><span class="nt">patient</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="s">&quot;Patient1&quot;</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="s">&quot;Patient2&quot;</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="s">&quot;Patient3&quot;</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">timepoint</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">1</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">1</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">1</span><span class="p p-Indicator">]</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">h5</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/sample2.dna+protein.h5</span>
<span class="w">        </span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Sample2</span>
<span class="w">        </span><span class="nt">patient</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="s">&quot;Patient1&quot;</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="s">&quot;Patient2&quot;</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">timepoint</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">2</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">2</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
<p>Some things to note about the configuration file:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">name</span></code> can be used to shorten the h5 file name. They are shown in the report and names that are
too long can hamper the readability of some figures. The same is true for <code class="docutils literal notranslate"><span class="pre">patient</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">timepoint</span></code> is optional. It is needed only if there are multiple timepoints for a patient. It
must be a list of integers and not strings. The samples are shown in the reports in ascending
order of timepoints. The value of the timepoint is not relevant, only the relative order of the
timepoints is important.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">patient</span></code> is only required for multiplexed samples. It is the list of patients multiplexed
together in each h5 file. The value in <code class="docutils literal notranslate"><span class="pre">patient</span></code> should match the <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> in the
demultiplexing and somatic csv files.</p></li>
<li><p>For single-sample runs the <code class="docutils literal notranslate"><span class="pre">name</span></code> should match the <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> in the somatic csv file.</p></li>
</ul>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>The pipeline generates the following outputs:</p>
<ol class="arabic simple">
<li><p><strong>Processed h5 Files</strong>: Located in the <code class="docutils literal notranslate"><span class="pre">h5</span></code> subdirectory. These contain all the relevant
information to generate the reports. No other files are needed. These can be loaded in a Jupyter
notebook to create custom figures that are not available in the standard output.</p></li>
<li><p><strong>Logs</strong>: Detailed logs of the pipeline execution in the <code class="docutils literal notranslate"><span class="pre">tertiary.log</span></code> file.</p></li>
<li><p><strong>Reports</strong>: HTML reports for individual samples and the cohort in the <code class="docutils literal notranslate"><span class="pre">reports</span></code> subdirectory.
The cohort figures are interactive HTML files. All figures can be downloaded as png files by
opening them in a web browser tab and using the download button on the top right of the figure.
The figures are:</p>
<ol class="loweralpha simple">
<li><p><strong>Cohort map</strong>: It shows the summary of clonal architecture, mutations, and correlation with
protein clusters for all samples. The protein cluster names can be renamed using the button at
the bottom of the figure.</p></li>
<li><p><strong>Protein UMAP</strong>: This figure is used to check the validity of the protein clustering. It can
be colored by cluster labels, sample names, point density, and sample density.</p></li>
<li><p><strong>Protein expression UMAP</strong>: It shows the expression of every antibody in the protein panel
on the UMAP. It can be used to identify the celltype for each cluster.</p></li>
<li><p><strong>Variant clonality</strong>: A figure and table that is helpful to identify any false positive
variants that might have been called by the pipeline. The whitelist/blacklist option can be
used when rerunning the pipeline to the correct variant calls.</p></li>
</ol>
</li>
</ol>
<p>Multiple other folders are created, those are intermediate files for each step of the pipeline.
These are kept for debugging purposes and might be removed in a future version.</p>
</section>
<section id="error-handling">
<h2>Error Handling<a class="headerlink" href="#error-handling" title="Permalink to this heading">#</a></h2>
<p>If the pipeline encounters an error, it will log the issue and terminate. Common issues include:</p>
<ul class="simple">
<li><p>Missing or invalid input files.</p></li>
<li><p>Incorrect configuration parameters.</p></li>
</ul>
<p>Refer to the error message in the last line of the log file to identify the issue.</p>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="data_structure.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Data structure</p>
      </div>
    </a>
    <a class="right-next"
       href="vignettes.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Vignettes</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prerequisites">Prerequisites</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pipeline-steps">Pipeline Steps</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-line-usage">Command-Line Usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#h5-files">H5 Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#demultiplex-csv-file-optional">Demultiplex CSV File (Optional)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#spike-in-csv-file-optional">Spike-in CSV File (Optional)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#somatic-variants-csv-file-optional">Somatic Variants CSV File (Optional)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#yaml-configuration-file">YAML Configuration File</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#error-handling">Error Handling</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Saurabh Parikh
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021, Mission Bio.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>