

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Data structure &#8212; Mosaic v3.1.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'pages/data_structure';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Vignettes" href="vignettes.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    <p class="title logo__title">Mosaic v3.1.1 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Data structure</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="vignettes.html">Vignettes</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/overview.html">Overview of mosaic</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/filtering.html">Filtering barcodes and ids</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/cnv.html">Manual CNV analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/plot-customizations.html">Plot customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/multi-sample-analysis.html">Multisample analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/additional-plots-and-customizations.html">Additional plots and customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/code-snippets.html">Code Snippets</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="curated_jupyter_notebooks.html">Curated Jupyter Notebooks</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna.html">Single Sample DNA-only notebook</a></li>







<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-multisample.html">Multisample DNA-only notebook</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-protein.html">Single Sample DNA + Protein notebook</a></li>








<li class="toctree-l2"><a class="reference internal" href="../notebooks/curated_notebooks/dna-protein-multisample.html">Multisample DNA + Protein notebook</a></li>








</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Additional resources</a></li>

<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Classes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.assay._Assay.html">_Assay</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.dna.Dna.html">Dna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.cnv.Cnv.html">Cnv</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.protein.Protein.html">Protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.sample.Sample.html">Sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.samplegroup.SampleGroup.html">SampleGroup</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interactive Workflows</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.workflows.copy_number.CopyNumber.html">CopyNumber</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html">VariantSubcloneTable</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Algorithms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.algorithms.nsp.NSP.html">NSP</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.algorithms.group_by_genotype.GroupByGenotype.html">GroupByGenotype</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.algorithms.compass.COMPASS.html">COMPASS</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Functional Modules</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.io.html">io</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.utils.html">utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.constants.html">constants</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Custom Plots</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.plots.heatmap.Heatmap.html">Heatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.plots.lineplot.LinePlot.html">LinePlot</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.plots.bargraph.BarGraph.html">BarGraph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.plots.multi_heatmap.MultiHeatmap.html">MultiHeatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.plots.fishplot.TreeGraph.html">TreeGraph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.plots.fishplot.Fishplot.html">Fishplot</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/pages/data_structure.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Data structure</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basics">Basics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dna">DNA</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#layers">Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#column-attributes">Column attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#row-attributes">Row attributes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv">CNV</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Column attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Row attributes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#protein">Protein</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id9">Column attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Row attributes</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
    <!--Add a white cover to hide the .bd-container grey on the right side of the page.-->
    <div class="background-cover" id="background-cover"></div>
    <section id="data-structure">
<span id="id1"></span><h1>Data structure<a class="headerlink" href="#data-structure" title="Permalink to this heading">#</a></h1>
<section id="basics">
<h2>Basics<a class="headerlink" href="#basics" title="Permalink to this heading">#</a></h2>
<p>Mosaic requires that you provide it an .h5 file before you do any analyses. The Tapestri Pipeline v2
generates this .h5 output file, which is a multi-omics file format stored in a structured
<a class="reference external" href="https://www.hdfgroup.org/solutions/hdf5/">HDF5</a> file. This file can contain data for one or more
runs. Each run can contain data for one or more “assay,” where each assay contains data for a different
analyte (e.g. DNA, protein, etc.). The Mission Bio <cite>.h5</cite> files are structured as follows:</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="n">missionbio</span><span class="o">.</span><span class="n">h5</span><span class="o">.</span><span class="n">assay</span><span class="o">.</span><span class="n">Assay</span>
      <span class="p">(</span><span class="n">h5</span> <span class="n">reader</span> <span class="ow">and</span> <span class="n">writer</span><span class="p">)</span>
               <span class="o">|</span>
               <span class="o">|</span>
 <span class="n">missionbio</span><span class="o">.</span><span class="n">mosaic</span><span class="o">.</span><span class="n">assay</span><span class="o">.</span><span class="n">_Assay</span>
<span class="p">(</span><span class="n">basic</span> <span class="n">interactivity</span> <span class="ow">and</span> <span class="n">plotting</span><span class="p">)</span>
               <span class="o">|</span>
               <span class="o">|</span>
   <span class="o">---------------------------</span>
   <span class="o">|</span>           <span class="o">|</span>             <span class="o">|</span>
   <span class="o">|</span>           <span class="o">|</span>             <span class="o">|</span>
   <span class="o">|</span>           <span class="o">|</span>             <span class="o">|</span>
  <span class="n">Dna</span>         <span class="n">Cnv</span>         <span class="n">Protein</span>
  <span class="p">(</span><span class="n">Assay</span><span class="o">-</span><span class="n">specific</span> <span class="n">functionality</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p><cite>missionbio</cite> is the top level package name Mission Bio uses for its software packages.
<cite>missionbio.h5</cite> is the module used to read/write/process .h5 files, which then get used by the <cite>missionbio.mosaic</cite> module.</p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p><code class="xref py py-class docutils literal notranslate"><span class="pre">missionbio.h5.data.assay.Assay</span></code>
This is a module for reading and writing .h5 files. You should not need to access this class directly.</p>
<p><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.assay._Assay.html#missionbio.mosaic.assay._Assay" title="missionbio.mosaic.assay._Assay"><code class="xref py py-class docutils literal notranslate"><span class="pre">_Assay</span></code></a>
This is the base class for each type of assay. It is built on top of <code class="xref py py-class docutils literal notranslate"><span class="pre">missionbio.h5.data.assay.Assay</span></code> to
incorporate functionality specific to tertiary analysis. It also provides generic functions for visualizing
the data. It stores four main groups of information for the assay:</p>
<blockquote>
<div><ol class="arabic simple">
<li><dl class="simple">
<dt>metadata (<a class="reference internal" href="../autosummary_pages/methods/missionbio.mosaic.assay._Assay.add_metadata.html#missionbio.mosaic.assay._Assay.add_metadata" title="missionbio.mosaic.assay._Assay.add_metadata"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_metadata()</span></code></a> / <a class="reference internal" href="../autosummary_pages/methods/missionbio.mosaic.assay._Assay.del_metadata.html#missionbio.mosaic.assay._Assay.del_metadata" title="missionbio.mosaic.assay._Assay.del_metadata"><code class="xref py py-meth docutils literal notranslate"><span class="pre">del_metadata()</span></code></a>):</dt><dd><p>Dictionary containing the metrics of the assay (e.g. sample name).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>row_attrs (<a class="reference internal" href="../autosummary_pages/methods/missionbio.mosaic.assay._Assay.add_row_attr.html#missionbio.mosaic.assay._Assay.add_row_attr" title="missionbio.mosaic.assay._Assay.add_row_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_row_attr()</span></code></a> / <a class="reference internal" href="../autosummary_pages/methods/missionbio.mosaic.assay._Assay.del_row_attr.html#missionbio.mosaic.assay._Assay.del_row_attr" title="missionbio.mosaic.assay._Assay.del_row_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">del_row_attr()</span></code></a>):</dt><dd><p>Dictionary containing meta information for each barcode/cell in the assay (e.g. basepair sequence for
each barcode, barcode label, etc.). All the values must be of the same length, i.e., match the number
of barcodes. This is the attribute where ‘label’, ‘pca’, and ‘umap’ values are added.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>col_attrs (<a class="reference internal" href="../autosummary_pages/methods/missionbio.mosaic.assay._Assay.add_col_attr.html#missionbio.mosaic.assay._Assay.add_col_attr" title="missionbio.mosaic.assay._Assay.add_col_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_col_attr()</span></code></a> / <a class="reference internal" href="../autosummary_pages/methods/missionbio.mosaic.assay._Assay.del_col_attr.html#missionbio.mosaic.assay._Assay.del_col_attr" title="missionbio.mosaic.assay._Assay.del_col_attr"><code class="xref py py-meth docutils literal notranslate"><span class="pre">del_col_attr()</span></code></a>)):</dt><dd><p>Dictionary containing meta information for each feature measured in the assay (i.e. variants from
a DNA assay, antibodies from a protein assay, etc.). All the values must be of the same length, i.e.,
match the number of ids. ‘ids’ contains variants for DNA assays and antibodies for Protein assays.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>layers (add_layer / del_layer):</dt><dd><p>Dictionary containing the values for each barcode and feature combination measured in the assay
(e.g. <cite>read_counts</cite>, <cite>genotype</cite>, etc.). All the values have the shape (number barcodes) x (number features).</p>
</dd>
</dl>
</li>
</ol>
</div></blockquote>
<p><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.dna.Dna.html#missionbio.mosaic.dna.Dna" title="missionbio.mosaic.dna.Dna"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dna</span></code></a>, <a class="reference internal" href="../autosummary_pages/missionbio.mosaic.cnv.Cnv.html#missionbio.mosaic.cnv.Cnv" title="missionbio.mosaic.cnv.Cnv"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cnv</span></code></a>, and
<a class="reference internal" href="../autosummary_pages/missionbio.mosaic.protein.Protein.html#missionbio.mosaic.protein.Protein" title="missionbio.mosaic.protein.Protein"><code class="xref py py-class docutils literal notranslate"><span class="pre">Protein</span></code></a> are subclasses of <a class="reference internal" href="../autosummary_pages/missionbio.mosaic.assay._Assay.html#missionbio.mosaic.assay._Assay" title="missionbio.mosaic.assay._Assay"><code class="xref py py-class docutils literal notranslate"><span class="pre">_Assay</span></code></a>, so
they inherit all of its plotting and interactive functionality. However, each class also includes methods
specific to the assay, such as normalization and cell type identification.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Sample</span>
  <span class="o">|</span>
  <span class="o">|--</span> <span class="n">dna</span>
  <span class="o">|--</span> <span class="n">cnv</span>
  <span class="o">|--</span> <span class="n">protein</span>
  <span class="o">|--</span> <span class="n">cnv_raw</span>
  <span class="o">|--</span> <span class="n">protein_raw</span>
</pre></div>
</div>
<p><a class="reference internal" href="../autosummary_pages/missionbio.mosaic.sample.Sample.html#missionbio.mosaic.sample.Sample" title="missionbio.mosaic.sample.Sample"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sample</span></code></a> object contains all three of the above assays along with the raw counts for the sample. The raw counts are also stored as the respective assay objects.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="dna">
<h2>DNA<a class="headerlink" href="#dna" title="Permalink to this heading">#</a></h2>
<p>Described here are common types and descriptions of data stored in each layer of the DNA assay.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>Mosaic provides a default level of cell and variant filtering (but does not remove any of the cells
by default). The results of the filtering are contained in <cite>sample.dna.layers[‘FILTER_MASK’]</cite>,
<cite>sample.dna.col_attrs[‘filtered’]</cite>, <cite>sample.dna.row_attrs[‘filtered’]</cite>, <cite>n_passing_variants</cite>,
<cite>n_passing_cells</cite>, and <cite>n_passing_variants_per_cell</cite>. The parameters that control that filtering are
<cite>af_cutoff</cite>, <cite>dp_cutoff</cite>, <cite>gq_cutoff</cite>, <cite>missing_cells_cutoff</cite>, <cite>missing_variants_cutoff</cite>, and
<cite>mutated_cells_cutoff</cite>.</p>
</div>
<section id="metadata">
<h3>Metadata<a class="headerlink" href="#metadata" title="Permalink to this heading">#</a></h3>
<table class="table" id="id11">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id11" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample_name</p></td>
<td><p>The sample name.</p></td>
</tr>
<tr class="row-odd"><td><p>ado_rate</p></td>
<td><p>The estimated rate of allelic dropout.</p></td>
</tr>
<tr class="row-even"><td><p>af_cutoff</p></td>
<td><p>Minimum variant allelic fraction used when computing <cite>sample.dna.layers[‘FILTER_MASK’]</cite>, <cite>sample.dna.col_attrs[‘filtered’]</cite>, and <cite>sample.dna.row_attrs[‘filtered’]</cite>, as well as the filtered variant metrics such as <cite>n_passing_variants</cite>, <cite>n_passing_cells</cite>, and <cite>n_passing_variants_per_cell</cite>. Does not actually remove any variants from the assay.</p></td>
</tr>
<tr class="row-odd"><td><p>avg_mapping_error_rate</p></td>
<td><p>Ratio of mismatches to bases mapped (see <cite>samtools stats</cite>).</p></td>
</tr>
<tr class="row-even"><td><p>avg_panel_uniformity</p></td>
<td><p>The proportion of amplicons that pass the panel uniformity test (see <a class="reference external" href="https://support.missionbio.com/hc/en-us/articles/360053187154-Run-Report-Metrics#h_01EHSC7F4BZJ3678B16S0CTVVW">here</a> for more information).</p></td>
</tr>
<tr class="row-odd"><td><p>chemistry_version</p></td>
<td><p>The version of the Tapestri kit used.</p></td>
</tr>
<tr class="row-even"><td><p>dp_cutoff</p></td>
<td><p>Minimum variant depth of coverage when computing <cite>sample.dna.layers[‘FILTER_MASK’]</cite>, <cite>sample.dna.col_attrs[‘filtered’]</cite>, and <cite>sample.dna.row_attrs[‘filtered’]</cite>, as well as the filtered variant metrics such as <cite>n_passing_variants</cite>, <cite>n_passing_cells</cite>, and <cite>n_passing_variants_per_cell</cite>. Does not actually remove any variants from the assay.</p></td>
</tr>
<tr class="row-odd"><td><p>genome_version</p></td>
<td><p>The genome version used for mapping and variant calling.</p></td>
</tr>
<tr class="row-even"><td><p>gq_cutoff</p></td>
<td><p>The minimum variant genotype quality when computing <cite>sample.dna.layers[‘FILTER_MASK’]</cite>, <cite>sample.dna.col_attrs[‘filtered’]</cite>, and <cite>sample.dna.row_attrs[‘filtered’]</cite>, as well as the filtered variant metrics such as <cite>n_passing_variants</cite>, <cite>n_passing_cells</cite>, and <cite>n_passing_variants_per_cell</cite>. Does not actually remove any variants from the assay.</p></td>
</tr>
<tr class="row-odd"><td><p>high_quality_variants</p></td>
<td><p>Number of variants with QUAL values greater than a minimum threshold.</p></td>
</tr>
<tr class="row-even"><td><p>missing_cells_cutoff</p></td>
<td><p>Variants that cannot be genotyped (i.e. <cite>NGT</cite> is equal to 0, 1, or 2) in more than this percentage of cells are not used when calculating <cite>sample.dna.col_attrs[‘filtered’]</cite> the filtered variant metrics such as <cite>n_passing_variants</cite> and <cite>n_passing_variants_per_cell</cite>. Does not actually remove any variants from the assay.</p></td>
</tr>
<tr class="row-odd"><td><p>missing_variants_cutoff</p></td>
<td><p>Cells that have more than this percentage of filtered variants that cannot be genotyped (i.e. <cite>NGT</cite> is equal to 0, 1, or 2) are not used when calculating <cite>sample.dna.row_attrs[‘filtered’]</cite> and the filtered variant metrics such as <cite>n_passing_cells</cite> and <cite>n_passing_variants_per_cell</cite>. Does not actually remove any cells from the assay.</p></td>
</tr>
<tr class="row-even"><td><p>mutated_cells_cutoff</p></td>
<td><p>Variants that are not mutated (i.e. <cite>NGT</cite> is equal to 1 or 2) in more than this percentage of cells are not used when calculating <cite>sample.dna.col_attrs[‘filtered’]</cite> and the filtered variant metrics such as <cite>n_passing_variants</cite> and <cite>n_passing_variants_per_cell</cite>. Does not actually remove any variants from the assay.</p></td>
</tr>
<tr class="row-odd"><td><p>n_amplicons</p></td>
<td><p>Number of amplicons in the panel.</p></td>
</tr>
<tr class="row-even"><td><p>n_bases_r1</p></td>
<td><p>Number of bases from the forward reads.</p></td>
</tr>
<tr class="row-odd"><td><p>n_bases_r1_q30</p></td>
<td><p>Number of bases that have a Q value of 30 or greater from the forward reads.</p></td>
</tr>
<tr class="row-even"><td><p>n_bases_r2</p></td>
<td><p>Number of bases from the reverse reads.</p></td>
</tr>
<tr class="row-odd"><td><p>n_bases_r2_q30</p></td>
<td><p>Number of bases that have a Q value of 30 or greater from the reverse reads.</p></td>
</tr>
<tr class="row-even"><td><p>n_cell_barcode_bases</p></td>
<td><p>Total number of bases from cell barcodes.</p></td>
</tr>
<tr class="row-odd"><td><p>n_cell_barcode_bases_q30</p></td>
<td><p>Total number of bases (with quality of 30 or greater) from cell barcodes.</p></td>
</tr>
<tr class="row-even"><td><p>n_cells</p></td>
<td><p>Number of cells.</p></td>
</tr>
<tr class="row-odd"><td><p>n_passing_cells</p></td>
<td><p>Number of passing cells.</p></td>
</tr>
<tr class="row-even"><td><p>n_passing_variants</p></td>
<td><p>Number of passing variants.</p></td>
</tr>
<tr class="row-odd"><td><p>n_passing_variants_per_cell</p></td>
<td><p>Median number of mutations (among the passing variants) per passing cell. Here, mutations refers to variants where the <cite>NGT</cite> is equal to <cite>1</cite> or <cite>2</cite>.</p></td>
</tr>
<tr class="row-even"><td><p>n_read_pairs</p></td>
<td><p>Number of read pairs.</p></td>
</tr>
<tr class="row-odd"><td><p>n_read_pairs_mapped_to_cells</p></td>
<td><p>Number of read pairs that mapped to cells.</p></td>
</tr>
<tr class="row-even"><td><p>n_read_pairs_trimmed</p></td>
<td><p>Number of read pairs that passed trimming.</p></td>
</tr>
<tr class="row-odd"><td><p>n_read_pairs_valid_cell_barcodes</p></td>
<td><p>Number of read pairs from cell barcodes that are valid.</p></td>
</tr>
<tr class="row-even"><td><p>n_reads_mapped</p></td>
<td><p>Number of read pairs that mapped to the genome.</p></td>
</tr>
<tr class="row-odd"><td><p>n_reads_mapped_insert</p></td>
<td><p>Total number of reads that map to the amplicons.</p></td>
</tr>
<tr class="row-even"><td><p>n_variants_per_cell</p></td>
<td><p>Median number of mutations per cell. Similar to <cite>n_passing_variants_per_cell</cite>, but the variants and cells were not filtered.</p></td>
</tr>
<tr class="row-odd"><td><p>panel_name</p></td>
<td><p>The name of the panel.</p></td>
</tr>
<tr class="row-even"><td><p>pipeline_version</p></td>
<td><p>The Tapestri pipeline version used.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="layers">
<h3>Layers<a class="headerlink" href="#layers" title="Permalink to this heading">#</a></h3>
<table class="table" id="id12">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id12" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>AF</p></td>
<td><p>The allelic fraction of the variant in each cell/barcode (ranges from 0 to 100). Missing values are set to 0.</p></td>
</tr>
<tr class="row-odd"><td><p>AF_MISSING</p></td>
<td><p>The allelic fraction of the variant in each cell/barcode (ranges from 0 to 100). Missing values are set to -50.</p></td>
</tr>
<tr class="row-even"><td><p>DP</p></td>
<td><p>Approximate read depth (reads with MQ=255 or with bad mates are filtered) of the variant position in each cell/barcode.</p></td>
</tr>
<tr class="row-odd"><td><p>FILTER_MASK</p></td>
<td><p>Contains values of 0 (variant passed) or 1 (variant did not pass). Variants that pass the GQ, DP, and AF thresholds listed in the metadata. Applies to only those variants mutated in at least x% of cells (default if 1%).</p></td>
</tr>
<tr class="row-even"><td><p>GQ</p></td>
<td><p>The genotype quality for the variant (from GATK) in each cell/barcode.</p></td>
</tr>
<tr class="row-odd"><td><p>NGT</p></td>
<td><p>The genotype in each cell/barcode. 0: is wildtype, 1: one allele is alternate, 2: both alleles are alternate, 3: Missing genotype (e.g. not enough depth of coverage).</p></td>
</tr>
<tr class="row-even"><td><p>RGQ</p></td>
<td><p>Unconditional reference genotype confidence, encoded as a phred quality -10*log10 p(genotype call is wrong).</p></td>
</tr>
<tr class="row-odd"><td><p>NGT_FILTERED</p></td>
<td><p>The genotype information after calling <cite>assay.dna.filter_variants()</cite> in each cell/barcode. 0: is wildtype, 1: one allele is alternate, 2: both alleles are alternate, 3: Missing genotype (e.g. not enough depth of coverage).</p></td>
</tr>
</tbody>
</table>
</section>
<section id="column-attributes">
<h3>Column attributes<a class="headerlink" href="#column-attributes" title="Permalink to this heading">#</a></h3>
<table class="table" id="id13">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id13" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ALT</p></td>
<td><p>The alternate allele for each variant.</p></td>
</tr>
<tr class="row-odd"><td><p>CHROM</p></td>
<td><p>The chromosome for each variant.</p></td>
</tr>
<tr class="row-even"><td><p>POS</p></td>
<td><p>The genomic position for each variant.</p></td>
</tr>
<tr class="row-odd"><td><p>QUAL</p></td>
<td><p>The genotype quality for each variant (from GATK).</p></td>
</tr>
<tr class="row-even"><td><p>REF</p></td>
<td><p>The reference allele for each variant.</p></td>
</tr>
<tr class="row-odd"><td><p>ado_gt_cells</p></td>
<td><p>The percentage of cells genotyped at the variant’s position. Set to -1 if ADO is not or cannot be estimated for the variant.</p></td>
</tr>
<tr class="row-even"><td><p>ado_rate</p></td>
<td><p>The estimated allelic dropout for the variant. Set to -1 if it cannot be estimated.</p></td>
</tr>
<tr class="row-odd"><td><p>amplicon</p></td>
<td><p>The ID of the amplicon the variant is in.</p></td>
</tr>
<tr class="row-even"><td><p>filtered</p></td>
<td><p>1 if the variant did not passed default filtering, otherwise 0.</p></td>
</tr>
<tr class="row-odd"><td><p>id</p></td>
<td><p>The id of the variant.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="row-attributes">
<h3>Row attributes<a class="headerlink" href="#row-attributes" title="Permalink to this heading">#</a></h3>
<table class="table" id="id14">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id14" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>barcode</p></td>
<td><p>The basepair sequence for the barcode/cell.</p></td>
</tr>
<tr class="row-odd"><td><p>filtered</p></td>
<td><p>1 if the cell/barcode did not passed default filtering, otherwise 0.</p></td>
</tr>
<tr class="row-even"><td><p>sample_name</p></td>
<td><p>The sample name.</p></td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="cnv">
<h2>CNV<a class="headerlink" href="#cnv" title="Permalink to this heading">#</a></h2>
<section id="id2">
<h3>Metadata<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h3>
<table class="table" id="id15">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id15" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample_name</p></td>
<td><p>The sample name.</p></td>
</tr>
<tr class="row-odd"><td><p>ado_rate</p></td>
<td><p>The estimated rate of allelic dropout.</p></td>
</tr>
<tr class="row-even"><td><p>avg_mapping_error_rate</p></td>
<td><p>Ratio of mismatches to bases mapped (see <cite>samtools stats</cite>).</p></td>
</tr>
<tr class="row-odd"><td><p>avg_panel_uniformity</p></td>
<td><p>The proportion of amplicons that pass the panel uniformity test (see <a class="reference external" href="https://support.missionbio.com/hc/en-us/articles/360053187154-Run-Report-Metrics#h_01EHSC7F4BZJ3678B16S0CTVVW">here</a> for more information).</p></td>
</tr>
<tr class="row-even"><td><p>genome_version</p></td>
<td><p>The genome version used for mapping and variant calling.</p></td>
</tr>
<tr class="row-odd"><td><p>n_amplicons</p></td>
<td><p>Number of amplicons in the panel.</p></td>
</tr>
<tr class="row-even"><td><p>n_bases_r1</p></td>
<td><p>Number of bases from the forward reads.</p></td>
</tr>
<tr class="row-odd"><td><p>n_bases_r1_q30</p></td>
<td><p>Number of bases that have a Q value of 30 or greater from the forward reads.</p></td>
</tr>
<tr class="row-even"><td><p>n_bases_r2</p></td>
<td><p>Number of bases from the reverse reads.</p></td>
</tr>
<tr class="row-odd"><td><p>n_bases_r2_q30</p></td>
<td><p>Number of bases that have a Q value of 30 or greater from the reverse reads.</p></td>
</tr>
<tr class="row-even"><td><p>n_cell_barcode_bases</p></td>
<td><p>Total number of bases from cell barcodes.</p></td>
</tr>
<tr class="row-odd"><td><p>n_cell_barcode_bases_q30</p></td>
<td><p>Total number of bases (with quality of 30 or greater) from cell barcodes.</p></td>
</tr>
<tr class="row-even"><td><p>n_cells</p></td>
<td><p>Total number of bases from cell barcodes.</p></td>
</tr>
<tr class="row-odd"><td><p>n_read_pairs</p></td>
<td><p>This is the number of total read pairs in the fastq file.</p></td>
</tr>
<tr class="row-even"><td><p>n_read_pairs_mapped_to_cells</p></td>
<td><p>Number of read pairs that mapped to cells.</p></td>
</tr>
<tr class="row-odd"><td><p>n_read_pairs_trimmed</p></td>
<td><p>Number of read pairs that passed trimming.</p></td>
</tr>
<tr class="row-even"><td><p>n_read_pairs_valid_cell_barcodes</p></td>
<td><p>Number of read pairs from cell barcodes that are valid.</p></td>
</tr>
<tr class="row-odd"><td><p>n_reads_mapped</p></td>
<td><p>Number of read pairs that mapped to the genome.</p></td>
</tr>
<tr class="row-even"><td><p>n_reads_mapped_insert</p></td>
<td><p>Total number of reads that map to the amplicons.</p></td>
</tr>
<tr class="row-odd"><td><p>panel_name</p></td>
<td><p>The panel name.</p></td>
</tr>
<tr class="row-even"><td><p>pipeline_version</p></td>
<td><p>Pipeline version used to produce the data.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="id4">
<h3>Layers<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h3>
<table class="table" id="id16">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id16" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>read_counts</p></td>
<td><p>The total read counts within the amplicon</p></td>
</tr>
</tbody>
</table>
</section>
<section id="id5">
<h3>Column attributes<a class="headerlink" href="#id5" title="Permalink to this heading">#</a></h3>
<table class="table" id="id17">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id17" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>id</p></td>
<td><p>The ID of the amplicon.</p></td>
</tr>
<tr class="row-odd"><td><p>CHROM</p></td>
<td><p>The chromosome for each variant.</p></td>
</tr>
<tr class="row-even"><td><p>start_pos</p></td>
<td><p>The start position of the amplicon.</p></td>
</tr>
<tr class="row-odd"><td><p>end_pos</p></td>
<td><p>The end position of the amplicon.</p></td>
</tr>
<tr class="row-even"><td><p>r1_counts</p></td>
<td><p>The count of R1 reads within the amplicon across all barcodes/cells.</p></td>
</tr>
<tr class="row-odd"><td><p>r2_counts</p></td>
<td><p>The count of R1 reads within the amplicon across all barcodes/cells.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="id6">
<h3>Row attributes<a class="headerlink" href="#id6" title="Permalink to this heading">#</a></h3>
<table class="table" id="id18">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id18" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>barcode</p></td>
<td><p>The basepair sequence for the barcode/cell.</p></td>
</tr>
<tr class="row-odd"><td><p>label</p></td>
<td><p>Per-barcode labels.</p></td>
</tr>
<tr class="row-even"><td><p>sample_name</p></td>
<td><p>The sample name.</p></td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="protein">
<h2>Protein<a class="headerlink" href="#protein" title="Permalink to this heading">#</a></h2>
<section id="id7">
<h3>Metadata<a class="headerlink" href="#id7" title="Permalink to this heading">#</a></h3>
<table class="table" id="id19">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id19" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample_name</p></td>
<td><p>The sample name.</p></td>
</tr>
<tr class="row-odd"><td><p>n_antibodies</p></td>
<td><p>Total number of antibodies measured in the assay.</p></td>
</tr>
<tr class="row-even"><td><p>n_bases_r1</p></td>
<td><p>Number of bases from the forward reads.</p></td>
</tr>
<tr class="row-odd"><td><p>n_bases_r1_q30</p></td>
<td><p>Number of bases that have a Q value of 30 or greater from the forward reads.</p></td>
</tr>
<tr class="row-even"><td><p>n_bases_r2</p></td>
<td><p>Number of bases from the reverse reads.</p></td>
</tr>
<tr class="row-odd"><td><p>n_bases_r2_q30</p></td>
<td><p>Number of bases that have a Q value of 30 or greater from the reverse reads.</p></td>
</tr>
<tr class="row-even"><td><p>n_cell_barcode_bases</p></td>
<td><p>Total number of bases from cell barcodes.</p></td>
</tr>
<tr class="row-odd"><td><p>n_cell_barcode_bases_q30</p></td>
<td><p>Total number of bases (with quality of 30 or greater) from cell barcodes.</p></td>
</tr>
<tr class="row-even"><td><p>n_read_pairs</p></td>
<td><p>This is the number of total read pairs in the fastq file.</p></td>
</tr>
<tr class="row-odd"><td><p>n_read_pairs_after_candidate_barcode_filtering</p></td>
<td><p>Total number of reads mapping to good quality barcodes.</p></td>
</tr>
<tr class="row-even"><td><p>n_reads_trimmed</p></td>
<td><p>Total number of reads that passed trimming.</p></td>
</tr>
<tr class="row-odd"><td><p>n_reads_valid_ab_barcodes</p></td>
<td><p>Total number of reads belonging to valid antibody.</p></td>
</tr>
<tr class="row-even"><td><p>n_reads_valid_cell_barcodes</p></td>
<td><p>Total number of reads belonging to valid cell barcodes.</p></td>
</tr>
<tr class="row-odd"><td><p>panel_name</p></td>
<td><p>The panel name.</p></td>
</tr>
<tr class="row-even"><td><p>pipeline_version</p></td>
<td><p>Pipeline version used to produce the data.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="id8">
<h3>Layers<a class="headerlink" href="#id8" title="Permalink to this heading">#</a></h3>
<table class="table" id="id20">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id20" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>read_counts</p></td>
<td><p>The total read counts within each barcode/antibody combination.</p></td>
</tr>
<tr class="row-odd"><td><p>normalized_counts</p></td>
<td><p>Normalized read counts. Only present after running <cite>sample.protein.normalize_reads()</cite>.</p></td>
</tr>
<tr class="row-even"><td><p>scaled_counts</p></td>
<td><p>Z-score normalized counts. Only present after running <cite>sample.protein.scale_data()</cite>.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="id9">
<h3>Column attributes<a class="headerlink" href="#id9" title="Permalink to this heading">#</a></h3>
<table class="table" id="id21">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id21" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>id</p></td>
<td><p>The ID of the antibody.</p></td>
</tr>
<tr class="row-odd"><td><p>antibody_id</p></td>
<td><p>The ID of the antibody.</p></td>
</tr>
<tr class="row-even"><td><p>antibody_sequence</p></td>
<td><p>The basepair sequenced tagged to the antibody.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="id10">
<h3>Row attributes<a class="headerlink" href="#id10" title="Permalink to this heading">#</a></h3>
<table class="table" id="id22">
<caption><span class="caption-text">Table</span><a class="headerlink" href="#id22" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 66.7%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>barcode</p></td>
<td><p>The basepair sequence for the barcode/cell.</p></td>
</tr>
<tr class="row-odd"><td><p>label</p></td>
<td><p>Per-barcode labels.</p></td>
</tr>
<tr class="row-even"><td><p>sample_name</p></td>
<td><p>The sample name.</p></td>
</tr>
<tr class="row-odd"><td><p>pca</p></td>
<td><p>PCA coordinates for each cell. Only present after running <cite>sample.protein.run_pca()</cite>.</p></td>
</tr>
<tr class="row-even"><td><p>umap</p></td>
<td><p>UMAP coordinates for each cell. Only present after running <cite>sample.protein.run_umap()</cite>.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>


                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="getting_started.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Getting Started</p>
      </div>
    </a>
    <a class="right-next"
       href="vignettes.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Vignettes</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basics">Basics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dna">DNA</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#layers">Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#column-attributes">Column attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#row-attributes">Row attributes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv">CNV</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Column attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Row attributes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#protein">Protein</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Layers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id9">Column attributes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Row attributes</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Saurabh Parikh
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021, Mission Bio.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>