
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Mosaic analysis walk-through &#8212; Mosaic v2.4 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="io" href="../pages/missionbio.mosaic.io.html" />
    <link rel="prev" title="Additional plots and customizations" href="additional-plots-and-customizations.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Mosaic v2.4 documentation</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../1_introduction.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../2_install.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../3_getting_started.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../4_data_structure.html">
   Data structure
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="index.html">
   Code Snippets
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="basics.html">
   Basic usage of mosaic
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cnv.html">
   Manual CNV analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="filtering.html">
   Filtering barcodes and ids
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="customizations.html">
   Plot customizations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="multi-sample-analysis.html">
   Multisample analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="additional-plots-and-customizations.html">
   Additional plots and customizations
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Mosaic analysis walk-through
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modules
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.io.html">
   io
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.utils.html">
   utils
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pages/missionbio.mosaic.constants.html">
   constants
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.COLORS.html">
     missionbio.mosaic.constants.COLORS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.PCA_LABEL.html">
     missionbio.mosaic.constants.PCA_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.UMAP_LABEL.html">
     missionbio.mosaic.constants.UMAP_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.SCALED_LABEL.html">
     missionbio.mosaic.constants.SCALED_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.LABEL.html">
     missionbio.mosaic.constants.LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.READS.html">
     missionbio.mosaic.constants.READS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.PALETTE.html">
     missionbio.mosaic.constants.PALETTE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pages/missionbio.mosaic.constants.NORMALIZED_READS.html">
     missionbio.mosaic.constants.NORMALIZED_READS
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Classes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.assay._Assay.html">
   _Assay
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.dna.Dna.html">
   Dna
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.cnv.Cnv.html">
   Cnv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.protein.Protein.html">
   Protein
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.sample.Sample.html">
   Sample
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/missionbio.mosaic.samplegroup.SampleGroup.html">
   SampleGroup
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/examples/analysis-walkthrough.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dna-analysis">
   DNA Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#basic-filtering">
     Basic filtering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#subsetting-data-for-variants-of-interest">
     Subsetting Data for Variants of Interest
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#annotation-addition">
     Annotation Addition
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#manual-variant-selection">
     Manual Variant Selection
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#clustering">
     Clustering
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#clustering-method-1">
       Clustering Method 1
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#clustering-method-2-pca-umap-clustering">
       Clustering Method 2 (PCA + UMAP + Clustering)
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#renaming-subclones-labels">
     Renaming Subclones (Labels)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cnv-analysis">
   CNV Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#filtering-and-normalization">
     Filtering And Normalization
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#genotype-guided-ploidy-clustering">
     Genotype-Guided Ploidy Clustering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#read-based-ploidy-clustering">
     Read-Based Ploidy Clustering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#statistics">
     Statistics
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#protein-analysis">
   Protein Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#normalization-and-data-inspection">
     Normalization And Data Inspection
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#clustering-and-visualization">
     Clustering And Visualization
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Statistics
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#combined-visualizations">
   Combined visualizations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-new-sample-object">
     Create a new sample object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#built-in-visualizations-for-combining-analytes">
     Built-in visualizations for combining analytes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#additional-visualizations">
     Additional visualizations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quantify-overlaps-between-dna-clones-and-protein-clusters">
     Quantify overlaps between DNA clones and protein clusters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#export-and-save-data">
     Export and Save Data
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Mosaic analysis walk-through</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dna-analysis">
   DNA Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#basic-filtering">
     Basic filtering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#subsetting-data-for-variants-of-interest">
     Subsetting Data for Variants of Interest
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#annotation-addition">
     Annotation Addition
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#manual-variant-selection">
     Manual Variant Selection
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#clustering">
     Clustering
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#clustering-method-1">
       Clustering Method 1
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#clustering-method-2-pca-umap-clustering">
       Clustering Method 2 (PCA + UMAP + Clustering)
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#renaming-subclones-labels">
     Renaming Subclones (Labels)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cnv-analysis">
   CNV Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#filtering-and-normalization">
     Filtering And Normalization
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#genotype-guided-ploidy-clustering">
     Genotype-Guided Ploidy Clustering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#read-based-ploidy-clustering">
     Read-Based Ploidy Clustering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#statistics">
     Statistics
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#protein-analysis">
   Protein Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#normalization-and-data-inspection">
     Normalization And Data Inspection
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#clustering-and-visualization">
     Clustering And Visualization
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Statistics
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#combined-visualizations">
   Combined visualizations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-new-sample-object">
     Create a new sample object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#built-in-visualizations-for-combining-analytes">
     Built-in visualizations for combining analytes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#additional-visualizations">
     Additional visualizations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quantify-overlaps-between-dna-clones-and-protein-clusters">
     Quantify overlaps between DNA clones and protein clusters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#export-and-save-data">
     Export and Save Data
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="mosaic-analysis-walk-through">
<h1>Mosaic analysis walk-through<a class="headerlink" href="#mosaic-analysis-walk-through" title="Permalink to this headline">#</a></h1>
<p><b>Objective</b>: This notebook contains a detailed step by step process of analyzing Tapestri runs. It is not a vignette to demonstrate how to use mosaic, but rather a guided analysis. The following cells have not been exectued but the notebook can be downloaded and run on any sample.</p>
<section id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">#</a></h2>
<p><b>Topics covered</b><br></p>
<ol class="arabic simple">
<li><p>Loading required packages and data.</p></li>
<li><p>Structure and contents of data objects.</p></li>
</ol>
<p><b>Load data</b><br></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import mosaic libraries</span>
<span class="kn">import</span> <span class="nn">missionbio.mosaic</span> <span class="k">as</span> <span class="nn">ms</span>

<span class="c1"># Import these to display entire dataframes</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>

<span class="c1"># Import graph_objects from the plotly package to display figures when saving the notebook as an HTML</span>
<span class="c1"># Import numpy for statistics</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># Import additional packages for specific visuals</span>
<span class="kn">import</span> <span class="nn">missionbio.mosaic.utils</span> <span class="k">as</span> <span class="nn">mutils</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># Import the colors</span>
<span class="kn">from</span> <span class="nn">missionbio.mosaic.constants</span> <span class="kn">import</span> <span class="n">COLORS</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1"># Note: when exporting the notebook as an HTML, plots that use the &quot;go.Figure(fig)&quot; command are saved</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check version; this notebook is designed for Mosaic 2.0 or higher</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Any function&#39;s parameters and default values can be looked up via the &#39;help&#39; function</span>
<span class="c1"># Here, the function is &#39;ms.load&#39;</span>
<span class="n">help</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">load</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the h5 file to be used in this analysis: h5path = &#39;/path/to/h5/file/test.h5&#39;</span>
<span class="c1"># If working with Windows, you may need to add an &#39;r&#39; before the path: h5path = r&#39;/path/to/h5/file/test.h5&#39;</span>
<span class="n">h5path</span> <span class="o">=</span> <span class="s1">&#39;/Users/robert_durruthy/Desktop/Mosaic notebooks/4-cell-lines-AML-multiomics.dna+protein.h5&#39;</span>

<span class="c1"># Load the data</span>
<span class="c1"># sample = ms.load(h5path, raw=False, apply_filter=True, single=True, whitelist = [])</span>

<span class="n">sample</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">load_example_dataset</span><span class="p">(</span><span class="s2">&quot;3 cell mix&quot;</span><span class="p">)</span>  <span class="c1"># Use the above `ms.load` for custom h5 files</span>

<span class="c1"># Always set raw=False; if raw=True, ALL barcodes will be loaded (rather than cell-associated barcodes)</span>
<span class="c1"># Always set apply_filter=True unless you can&#39;t detect an expected (target) variant. Additional filtering options are included in the DNA section below</span>
<span class="c1"># The single=True option loads multi-sample h5 files as a single sample object (compatible with this notebook)</span>
<span class="c1"># The whitelist option loads any variant that is in the vcf.gz file (e.g. &quot;chr1:179520511:C/G&quot;); similar to whitelist feature in Tapestri Insights v2</span>
</pre></div>
</div>
</div>
</div>
<p>All the interactive plotting functions return a plotly figure. If the layout or the color
scheme is not suitable for your data type, they can be changed before creating the final figure.</p>
<p>The colors for the plots are stored either in the individual traces or the layout attributes of the plotly figure.</p>
<p>Mosaic also contains a list of colors that can be used to customize the plots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Explore the colors</span>
<span class="c1"># Additional color palettes: https://seaborn.pydata.org/tutorial/color_palettes.html</span>
<span class="c1"># Plot the first few colors</span>
<span class="n">sns</span><span class="o">.</span><span class="n">palplot</span><span class="p">(</span><span class="n">COLORS</span><span class="p">[:</span><span class="mi">40</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Alternatively plot another palette</span>
<span class="n">sns</span><span class="o">.</span><span class="n">palplot</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;magma&quot;</span><span class="p">,</span> <span class="n">n_colors</span><span class="o">=</span><span class="mi">20</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Print the corresponding hex codes</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;magma&quot;</span><span class="p">,</span> <span class="n">n_colors</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">as_hex</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<p><b>Data Structure</b></p>
<p>DNA, CNV, and Protein are sub-classes of the Assay class. The information is stored in four categories, and the user can modify each of those:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>1. metadata (add_metadata / del_metadata):
    - dictionary containing metrics of the assay

2. row_attrs (add_row_attr / del_row_attr):
    - dictionary which contains &#39;barcode&#39; as one of the keys 
      (where the value is a list of all barcodes)
    - for all other keys, the values must be of the same 
      length, i.e. match the number of barcodes
    - this is the attribute where &#39;label&#39;, &#39;pca&#39;, 
      and &#39;umap&#39; values are added

3. col_attrs (add_col_attr / del_col_attr):
    - dictionary which contains &#39;id&#39; as one of the keys 
      (where the value is a list of all ids)
    - for DNA assays, &#39;ids&#39; are variants; for Protein assays, 
      &#39;ids&#39; are antibodies
    - for all other keys, the values must be of the same
      length, i.e. match the number ids

4. layers (add_layer / del_layer):
    - dictionary which contains critically important assay metrics
    - all the values have the shape (num barcodes) x (num ids) 
    - for DNA assays, this includes AF, GQ, DP, etc. 
      (per cell, per variant)
    - for Protein assays, this includes read counts 
      (per cell, per antibody)
    - this is the attribute where &#39;normalized_counts&#39; will be added
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Summary of DNA assay </span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">sample.dna</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">row_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">col_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">layers</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">metadata</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="s2">&quot;: &quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Summary of Protein assay </span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">sample.protein</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">row_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">col_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">layers</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">metadata</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="s2">&quot;: &quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For DNA, ids are variants</span>
<span class="c1"># sample.dna.ids() is a shortcut for sample.dna.col_attrs[&#39;id&#39;]</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For Protein, ids are AOCs</span>
<span class="c1"># sample.protein.ids() is a shortcut for sample.protein.col_attrs[&#39;id&#39;]</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="dna-analysis">
<h2>DNA Analysis<a class="headerlink" href="#dna-analysis" title="Permalink to this headline">#</a></h2>
<p><b>Topics covered</b><br></p>
<ol class="arabic simple">
<li><p>Standard filtering of DNA variants.</p></li>
<li><p>Subsetting dataset for variants of interest, including whitelisted variants.</p></li>
<li><p>Addition of annotations to the variants.</p></li>
<li><p>Manual variant selection. [OPTIONAL]</p></li>
<li><p>Clustering sub-clones.</p></li>
</ol>
<section id="basic-filtering">
<h3>Basic filtering<a class="headerlink" href="#basic-filtering" title="Permalink to this headline">#</a></h3>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>There are many options for filtering DNA variants. 
Use the help() function to understand the approach listed below.
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For additional information visit: https://support.missionbio.com/hc/en-us/articles/360047303654-How-do-the-Advanced-Filters-work-</span>
<span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter the variants, similar to the &quot;Advanced Filters&quot; in Tapestri Insights v2.2</span>

<span class="c1"># One major difference: The filter &quot;REMOVE GENOTYPE IN CELLS WITH ALTERNATE ALLELE FREQ&quot; </span>
<span class="c1"># is replaced with the following 3 zygosity-specific filters: vaf_ref, vaf_hom, vaf_het</span>

<span class="c1"># In general, these additional filters remove additional false-positive from the data.</span>

<span class="c1"># Define dna_vars variable</span>
<span class="n">dna_vars</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">()</span>
<span class="n">dna_vars</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check the number of filtered variants. When using the default filters, the number of </span>
<span class="c1"># variants is likely smaller compared to the originally loaded variants due to the more </span>
<span class="c1"># stringent filtering criteria (e.g., vaf_ref=5, vaf_hom=95, vaf_het=35).</span>
<span class="nb">len</span><span class="p">(</span><span class="n">dna_vars</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Adjust filters if needed by overwriting dna_vars</span>
<span class="n">dna_vars</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">(</span>
    <span class="n">min_dp</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">min_gq</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
    <span class="n">vaf_ref</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">vaf_hom</span><span class="o">=</span><span class="mi">95</span><span class="p">,</span>
    <span class="n">vaf_het</span><span class="o">=</span><span class="mi">35</span><span class="p">,</span>
    <span class="n">min_prct_cells</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">min_mut_prct_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># List all filtered variants</span>
<span class="n">dna_vars</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Re-check the number of filtered variants</span>
<span class="nb">len</span><span class="p">(</span><span class="n">dna_vars</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="subsetting-data-for-variants-of-interest">
<h3>Subsetting Data for Variants of Interest<a class="headerlink" href="#subsetting-data-for-variants-of-interest" title="Permalink to this headline">#</a></h3>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>First, specify variants of interest using one of the three options below:
    1. Use the filtered variants from the above section (dna_vars)
    2. Use specific variants of interest (whitelist) 
    3. Combine 1 &amp; 2: filtered variants plus whitelist

Then, this list (final_vars) is used to reduce the larger data set to only your variants of interest.
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Option 1: filtered variants only (no whitelist)</span>

<span class="n">final_vars</span><span class="o">=</span><span class="n">dna_vars</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Option 2: whitelisted variants only </span>

<span class="c1"># Specify the whitelist; variants may be copy/pasted from Tapestri Insights v2.2,</span>
<span class="c1"># but ensure correct nomenclature, ie whitelist = [&quot;chr13:28589657:T/G&quot;,&quot;chrX:39921424:G/A&quot;]</span>
<span class="n">final_vars</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr13:28597686:G/A&quot;</span><span class="p">,</span><span class="s2">&quot;chr11:32421651:T/C&quot;</span><span class="p">,</span><span class="s2">&quot;chr4:106158216:G/A&quot;</span><span class="p">,</span><span class="s2">&quot;chr4:106196829:T/G&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Option 3: filtered variants plus whitelisted variants</span>

<span class="c1"># Specify the whitelist; variants may be copy/pasted from Tapestri Insights v2.2,</span>
<span class="c1"># but ensure correct nomenclature, ie whitelist = [&quot;chr13:28589657:T/G&quot;,&quot;chrX:39921424:G/A&quot;]</span>
<span class="n">target_variants</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr13:28597686:G/A&quot;</span><span class="p">,</span><span class="s2">&quot;chr11:32421651:T/C&quot;</span><span class="p">,</span><span class="s2">&quot;chr4:106158216:G/A&quot;</span><span class="p">,</span><span class="s2">&quot;chr4:106196829:T/G&quot;</span><span class="p">]</span>

<span class="c1"># Combine whitelisted and filtered variants</span>
<span class="n">final_vars</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">dna_vars</span><span class="p">)</span> <span class="o">+</span> <span class="n">target_variants</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check the lenght of your final list of variants</span>
<span class="nb">len</span><span class="p">(</span><span class="n">final_vars</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dimensionality of the original sample.dna dataframe</span>
<span class="c1"># First number = number of cells (rows); second number = number of variants (columns)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Before subsetting, verify that all the chosen variants are in the current sample.dna ids (should return True)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">final_vars</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">())))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subsetting sample.dna (columns) based on reduced variant list. Keeping all cells that passed filtering</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">[</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(),</span> <span class="n">final_vars</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check the shape of the final filtered DNA object, i.e. (number of barcodes/cells, number of ids/variants)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotation-addition">
<h3>Annotation Addition<a class="headerlink" href="#annotation-addition" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_annotations</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Fetch annotations using varsome</span>
<span class="c1"># Note: run this on a filtered DNA sample - too many variants (e.g., 100+) are not handled correctly by the method</span>
<span class="n">annotation</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_annotations</span><span class="p">()</span>  

<span class="c1"># Store the annotations in the dna assay as a new column attribute</span>
<span class="k">for</span> <span class="n">col</span><span class="p">,</span> <span class="n">content</span> <span class="ow">in</span> <span class="n">annotation</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_col_attr</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">content</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sort the values as desired and display a table of all variants (similar to Tapestri Insights)</span>
<span class="n">ann</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;DANN&quot;</span><span class="p">,</span> <span class="s2">&quot;Coding impact&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">ann</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The data can also be presented as an interactive table</span>
<span class="c1"># There are multiple pages to the table - check top right corner for page number</span>
<span class="c1"># Clicking will select a particular row and store it in table.selected_rows</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">Table</span><span class="p">(</span><span class="n">ann</span><span class="p">)</span>
<span class="n">table</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check the selected variants</span>
<span class="c1"># If desired, this list can be used to subset the data further (similar to the subsetting done above with &quot;final_vars&quot;)</span>
<span class="n">table</span><span class="o">.</span><span class="n">selected_rows</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Any of these column values can be added to the id names</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add annotation to the id names</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">set_ids_from_cols</span><span class="p">([</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;ALT&quot;</span><span class="p">])</span>

<span class="c1"># Annotations are now added to the variants</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>

<span class="c1"># Use sample.dna.reset_ids() to get the original ids</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="manual-variant-selection">
<h3>Manual Variant Selection<a class="headerlink" href="#manual-variant-selection" title="Permalink to this headline">#</a></h3>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>If using only whitelisted variants (e.g., target variants are already known), this section can be skipped.

This section includes a variety of plots to help assess variant quality.
Heatmaps are interactive. Clicking on a column selects the corresponding id,
whose value is stored in the `selected_ids` attribute. Selected ids can then 
be removed from the data set. I.e., click on the low quality variants you wish to discard.
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">stripplot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># First diagnostic plot to evaluate variant quality</span>
<span class="c1"># The &#39;attribute&#39; or &#39;colorby&#39; arguments may be changed</span>
<span class="c1"># Variants with the same genotype for every cell are likley germline</span>
<span class="c1"># Variants with a HET population distributed below AF=35 are likely false positives (expect HET cells around AF=50)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">stripplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;AF_MISSING&#39;</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;GQ&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Second diagnostic plot: use this heatmap to (de)-select variants with little variance </span>
<span class="c1"># For example, a germline mutation with identical genotype across all cells</span>
<span class="c1"># Additionally, variants with excessive missing data can be selected for removal</span>
<span class="c1"># Variants may be selected by clicking on the heatmap (variant name changes color from black to red)</span>
<span class="c1"># Don&#39;t use &quot;fig =&quot; as the heatmap won&#39;t become responsive for variant selection</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Array that lists all variants that were selected in the heatmap</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">selected_ids</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subset the sample.dna variable to removing all selected variants</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">selected_ids</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">selected_ids</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Redraw the heatmap with GQ values to select variants that display low quality in majority of cells</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;GQ&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Array that lists all variants that were selected in the heatmap</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">selected_ids</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subset the sample.dna variable to removing all selected variants</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">selected_ids</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">selected_ids</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Confirm new number of columns (variants)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Redraw heatmap with undesired variants removed</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;AF_MISSING&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="clustering">
<h3>Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline">#</a></h3>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The DNA assay class comes with three different methods of clustering:
Method 1: f(x) = group_by_genotype (akin to Tapestri Insights v2.2)
Method 2: PCA + UMAP + clustering (customizable)
</pre></div>
</div>
<section id="clustering-method-1">
<h4>Clustering Method 1<a class="headerlink" href="#clustering-method-1" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">group_by_genotype</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clustering Method #1</span>

<span class="c1"># Recommended for 1-10 variants</span>
<span class="c1"># Cluster with Tapestri Insights v2.2 count-based method </span>

<span class="c1"># The created table includes a &#39;score&#39; row that intends to help with the identification of allelic dropout (ADO) clones</span>
<span class="c1"># Scores greater than 0.8 are typically considered artifacts (ADO) and may be labeled as such to be discarded</span>
<span class="c1"># Clones are ordered based on their size (1=largest clone, 2=second largest clone, etc.)</span>

<span class="n">clone_data</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">group_by_genotype</span><span class="p">(</span>
    <span class="n">features</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()[</span><span class="mi">4</span><span class="p">:</span><span class="mi">6</span><span class="p">],</span>
    <span class="n">group_missing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">min_clone_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;NGT_FILTERED&quot;</span><span class="p">,</span>
    <span class="n">show_plot</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>

<span class="c1"># Display as a static html or interactive table using ms.Table.</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">clone_data</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot heatmap using NGT_FILTERED.</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Store the clustering in the row attributes</span>

<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_row_attr</span><span class="p">(</span><span class="s2">&quot;clustering-1&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="clustering-method-2-pca-umap-clustering">
<h4>Clustering Method 2 (PCA + UMAP + Clustering)<a class="headerlink" href="#clustering-method-2-pca-umap-clustering" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># CLUSTERING METHOD #2</span>
<span class="c1"># An alternative clustering recommended for high-feature/dimensional space (e.g. 10+ variants)</span>

<span class="c1"># First: run PCA on filtered data frame (all cells, chosen variants) using the AF_MISSING layer, for instance</span>
<span class="c1"># Note: AF_MISSING includes -50 values for all 0 values for which the data was filtered and is missing</span>
<span class="c1"># The number of components should equal the number of pre-filtered variants</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">components</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;AF_MISSING&#39;</span><span class="p">,</span><span class="n">show_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> 

<span class="c1"># Assess &#39;elbow&#39; plot and determine number of PCs where the distribution starts to plateau</span>
<span class="c1"># Note that PC1 = 0 in the plot</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rerun PCA with optimal number of components based on elbow plot analysis</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">components</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;AF_MISSING&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run UMAP on top of the newly created PC dataframe</span>
<span class="c1"># See https://jlmelville.github.io/uwot/abparams.html for appropriate spread and min_dist values</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span> <span class="c1">#, min_dist=0.2, spread=1.5)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Review other clustering methods</span>
<span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">cluster</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize data and cluster it using different methods</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;dbscan&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Re-plot UMAP projection w/ alternative clustering results</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot heatmap using NGT</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Store this clustering in the row attributes</span>

<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_row_attr</span><span class="p">(</span><span class="s2">&quot;clustering-2&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<p><b>The next few commands allow you to customize the colors of the genotypes in the heatmap.</b></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">palplot</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span> <span class="n">n_colors</span><span class="o">=</span><span class="mi">20</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vir20</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span> <span class="n">n_colors</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">as_hex</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">vir20</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vir20</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The DNA heatmap and scatterplot colors are stored in the layout.coloraxis.colorscale attribute</span>

<span class="c1"># This value must be updated to customize the plot</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assuming these are new desired colors</span>
<span class="c1"># NGT=0 (WT) - blue</span>
<span class="c1"># NGT=1 (HET) - orange</span>
<span class="c1"># NGT=2 (HOM) - red</span>
<span class="c1"># NGT=3 (missing) - black</span>

<span class="c1"># Additional information re: color palettes here: https://seaborn.pydata.org/tutorial/color_palettes.html</span>
<span class="n">wt_col</span> <span class="o">=</span> <span class="n">vir20</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">het_col</span> <span class="o">=</span> <span class="n">vir20</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="n">hom_col</span> <span class="o">=</span> <span class="n">vir20</span><span class="p">[</span><span class="mi">19</span><span class="p">]</span>
<span class="n">miss_col</span> <span class="o">=</span> <span class="n">COLORS</span><span class="p">[</span><span class="mi">19</span><span class="p">]</span>

<span class="n">sns</span><span class="o">.</span><span class="n">palplot</span><span class="p">([</span><span class="n">wt_col</span><span class="p">,</span> <span class="n">het_col</span><span class="p">,</span> <span class="n">hom_col</span><span class="p">,</span> <span class="n">miss_col</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Update the coloraxis to make a plot with the new colors</span>
<span class="n">new_colors</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">0</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">wt_col</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">wt_col</span><span class="p">),</span>
              <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">het_col</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">het_col</span><span class="p">),</span>
              <span class="p">(</span><span class="mi">2</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">hom_col</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">hom_col</span><span class="p">),</span>
              <span class="p">(</span><span class="mi">3</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">miss_col</span><span class="p">),</span> <span class="p">(</span><span class="mi">4</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="n">miss_col</span><span class="p">)]</span>

<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="n">new_colors</span>
<span class="n">fig</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="renaming-subclones-labels">
<h3>Renaming Subclones (Labels)<a class="headerlink" href="#renaming-subclones-labels" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use clustering-1 for further analysis</span>

<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">set_labels</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;clustering-1&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rename clusters based on genotypes</span>
<span class="c1"># Merge clusters by giving them the same name</span>
<span class="c1"># Rename clusters identically that are to be merged into one and discarded (e.g. &quot;FP&quot; for false-positive)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">rename_labels</span><span class="p">(</span>
  <span class="p">{</span>
    <span class="s2">&quot;1&quot;</span><span class="p">:</span> <span class="s2">&quot;Cell 1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;2&quot;</span><span class="p">:</span> <span class="s2">&quot;Cell 2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;3&quot;</span><span class="p">:</span> <span class="s2">&quot;Cell 3&quot;</span><span class="p">,</span>
    <span class="s2">&quot;missing&quot;</span><span class="p">:</span> <span class="s2">&quot;FP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;small&quot;</span><span class="p">:</span> <span class="s2">&quot;FP&quot;</span><span class="p">,</span>
  <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove barcodes (clones) from data based on renamed labels</span>
<span class="c1"># The reduced data set will now be called &#39;sample.dna2&#39;</span>
<span class="k">if</span> <span class="s2">&quot;FP&quot;</span> <span class="ow">in</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">():</span>
    <span class="n">fp_barcodes</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">({</span><span class="s2">&quot;FP&quot;</span><span class="p">})</span>
    <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">fp_barcodes</span><span class="p">)</span> 
<span class="k">else</span><span class="p">:</span>
    <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span>

<span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Redraw heatmap</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Redraw again with new (previously defined) color palette</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">,</span><span class="n">splitby</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="n">new_colors</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># If analyzing multiple samples, re-plot heatmap with barcodes ordered based on sample</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">,</span><span class="n">splitby</span><span class="o">=</span><span class="s2">&quot;sample_name&quot;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Adding write_image function saves the picture in high-res</span>
<span class="c1"># See help() to change resolution or file type </span>
<span class="n">help</span><span class="p">(</span><span class="n">fig</span><span class="o">.</span><span class="n">write_image</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="o">.</span><span class="n">write_image</span><span class="p">(</span><span class="s2">&quot;heatmap_example.svg&quot;</span><span class="p">,</span><span class="n">scale</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Evaluate new total number of cells after the above filtering</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compare to original cell number</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="cnv-analysis">
<h2>CNV Analysis<a class="headerlink" href="#cnv-analysis" title="Permalink to this headline">#</a></h2>
<p><b>Topics covered</b></p>
<ol class="arabic simple">
<li><p>Amplicon filtering and read normalization.</p></li>
<li><p>Genotype-guided ploidy computation and CNV clustering.</p></li>
<li><p>Read-based ploidy computation and CNV clustering.</p></li>
<li><p>Statistical significance analysis.</p></li>
</ol>
<section id="filtering-and-normalization">
<h3>Filtering And Normalization<a class="headerlink" href="#filtering-and-normalization" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filtering Amplicons</span>
<span class="c1"># This returns a table of the reads for each amplicon in each cell</span>
<span class="n">reads</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="s1">&#39;read_counts&#39;</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row+col&#39;</span><span class="p">)</span>
<span class="n">reads</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Only amplicons found in more than half the cells are analyzed </span>
<span class="c1"># The other amplicons are dropped</span>
<span class="c1"># Note: optional for samples with expected biological missing data</span>
<span class="n">working_amplicons</span> <span class="o">=</span> <span class="p">(</span><span class="n">reads</span><span class="o">.</span><span class="n">median</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[:,</span> <span class="n">working_amplicons</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Additionally, we keep only valid barcodes from the DNA analysis and are storing the data</span>
<span class="c1"># in a new variable sample.cnv2 (e.g., in order to not overwrite sample.cnv)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(),:]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compare amplicon number and cell number original</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compare amplicon number and cell number</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Normalize the reads </span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="genotype-guided-ploidy-clustering">
<h3>Genotype-Guided Ploidy Clustering<a class="headerlink" href="#genotype-guided-ploidy-clustering" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assuming the presence of a cell population that is considered diploid across</span>
<span class="c1"># all the amplicons, we can compute the ploidy of all other cell populations (previously defined)</span>

<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">compute_ploidy</span><span class="p">(</span><span class="n">diploid_cells</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(</span><span class="s1">&#39;Cell 1&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign the DNA labels to the CNV assay class</span>
<span class="c1"># We want to ensure the labels (e.g., names of cell populations) are identical across assay classes</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">set_labels</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">set_palette</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">get_palette</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Heatmap with the features ordered by the default amplicon order</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s1">&#39;positions&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Scale the figure width and plot as a static image</span>
<span class="c1"># Double click on the plot to zoom-in and improve the resolution</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s1">&#39;positions&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">1600</span>
<span class="n">mutils</span><span class="o">.</span><span class="n">static_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Heatmap for a subset of the chromosomes</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;7&#39;</span><span class="p">,</span> <span class="s1">&#39;17&#39;</span><span class="p">,</span> <span class="s1">&#39;20&#39;</span><span class="p">])</span>

<span class="c1"># Optionally, restrict the range of ploidy values based on observed/expected CNV events (commented out)</span>
<span class="c1">#fig.layout.coloraxis.cmax = 4</span>
<span class="c1">#fig.layout.coloraxis.cmin = 0</span>

<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Heatmap with the features grouped by the genes</span>
<span class="c1"># The first time this runs, it fetches the gene names from ensembl</span>
<span class="c1"># The annotation can also be fetched using sample.cnv.get_gene_names()</span>
<span class="c1"># The plots can also be smoothed using a moving average with the convolve parameter</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="n">convolve</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Optionally, restrict the range of ploidy values based on observed/expected CNV events (commented out)</span>
<span class="c1">#fig.layout.coloraxis.cmax = 4</span>
<span class="c1">#fig.layout.coloraxis.cmin = 0</span>

<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Change the color scale to &quot;magma&quot; - other suitable options might be &quot;viridis&quot;, &quot;plasma&quot;, &quot;blues&quot;, &quot;blues_r&quot; ...</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="s1">&#39;magma&#39;</span>

<span class="c1"># Update the separating lines to be black</span>
<span class="k">for</span> <span class="n">shape</span> <span class="ow">in</span> <span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">shapes</span><span class="p">:</span>
    <span class="n">shape</span><span class="o">.</span><span class="n">line</span><span class="o">.</span><span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;#000000&#39;</span>

<span class="c1"># Optionally, restrict the range of ploidy values based on observed/expected CNV events (commented out)</span>
<span class="c1">#fig.layout.coloraxis.cmax = 4</span>
<span class="c1">#fig.layout.coloraxis.cmin = 0</span>

<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Heatmap for a subset of the genes</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ASXL1&quot;</span><span class="p">,</span> <span class="s2">&quot;EZH2&quot;</span><span class="p">,</span><span class="s1">&#39;TP53&#39;</span><span class="p">])</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="s1">&#39;magma&#39;</span>

<span class="c1"># Update the separating lines to be black</span>
<span class="k">for</span> <span class="n">shape</span> <span class="ow">in</span> <span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">shapes</span><span class="p">:</span>
    <span class="n">shape</span><span class="o">.</span><span class="n">line</span><span class="o">.</span><span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;#000000&#39;</span>

<span class="c1"># Optionally, restrict the range of ploidy values based on observed/expected CNV events (commented out)</span>
<span class="c1">#fig.layout.coloraxis.cmax = 2</span>
<span class="c1">#fig.layout.coloraxis.cmin = 0</span>

<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">plot_ploidy</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Ploidy line plot across amplicons for all WT cells</span>
<span class="c1"># WT clone cells normalized to ploidy = 2 by default</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">plot_ploidy</span><span class="p">(</span><span class="s1">&#39;Cell 1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Same as above, only using different feature groupings (genes+amplicons)</span>
<span class="c1"># Red dots represent the median per gene, blue dots represent the individual amplicons in each gene</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">plot_ploidy</span><span class="p">(</span><span class="s1">&#39;Cell 1&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s2">&quot;genes+amplicons&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Evaluate ploidy for other genotype-based defined clones by updating the code</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">plot_ploidy</span><span class="p">(</span><span class="s1">&#39;Cell 2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Evaluate ploidy for other genotype-based defined clones by updating the code</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">plot_ploidy</span><span class="p">(</span><span class="s1">&#39;Cell 2&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s2">&quot;genes+amplicons&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="read-based-ploidy-clustering">
<h3>Read-Based Ploidy Clustering<a class="headerlink" href="#read-based-ploidy-clustering" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster the CNV data using the reads via PCA/UMAP</span>
<span class="c1"># Note: all cell barcodes are now used, including cells that were previously filtered out due to incomplete genotype data</span>

<span class="c1"># First normalize the reads and plot a heatmap (which also adds the annotations)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">set_labels</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">()</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>  <span class="n">features</span><span class="o">=</span><span class="s1">&#39;genes&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## OPTIONAL</span>

<span class="c1"># Clustering with all amplicons can often result in low resolution of CNV events</span>
<span class="c1"># If genes or chromosomes of interest are known, you can use these features for clustering</span>
<span class="n">amps_of_interest</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;TP53&quot;</span><span class="p">,</span> <span class="s2">&quot;EZH2&quot;</span><span class="p">,</span> <span class="s2">&quot;TET2&quot;</span><span class="p">])</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[:,</span> <span class="n">amps_of_interest</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The appropriate number of components for dimensionality reduction can be determined using the elbow plot</span>
<span class="c1"># We start with components = the number of amplicons</span>
<span class="n">amp_number</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">components</span><span class="o">=</span><span class="n">amp_number</span><span class="p">,</span> <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span><span class="n">show_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># After assessing the elbow plot, re-run PCA with your chosen number of PCs</span>
<span class="c1"># Too many PCA components may result in merging of clusters</span>
<span class="c1"># Too few PCA components may result in splitting of clusters</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">components</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span><span class="n">show_plot</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run UMAP on top of the newly created PC dataframe</span>
<span class="c1"># See https://jlmelville.github.io/uwot/abparams.html for appropriate spread/min_dist values</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span> <span class="c1">#, min_dist=0.2, spread=1.5)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster data using different methods</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;graph-community&#39;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the UMAP scatterplot colored by cluster</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span><span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># If analyzing multiple samples, color the UMAP by sample</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span><span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;sample_name&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the heatmap of normalized counts, stratified by cluster</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Note, two different versions of the CNV subclass exist now:</p>
<ul class="simple">
<li><p>sample.cnv: contains data based on the read-based ploidy calculation</p></li>
<li><p>sample.cnv2: contains data based on the genotype-guided ploidy calculation</p></li>
</ul>
</section>
<section id="statistics">
<h3>Statistics<a class="headerlink" href="#statistics" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_cnv</span><span class="p">,</span> <span class="n">tstat_cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span><span class="o">.</span><span class="n">test_signature</span><span class="p">(</span><span class="s2">&quot;normalized_counts&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_cnv</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_cnv</span> <span class="o">=</span> <span class="n">pval_cnv</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">**</span> <span class="o">-</span><span class="mi">50</span>
<span class="n">pvals_cnv</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pval_cnv</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">tstat_cnv</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Colored tiles indicate strong association of a ploidy with a particular cluster</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pvals_cnv</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="protein-analysis">
<h2>Protein Analysis<a class="headerlink" href="#protein-analysis" title="Permalink to this headline">#</a></h2>
<p><b>Topics covered</b></p>
<ol class="arabic simple">
<li><p>Normalization and assessment of AOC abundance</p></li>
<li><p>Clustering and visualization</p></li>
<li><p>Statistical significance analysis</p></li>
</ol>
<section id="normalization-and-data-inspection">
<h3>Normalization And Data Inspection<a class="headerlink" href="#normalization-and-data-inspection" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Normalize reads</span>
<span class="c1"># Three different methods available: CLR = Centered Log Ratio</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">(</span><span class="s1">&#39;CLR&#39;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Ridge plot: shows distribution of AOC counts across all cells (bimodal = positive and negative cell populations)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ridgeplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
                         <span class="n">features</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plotting combinations of AOCs on a biaxial scatterplot, to mimic FACS data</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">feature_scatter</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">])</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#List all AOCs</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Optionally, remove AOCs from the data set prior to clustering </span>
<span class="c1"># For example, those that don&#39;t display any signal in any cells</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;Mouse IgG1k&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="clustering-and-visualization">
<h3>Clustering And Visualization<a class="headerlink" href="#clustering-and-visualization" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_pca</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clustering similar to &quot;CLUSTERING METHOD #2&quot; of DNA clustering using genotypes</span>
<span class="c1"># Select fewer components when analyzing custom panels with 2-20 AOCs</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span><span class="n">show_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rerun PCA with optimal number of components based on elbow plot analysis</span>
<span class="c1"># Typically, components = 10 is appropriate for 45-plex Biolegend panel</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">show_plot</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now run UMAP on the chosen PCs </span>
<span class="c1"># UMAPs rely on an initial randomization which leads to different projections every time</span>
<span class="c1"># To address this, pass random_state to the run_umap method</span>
<span class="c1"># See https://jlmelville.github.io/uwot/abparams.html for appropriate spread/min_dist values</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span> <span class="c1">#, spread=, min_dist=</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster the AOCs. Several options are available, see help info for more options</span>
<span class="c1"># Graph-community clustering: the higher the k value, the smaller the number of clusters --&gt; adjust if necessary</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;graph-community&#39;</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">cluster</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot heatmap of normalized, clustered protein data</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># If analyzing multiple samples, order barcodes by sample and then by cluster</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span><span class="n">splitby</span><span class="o">=</span><span class="s1">&#39;sample_name&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot UMAP for visualization of clusters</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span><span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP colored by the expression of each AOC</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span>
                           <span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
                           <span class="n">features</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">())</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Relabel clone names acccording to biology, combine clones by assigned identical names</span>
<span class="c1"># Just an example! Need to fill this in based on knowledge of the sample and markers</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">rename_labels</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s1">&#39;1&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type A&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;2&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type B&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;3&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type C&#39;</span><span class="p">,</span>
        <span class="s1">&#39;4&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type D&#39;</span><span class="p">,</span>
        <span class="s1">&#39;5&#39;</span><span class="p">:</span> <span class="s2">&quot;FP&quot;</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run heatmap again with new labels</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Drop cells (barcodes) that need to be removed</span>
<span class="c1"># Instead of overwriting the sample.protein variable, we define new variable called sample.protein2</span>
<span class="c1"># Note: if you do not wish to drop any cells, this code will producs an error (ignore)</span>

<span class="n">fp_barcodes2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(</span><span class="s2">&quot;FP&quot;</span><span class="p">)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">fp_barcodes2</span><span class="p">)</span>

<span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use only if no barcodes are removed by the above-listed command</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Additionally, option to remove AOCs from the data </span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;CD3&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Shape of cleaned up protein data (number of cells, number of AOCs)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Re-run heatmap with new filtered data</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="id1">
<h3>Statistics<a class="headerlink" href="#id1" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_protein</span><span class="p">,</span> <span class="n">tstat_protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">test_signature</span><span class="p">(</span><span class="s2">&quot;normalized_counts&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_protein</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_protein</span> <span class="o">=</span> <span class="n">pval_protein</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">**</span> <span class="o">-</span><span class="mi">50</span>
<span class="n">pvals_protein</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pval_protein</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">tstat_protein</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Colored tiles indicate strong association of a protein with a particular cluster</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pvals_protein</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="combined-visualizations">
<h2>Combined visualizations<a class="headerlink" href="#combined-visualizations" title="Permalink to this headline">#</a></h2>
<p><b>Topics covered</b></p>
<ol class="arabic simple">
<li><p>Defining a new sample object by combining the cleaned-up versions of the DNA, CNV and protein data</p></li>
<li><p>Built-in methods for combined visualizations</p></li>
<li><p>Additional visualizations with analyte-specific color schemes</p></li>
<li><p>Quantification of DNA clone and protein cluster overlaps</p></li>
<li><p>Export and save final data set</p></li>
</ol>
<section id="create-a-new-sample-object">
<h3>Create a new sample object<a class="headerlink" href="#create-a-new-sample-object" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># First, we ensure that all individual analytes are using the same cells.</span>
<span class="c1"># Since no cells are removed during CNV analysis, we take the intersect of cells retained in the DNA and protein data.</span>
<span class="c1"># In this and the following lines of code, use the final version of each assay, </span>
<span class="c1"># which may differ from the defaults listed here (e.g. sample.dna3, sample.protein4, etc.)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">barcodes</span><span class="p">()</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">barcodes</span><span class="p">()</span>

<span class="n">z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subset protein data to include only barcodes that were included in both analytes and store in sample.protein.3 variable</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein3</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="p">[</span><span class="n">z</span><span class="p">,</span><span class="n">sample</span><span class="o">.</span><span class="n">protein2</span><span class="o">.</span><span class="n">ids</span><span class="p">()]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subset dna data to include only barcodes that were included in both analytes and store in sample.dna.3 variable</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna3</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="p">[</span><span class="n">z</span><span class="p">,</span><span class="n">sample</span><span class="o">.</span><span class="n">dna2</span><span class="o">.</span><span class="n">ids</span><span class="p">()]</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Update each sub-class (e.g., DNA, CNV, Protein) in order to visualize &quot;clean&quot; data only with relevant features and cells.</span>
<span class="n">sample2</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[:]</span>
<span class="c1">#sub-assays</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna3</span> 
<span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein3</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check dimensionality for each subclass; the number of cells (first number) should be the same in each data set</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="built-in-visualizations-for-combining-analytes">
<h3>Built-in visualizations for combining analytes<a class="headerlink" href="#built-in-visualizations-for-combining-analytes" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># A heatmap presenting the data for up to three analytes</span>
<span class="c1"># Note, the default layer in the DNA subclass is NGT, not NGT_FILTERED</span>
<span class="c1"># If you wish to use NGT_FILTERED, you must first run the following line (commented out)</span>
<span class="c1"># Note that this will erase the NGT layer and replace it with NGT_FILTERED (which includes missing data)</span>
<span class="c1">#sample2.dna.layers[&#39;NGT&#39;] = sample2.dna.layers[&#39;NGT_FILTERED&#39;]</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">clusterby</span><span class="o">=</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="n">sortby</span><span class="o">=</span><span class="s1">&#39;protein&#39;</span><span class="p">,</span> <span class="n">flatten</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Often the number of amplicons in the CNV data will take over the heatmap, making the plot uninterpretable. Moreover, there might be certain non-differentiating variants and antibod in the panel. These can be dropped before making the final heatmap.</p>
<p>Note: when subsetting any of the analytes, the assay will “shrink” to contain only those features. To go back to the full data set, you must re-initialize your assay (see below).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter the CNV with amplicons only from the relevant genes</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">relevant_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;ASXL1&#39;</span><span class="p">,</span><span class="s1">&#39;EZH2&#39;</span><span class="p">,</span><span class="s1">&#39;TP53&#39;</span><span class="p">])</span>

<span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span><span class="p">[:,</span> <span class="n">relevant_ids</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot again w/ customization (e.g., color code)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">clusterby</span><span class="o">=</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="n">sortby</span><span class="o">=</span><span class="s1">&#39;cnv&#39;</span><span class="p">,</span><span class="n">drop</span><span class="o">=</span><span class="s1">&#39;protein&#39;</span><span class="p">,</span> <span class="n">flatten</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Update the width of the plot [See the section on CNV heatmaps]</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">3000</span>

<span class="c1"># Change the CNV colorscale [See the section on CNV heatmaps]</span>
<span class="n">fig</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">zmax</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">fig</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">zmin</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">fig</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="s1">&#39;magma&#39;</span>

<span class="c1"># Updating the ticktexts to show the gene names instead</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">xaxis3</span><span class="o">.</span><span class="n">ticktext</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">col_attrs</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Show as a static plot</span>
<span class="n">mutils</span><span class="o">.</span><span class="n">static_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To reset the CNV data (include all filtered amplicons instead of this new subset), simply re-initialize</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv2</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Alternatively, try the following combined heatmaps</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">clusterby</span><span class="o">=</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="n">sortby</span><span class="o">=</span><span class="s1">&#39;protein&#39;</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="s1">&#39;cnv&#39;</span><span class="p">,</span> <span class="n">flatten</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># sample2.heatmap(clusterby=&#39;protein&#39;, sortby=&#39;dna&#39;, drop=&#39;cnv&#39;, flatten=True)</span>
<span class="c1"># sample2.heatmap(clusterby=&#39;dna&#39;, sortby=&#39;protein&#39;, flatten=False)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample2</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the CNV data stratified by clone</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">(</span><span class="s1">&#39;cnv&#39;</span><span class="p">)</span>
<span class="n">fig</span>
<span class="c1">#fig.savefig(&#39;genotype_cnv.png&#39;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the protein data (violin plots) stratified by clone</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">(</span><span class="s1">&#39;protein&#39;</span><span class="p">)</span>
<span class="n">fig</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To review a subset of features, first subset the assay itself </span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="p">[:,</span> <span class="p">[</span><span class="s1">&#39;CD11b&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD38&#39;</span><span class="p">,</span> <span class="s1">&#39;CD90&#39;</span><span class="p">]]</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">(</span><span class="s1">&#39;protein&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To reset the Protein data (include all AOCs instead of this new subset), simply re-initialize</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein3</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="additional-visualizations">
<h3>Additional visualizations<a class="headerlink" href="#additional-visualizations" title="Permalink to this headline">#</a></h3>
<p>In addition to the built-in methods above, the analytes can be intersected by re-plotting some of the single-analyte visualizations, but coloring them by the labels of a different analyte.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the protein heatmap with the DNA clone labels</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span><span class="n">splitby</span><span class="o">=</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the protein UMAP and color with the DNA clone labels</span>
<span class="n">protein_umap</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;umap&quot;</span><span class="p">]</span>
<span class="n">fig</span><span class="o">=</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="n">protein_umap</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the protein UMAP and color with variant layers (e.g., VAFs) for specific variants</span>
<span class="n">protein_umap</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;umap&quot;</span><span class="p">]</span>
<span class="n">feats</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span> <span class="c1"># plot the first 3 variants</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="n">protein_umap</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s2">&quot;AF&quot;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">feats</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the CNV UMAP with the DNA NGT color code</span>
<span class="c1"># Note, this function is only applicable in the case of defining &quot;CNV clusters&quot; using the read-based approach (not the genotype-based approach)</span>
<span class="c1"># In this example, we have the read-based clustering saved in sample.cnv</span>
<span class="c1"># Verify that this object has the umap attribute</span>

<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># First, take only the barcodes retained in the dna data</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="p">[</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(),:]</span>

<span class="c1"># Then, plot the umap colored by NGT_FILTERED for certain variants</span>
<span class="n">cnv_umap</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;umap&quot;</span><span class="p">]</span>
<span class="n">feats</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span> <span class="c1"># plot the first 3 variants</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="n">cnv_umap</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s2">&quot;NGT_FILTERED&quot;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">feats</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the AOC ridge plots split by DNA clone</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ridgeplot</span><span class="p">(</span>
    <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD38&#39;</span><span class="p">,</span> <span class="s1">&#39;CD45&#39;</span><span class="p">,</span> <span class="s1">&#39;CD110&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">],</span>
    <span class="n">splitby</span><span class="o">=</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">(),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the AOC ridge plots split by protein cluster</span>
<span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ridgeplot</span><span class="p">(</span>
    <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD38&#39;</span><span class="p">,</span> <span class="s1">&#39;CD45&#39;</span><span class="p">,</span> <span class="s1">&#39;CD110&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">],</span>
    <span class="n">splitby</span><span class="o">=</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">(),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the biaxial AOC plot split by DNA clone</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">feature_scatter</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">],</span> <span class="n">colorby</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the biaxial AOC plot colored by genotype (or AF_MISSING, etc.) for specific variants</span>
<span class="n">variants</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DNMT3A:2:25458546:C:T&#39;</span><span class="p">,</span> <span class="s1">&#39;WT1:11:32417945:T:C&#39;</span><span class="p">]</span>
<span class="n">layer</span> <span class="o">=</span> <span class="s1">&#39;NGT_FILTERED&#39;</span>
<span class="n">AOCs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">]</span> <span class="c1"># only two!</span>

<span class="c1"># Set the layout and size of the resulting plots (rows, columns) - may need to adjust based on number of variants</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">variants</span><span class="p">),</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="c1"># Generate plots for each variant</span>
<span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
    <span class="n">color_variant</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layer</span><span class="p">][:,</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span> <span class="o">==</span> <span class="n">variant</span><span class="p">]</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">feature_scatter</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="n">AOCs</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="n">color_variant</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">variant</span><span class="p">))</span>
    <span class="n">mutils</span><span class="o">.</span><span class="n">static_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">counter</span><span class="p">])</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="n">counter</span> <span class="o">+</span> <span class="mi">1</span>

<span class="c1"># Show the plot</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">counter</span><span class="p">):</span>
    <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>    
</pre></div>
</div>
</div>
</div>
</section>
<section id="quantify-overlaps-between-dna-clones-and-protein-clusters">
<h3>Quantify overlaps between DNA clones and protein clusters<a class="headerlink" href="#quantify-overlaps-between-dna-clones-and-protein-clusters" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Count total cells in each cluster (cell type) and stratify by clone</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
<span class="n">all_cells</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Abundance in all clones&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">all_cells</span><span class="p">):</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">all_cells</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">y</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">x</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">all_cells</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;    &quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="p">,</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">)</span>
    
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]):</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Abundance in&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
    <span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="n">all_cells</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_cells</span><span class="p">))</span> <span class="k">if</span> <span class="n">names</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">subset</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">subset</span><span class="p">)[</span><span class="n">j</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">x</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;    &quot;</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="p">,</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For each cell type, measure abundance of each DNA clone</span>
<span class="n">all_cells</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">all_cells</span><span class="p">):</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Clonal abundance in&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
    <span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="n">names</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">names</span><span class="p">))</span> <span class="k">if</span> <span class="n">all_cells</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">subset</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">subset</span><span class="p">)[</span><span class="n">j</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">x</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;    &quot;</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="p">,</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># If analyzing multiple samples: count total cells in each cluster (cell type) and stratify by sample</span>
<span class="n">all_cells</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Abundance in merged data&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">all_cells</span><span class="p">):</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">all_cells</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">y</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">x</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">all_cells</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;    &quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="p">,</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">)</span>
    
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;sample_name&#39;</span><span class="p">]):</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;sample_name&#39;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Abundance in&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
    <span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="n">all_cells</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_cells</span><span class="p">))</span> <span class="k">if</span> <span class="n">names</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">subset</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">subset</span><span class="p">)[</span><span class="n">j</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">x</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">))</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;    &quot;</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="p">,</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="export-and-save-data">
<h3>Export and Save Data<a class="headerlink" href="#export-and-save-data" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save new h5 file that includes only the final, cleaned dataset</span>
<span class="n">ms</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">sample2</span><span class="p">,</span> <span class="s2">&quot;FilteredData.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Export data into csv format</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="c1">## set directory that does NOT exist</span>
<span class="n">folder_to_save</span> <span class="o">=</span> <span class="s1">&#39;./h5_extract/&#39;</span>

<span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">folder_to_save</span><span class="p">)</span>

<span class="k">for</span> <span class="n">assay</span> <span class="ow">in</span> <span class="p">[</span><span class="n">sample2</span><span class="o">.</span><span class="n">dna</span><span class="p">,</span> <span class="n">sample2</span><span class="o">.</span><span class="n">cnv</span><span class="p">,</span> <span class="n">sample2</span><span class="o">.</span><span class="n">protein</span><span class="p">]:</span>
    <span class="k">if</span> <span class="n">assay</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">folder_to_save</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">assay</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">folder_to_save</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">assay</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">/layers&#39;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">folder_to_save</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">assay</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">/rows&#39;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">layer</span> <span class="ow">in</span> <span class="n">assay</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">assay</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="n">layer</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row+col&#39;</span><span class="p">)</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assay</span><span class="o">.</span><span class="n">get_labels</span><span class="p">()</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">cols</span><span class="p">]</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">folder_to_save</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">assay</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">/layers/</span><span class="si">{</span><span class="n">layer</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">assay</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">assay</span><span class="o">.</span><span class="n">get_attribute</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">constraint</span><span class="o">=</span><span class="s1">&#39;row&#39;</span><span class="p">)</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assay</span><span class="o">.</span><span class="n">get_labels</span><span class="p">()</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">cols</span><span class="p">]</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">folder_to_save</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">assay</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">/rows/</span><span class="si">{</span><span class="n">row</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>


              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="additional-plots-and-customizations.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Additional plots and customizations</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../pages/missionbio.mosaic.io.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">io</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Saurabh Parikh<br/>
  
      &copy; Copyright 2021, Mission Bio.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>