
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Dna.group_by_genotype &#8212; Mosaic v2.4 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Mosaic v2.4 documentation</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../1_introduction.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../2_install.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../3_getting_started.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../4_data_structure.html">
   Data structure
  </a>
 </li>
</ul>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/index.html">
   Code Snippets
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/basics.html">
   Basic usage of mosaic
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/cnv.html">
   Manual CNV analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/filtering.html">
   Filtering barcodes and ids
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/customizations.html">
   Plot customizations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/multi-sample-analysis.html">
   Multisample analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/additional-plots-and-customizations.html">
   Additional plots and customizations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../examples/analysis-walkthrough.html">
   Mosaic analysis walk-through
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modules
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.io.html">
   io
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.utils.html">
   utils
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../missionbio.mosaic.constants.html">
   constants
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.COLORS.html">
     missionbio.mosaic.constants.COLORS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.PCA_LABEL.html">
     missionbio.mosaic.constants.PCA_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.UMAP_LABEL.html">
     missionbio.mosaic.constants.UMAP_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.SCALED_LABEL.html">
     missionbio.mosaic.constants.SCALED_LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.LABEL.html">
     missionbio.mosaic.constants.LABEL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.READS.html">
     missionbio.mosaic.constants.READS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.PALETTE.html">
     missionbio.mosaic.constants.PALETTE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../missionbio.mosaic.constants.NORMALIZED_READS.html">
     missionbio.mosaic.constants.NORMALIZED_READS
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Classes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.assay._Assay.html">
   _Assay
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.dna.Dna.html">
   Dna
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.cnv.Cnv.html">
   Cnv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.protein.Protein.html">
   Protein
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.sample.Sample.html">
   Sample
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../missionbio.mosaic.samplegroup.SampleGroup.html">
   SampleGroup
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../../_sources/pages/methods/missionbio.mosaic.dna.Dna.group_by_genotype.rst"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.rst</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Dna.group_by_genotype</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="dna-group-by-genotype">
<h1>Dna.group_by_genotype<a class="headerlink" href="#dna-group-by-genotype" title="Permalink to this headline">#</a></h1>
<dl class="py method">
<dt class="sig sig-object py" id="missionbio.mosaic.dna.Dna.group_by_genotype">
<span class="sig-prename descclassname"><span class="pre">Dna.</span></span><span class="sig-name descname"><span class="pre">group_by_genotype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'NGT'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_missing</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_ado_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_clone_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_zygosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_plot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#missionbio.mosaic.dna.Dna.group_by_genotype" title="Permalink to this definition">#</a></dt>
<dd><p>Clusters cells into clones based on the provided variants
and returns a dataframe of per-clone and per-variant statistics.
The identified clone labels are stored in the ‘LABEL’ layer.
This algorithm also takes into consideration allele dropout out (ADO)
to identify potential false positive clones.</p>
<p><strong>Algorithm details</strong></p>
<p>In an ADO event HET goes to WT and HOM for a given
variant in a subset of the cells. Here, the HET clone is
called the parent clone, the HOM and WT clones are
the ADO clones, together called the sister clones.</p>
<p>The parent and sister clones will be <cite>np.nan</cite> if the
score is zero. Otherwise, it is the name of the clone
from which the subclone was obtained due to an ADO event.</p>
<p>The score for each subclone measures the possibility that
it’s a flase positive subclone obtained due to an ADO
event. The score is 0 if it unlikely to be a clone due
to ADO and 1 if it is highly likely to be an ADO clone.</p>
<dl class="simple">
<dt>The score takes into account the following metrics.</dt><dd><ol class="arabic simple">
<li><p>NGT values of the clones</p></li>
<li><p>Relative proportions of the clones</p></li>
<li><p>Absolute proportions of the clones (uses min_clone_size as a parameter)</p></li>
<li><p>Mean GQ of the clones</p></li>
<li><p>Mean DP of the clones</p></li>
</ol>
</dd>
<dt>The score is calculated using four sub scores.</dt><dd><p><code class="docutils literal notranslate"><span class="pre">score</span> <span class="pre">=</span> <span class="pre">(ss</span> <span class="pre">+</span> <span class="pre">ds</span> <span class="pre">+</span> <span class="pre">gs)</span> <span class="pre">*</span> <span class="pre">ps</span></code></p>
<ol class="arabic simple">
<li><dl class="simple">
<dt><cite>ss - sister score (0 - 0.8)</cite></dt><dd><p>It measures the proportion of the clone
with resepect to its sister clone. This
score is closer to 0.8 when the sister clones
have similar proportions and exactly 0.8 when
their proportions are within the min_clone_size.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>ds - DP score (0 - 0.1)</cite></dt><dd><p>It measures the mean DP of the clone
with resepect to its parent clone. It is
closer to 0.1 if the DP of the clone is
lower than the parents’ DP.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>gs - GQ score (0 - 0.1)</cite></dt><dd><p>It measures the mean GQ of the clone
with resepect to its parent clone. It is
closer to 0.1 if the GQ of the clone is
lowert than the parents’ GQ.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>ps - parent score (0 - 1)</cite></dt><dd><p>It measures the proportion of the clone with
respect to the parent clone. This score is closer
to 1 the larger the parent is compared to the
clone, and closer to 0 the smaller the parent
compared to the clone.</p>
</dd>
</dl>
</li>
</ol>
</dd>
</dl>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>features</strong><span class="classifier">Sequence</span></dt><dd><p>The features which are to be considered while
allocating the groups formed by the genotype.</p>
</dd>
<dt><strong>layer</strong><span class="classifier">str</span></dt><dd><p>Name of the layer used to count the cell types.
Expected values are NGT or NGT_FILTERED as obtained
from the <a class="reference internal" href="missionbio.mosaic.dna.Dna.filter_variants.html#missionbio.mosaic.dna.Dna.filter_variants" title="missionbio.mosaic.dna.Dna.filter_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Dna.filter_variants()</span></code></a> method.</p>
</dd>
<dt><strong>group_missing</strong><span class="classifier">bool</span></dt><dd><p>Whether the clusters caused due to missing values
are merged together under one cluster named ‘Missing’.</p>
</dd>
<dt><strong>max_ado_score</strong><span class="classifier">float [0, 1]</span></dt><dd><p>The maximum ADO score of a clone before it is grouped
into the “ADO” clone category.</p>
</dd>
<dt><strong>min_clone_size</strong><span class="classifier">float [0, 100]</span></dt><dd><p>The minimumum proportion of total cells to be present
in the clone to count it as a separate clone.</p>
</dd>
<dt><strong>ignore_zygosity</strong><span class="classifier">bool</span></dt><dd><p>Whether HET and HOM are considered the same or not</p>
</dd>
<dt><strong>show_plot</strong><span class="classifier">bool | str</span></dt><dd><p>If True, a plot showing the ADO identification process
is shown. The same plot is shown if the value is “all”.
True as option is kept for backwards compatibility. If
the value is “collapsed” then only the non-small and
non-ADO clones are shown separately.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt>pd.DataFrame / None</dt><dd><p><cite>None</cite> is returned if <cite>ignore_zygosity</cite> is True or
<cite>group_missing</cite> is False otherwise a pandas dataframe
is returned.</p>
<p><strong>Columns:</strong></p>
<ul class="simple">
<li><p>The initial columns in the dataframe contain per-variant information.
Each cell contains the genotype and average variant allele frequency
for the respective clone.</p></li>
<li><p>The next set of columns show the number of cells (and %) in each
clone, with one column for all samples combined together
(<cite>Total Cell Number</cite>) and separate columns for each individual
sample (e.g. <cite>Sample-A Cell Number</cite>).</p></li>
<li><p>The last three columns contain data related to identifying false
positive ADO clones. The columns are: parent clone, sister ADO
clone(s), and a score. The indices are the subclone names.</p></li>
</ul>
<p><strong>Additional rows:</strong></p>
<ul class="simple">
<li><p>The second-to-last row shows the percentage of cells where the
variant is missing.</p></li>
<li><p>The last row displays statistics for small subclones (determeined
by the <cite>min_clone_size</cite> parameter): the number and percentage of
total cells.</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>


              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Saurabh Parikh<br/>
  
      &copy; Copyright 2021, Mission Bio.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>