

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Single Sample DNA + Protein notebook &#8212; Mosaic v3.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/curated_notebooks/dna-protein';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Multisample DNA + Protein notebook" href="dna-protein-multisample.html" />
    <link rel="prev" title="Multisample DNA-only notebook" href="dna-multisample.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    <p class="title logo__title">Mosaic v3.0.1 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pages/install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pages/getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pages/data_structure.html">Data structure</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../pages/vignettes.html">Vignettes</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../overview.html">Overview of mosaic</a></li>
<li class="toctree-l2"><a class="reference internal" href="../filtering.html">Filtering barcodes and ids</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cnv.html">Manual CNV analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plot-customizations.html">Plot customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multi-sample-analysis.html">Multisample analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../additional-plots-and-customizations.html">Additional plots and customizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../code-snippets.html">Code Snippets</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../../pages/curated_jupyter_notebooks.html">Curated Jupyter Notebooks</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dna.html">Single Sample DNA-only notebook</a></li>







<li class="toctree-l2"><a class="reference internal" href="dna-multisample.html">Multisample DNA-only notebook</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Single Sample DNA + Protein notebook</a></li>








<li class="toctree-l2"><a class="reference internal" href="dna-protein-multisample.html">Multisample DNA + Protein notebook</a></li>








</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../pages/help.html">Additional resources</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../pages/changelog.html">Changelog</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Classes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.assay._Assay.html">_Assay</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.dna.Dna.html">Dna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.cnv.Cnv.html">Cnv</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.protein.Protein.html">Protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.sample.Sample.html">Sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.samplegroup.SampleGroup.html">SampleGroup</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interactive Workflows</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.workflows.copy_number.CopyNumber.html">CopyNumber</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.workflows.variant_subclone_table.VariantSubcloneTable.html">VariantSubcloneTable</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Algorithms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.algorithms.nsp.NSP.html">NSP</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.algorithms.group_by_genotype.GroupByGenotype.html">GroupByGenotype</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.algorithms.compass.COMPASS.html">COMPASS</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Functional Modules</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.io.html">io</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.utils.html">utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.constants.html">constants</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Custom Plots</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.plots.heatmap.Heatmap.html">Heatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.plots.lineplot.LinePlot.html">LinePlot</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.plots.bargraph.BarGraph.html">BarGraph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.plots.multi_heatmap.MultiHeatmap.html">MultiHeatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.plots.fishplot.TreeGraph.html">TreeGraph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autosummary_pages/missionbio.mosaic.plots.fishplot.Fishplot.html">Fishplot</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/notebooks/curated_notebooks/dna-protein.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Single Sample DNA + Protein notebook</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Single Sample DNA + Protein notebook</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-structure">Data Structure</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#dna-analysis">DNA Analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-filtering">Basic filtering</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-addition">Annotation Addition</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variant-selection-and-subclone-identification">Variant selection and Subclone identification</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adjusting-subclone-colors-or-heatmap-colors">Adjusting subclone colors or heatmap colors</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv-analysis">CNV Analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv-workflow">CNV workflow</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#protein-analysis">Protein Analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#normalization-and-data-inspection">Normalization And Data Inspection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clustering-and-visualization">Clustering And Visualization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistics">Statistics</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#combined-visualizations">Combined Visualizations</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantify-overlaps-between-dna-clones-and-protein-clusters">Quantify overlaps between DNA clones and Protein clusters</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#export-and-save-data">Export and Save Data</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#appendix">Appendix</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dna-signature">DNA signature</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#compass-imputation">Compass imputation</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
    <!--Add a white cover to hide the .bd-container grey on the right side of the page.-->
    <div class="background-cover" id="background-cover"></div>
    <section id="single-sample-dna-protein-notebook">
<h1>Single Sample DNA + Protein notebook<a class="headerlink" href="#single-sample-dna-protein-notebook" title="Permalink to this heading">#</a></h1>
<p><b>This notebook is set up to work with single-sample h5 files, generated with v2 or v3 chemistry. This notebook will only work with Mosaic versions 3.0.1 and higher.</b></p>
<p><b>Objective</b>: To showcase the minimum number of steps required for tertiary analysis of DNA (single-cell genotyping and CNV) and Protein analytes and explore different ways of visualizing the data.<br></p>
<p><b>Major questions answered:</b></p>
<ol class="arabic simple">
<li><p>Can we identify DNA clones based on genotypes (SNVs/Indels)?<br></p></li>
<li><p>Do we detect CNV events (e.g., copy number amplification, copy number loss)?<br></p></li>
<li><p>Can we delineate different cell types from cell surface protein expression data?<br></p></li>
<li><p>Do DNA clones (genotypes) correlate with specific protein-defined cell types and/or CNV profiles?<br></p></li>
</ol>
<p><b>Sections:</b></p>
<ol class="arabic simple">
<li><p>Setup</p></li>
<li><p>Data Structure</p></li>
<li><p>DNA Analysis</p></li>
<li><p>CNV Analysis</p></li>
<li><p>Protein Analysis</p></li>
<li><p>Combined Visualizations</p></li>
<li><p>Export and Save Data</p></li>
<li><p>Appendix</p></li>
</ol>
<p><b>Not shown:</b>
All available methods and options - documented <a class="reference external" href="https://missionbio.github.io/mosaic/index.html">here</a><br></p>
</section>
<section id="setup">
<h1>Setup<a class="headerlink" href="#setup" title="Permalink to this heading">#</a></h1>
<p><b>Topics covered</b><br></p>
<ol class="arabic simple">
<li><p>Loading required packages and data.</p></li>
<li><p>Structure and contents of data objects.</p></li>
</ol>
<p><b>Load data</b><br>
<b>Note:</b> importing dependencies can sometimes take a couple of minutes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import mosaic libraries</span>
<span class="kn">import</span> <span class="nn">missionbio.mosaic</span> <span class="k">as</span> <span class="nn">ms</span>

<span class="c1"># Import these to display entire dataframes</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>

<span class="c1"># Import graph_objects from the plotly package to display figures when saving the notebook as an HTML</span>
<span class="kn">import</span> <span class="nn">plotly</span> <span class="k">as</span> <span class="nn">px</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>

<span class="c1"># Import additional packages for specific visuals</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">plotly.offline</span> <span class="k">as</span> <span class="nn">pyo</span>
<span class="n">pyo</span><span class="o">.</span><span class="n">init_notebook_mode</span><span class="p">()</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1"># Import COMPASS for imputation</span>
<span class="kn">from</span> <span class="nn">missionbio.mosaic.algorithms.compass</span> <span class="kn">import</span> <span class="n">COMPASS</span>

<span class="c1"># Note: when exporting the notebook as an HTML, plots that use the &quot;go.Figure(fig)&quot; command are saved</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This code is optional, but will make the notebook cells/figures display across the entire width of your browser</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="s2">&quot;&lt;style&gt;.container { width:100% !important; }&lt;/style&gt;&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check version; this notebook is designed for Mosaic v3.0.1 or higher</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Any function&#39;s parameters and default values can be looked up via the &#39;help&#39; function</span>
<span class="c1"># Here, the function is &#39;ms.load&#39;</span>
<span class="n">help</span><span class="p">(</span><span class="n">ms</span><span class="o">.</span><span class="n">load</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the h5 file to be used in this analysis: h5path = &#39;/path/to/h5/file/test.h5&#39;</span>
<span class="c1"># If working with Windows, you may need to add an &#39;r&#39; before the path: h5path = r&#39;/path/to/h5/file/test.h5&#39;</span>
<span class="n">h5path</span> <span class="o">=</span> <span class="s1">&#39;/Users/botero/Desktop/Sample data/4-cell-lines-AML-multiomics/4-cell-lines-AML-multiomics.dna+protein.h5&#39;</span>

<span class="c1"># Load the data</span>
<span class="n">sample</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">h5path</span><span class="p">,</span> <span class="n">raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">filter_variants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">single</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">whitelist</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">filter_cells</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> 

<span class="c1"># Always set raw=False; if raw=True, ALL barcodes will be loaded (rather than cell-associated barcodes)</span>
<span class="c1"># Always set filter_variants=True unless you can&#39;t detect an expected (target) variant. Additional filtering options are included in the DNA section below</span>
<span class="c1"># The single=True option loads multi-sample h5 files as a single sample object (compatible with this notebook)</span>
<span class="c1"># The whitelist option loads any variant that is in the vcf.gz file (e.g. &quot;chr1:179520511:C/G&quot;); similar to whitelist feature in Tapestri Insights v2</span>
<span class="c1"># Always set filter_cells=False; if filter_cells=True, only genomically complete cells will be loaded </span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-structure">
<h1>Data Structure<a class="headerlink" href="#data-structure" title="Permalink to this heading">#</a></h1>
<p>DNA, CNV, and Protein are sub-classes of the Assay class. The information is stored in four categories, and the user can modify each of those:</p>
<p><b>1. metadata (add_metadata / del_metadata):</b></p>
<ul class="simple">
<li><p>dictionary containing metrics of the assay</p></li>
</ul>
<p><b>2. row_attrs (add_row_attr / del_row_attr):</b></p>
<ul class="simple">
<li><p>dictionary which contains ‘barcode’ as one of the keys
(where the value is a list of all barcodes)</p></li>
<li><p>for all other keys, the values must be of the same
length, i.e. match the number of barcodes</p></li>
<li><p>this is the attribute where ‘label’, ‘pca’,
and ‘umap’ values are added</p></li>
</ul>
<p><b>3. col_attrs (add_col_attr / del_col_attr):</b></p>
<ul class="simple">
<li><p>dictionary which contains ‘id’ as one of the keys
(where the value is a list of all ids)</p></li>
<li><p>for DNA assays, ‘ids’ are variants; for Protein assays,
‘ids’ are antibodies</p></li>
<li><p>for all other keys, the values must be of the same
length, i.e. match the number ids</p></li>
</ul>
<p><b>4. layers (add_layer / del_layer):</b></p>
<ul class="simple">
<li><p>dictionary which contains critically important assay metrics</p></li>
<li><p>all the values have the shape (num barcodes) x (num ids)</p></li>
<li><p>for DNA assays, this includes AF, GQ, DP, etc.
(per cell, per variant)</p></li>
<li><p>for Protein assays, this includes read counts
(per cell, per antibody)</p></li>
<li><p>this is the attribute where ‘normalized_counts’ will be added</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Summary of DNA assay </span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">sample.dna</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">row_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">col_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">layers</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">metadata</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="s1">&#39;: &#39;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For DNA, ids are variants</span>
<span class="c1"># sample.dna.ids() is a shortcut for sample.dna.col_attrs[&#39;id&#39;]</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Summary of Protein assay </span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">sample.protein</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">row_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">col_attrs</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">layers</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&#39;</span><span class="s2">metadata</span><span class="se">\&#39;</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="s2">&quot;: &quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For Protein, ids are AOCs</span>
<span class="c1"># sample.protein.ids() is a shortcut for sample.protein.col_attrs[&#39;id&#39;]</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="dna-analysis">
<h1>DNA Analysis<a class="headerlink" href="#dna-analysis" title="Permalink to this heading">#</a></h1>
<p><b>Topics covered</b><br></p>
<ol class="arabic simple">
<li><p>Standard filtering of DNA variants.</p></li>
<li><p>Subsetting dataset for variants of interest, including whitelisted variants.</p></li>
<li><p>Addition of annotations to the variants.</p></li>
<li><p>Manual variant selection and clone identification</p></li>
<li><p><b> OPTIONAL: </b> Use Compass to impute missing data and a clonal phylogeny</p></li>
<li><p>Visualizations and customization options</p></li>
</ol>
<section id="basic-filtering">
<h2>Basic filtering<a class="headerlink" href="#basic-filtering" title="Permalink to this heading">#</a></h2>
<p>There are many options for filtering DNA variants.
Use the <code class="docutils literal notranslate"><span class="pre">help()</span></code> function to understand the approach listed below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For additional information visit: https://support.missionbio.com/hc/en-us/articles/360047303654-How-do-the-Advanced-Filters-work-</span>
<span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter the variants, similar to the &quot;Advanced Filters&quot; in Tapestri Insights v2.2</span>

<span class="c1"># One major difference: The filter &quot;REMOVE GENOTYPE IN CELLS WITH ALTERNATE ALLELE FREQ&quot; </span>
<span class="c1"># is replaced with the following 3 zygosity-specific filters: vaf_ref, vaf_hom, vaf_het</span>

<span class="c1"># In general, these additional filters remove additional false-positive from the data.</span>

<span class="c1"># Adjust filters if needed by overwriting dna_vars</span>
<span class="n">dna_vars</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">(</span>
    <span class="n">min_dp</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">min_gq</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
    <span class="n">vaf_ref</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">vaf_hom</span><span class="o">=</span><span class="mi">95</span><span class="p">,</span>
    <span class="n">vaf_het</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
    <span class="n">min_prct_cells</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">min_mut_prct_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">iterations</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Check the number of filtered variants. When using the default filters, the number of </span>
<span class="c1"># variants is likely smaller compared to the originally loaded variants due to the more </span>
<span class="c1"># stringent filtering criteria (e.g., vaf_ref=5, vaf_hom=95, vaf_het=30).</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of variants:&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dna_vars</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dna_vars</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>First, specify variants of interest using one of the three options below:</p>
<ol class="arabic simple">
<li><p>Use the filtered variants from the above section (dna_vars)</p></li>
<li><p>Use specific variants of interest (whitelist)</p></li>
<li><p>Combine 1 &amp; 2: filtered variants plus whitelist</p></li>
</ol>
<p>Then, this list (final_vars) is used to reduce the larger data set to only your variants of interest.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Specify the whitelist; variants may be copy/pasted from Tapestri Insights v2.2,</span>
<span class="c1"># but ensure correct nomenclature, ie whitelist = [&quot;chr13:28589657:T/G&quot;,&quot;chrX:39921424:G/A&quot;]</span>
<span class="c1"># If there are no whitelist variants, you can leave target variants blank</span>

<span class="n">white_list</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Combine whitelisted and filtered variants</span>
<span class="n">final_vars</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">dna_vars</span><span class="p">)</span> <span class="o">+</span> <span class="n">white_list</span><span class="p">))</span>

<span class="c1"># Want to use your whitelist only?</span>
<span class="c1"># final_vars = white_list</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check the length of your final list of variants</span>
<span class="nb">len</span><span class="p">(</span><span class="n">final_vars</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dimensionality of the original sample.dna dataframe</span>
<span class="c1"># First number = number of cells (rows); second number = number of variants (columns)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Before subsetting, verify that all the chosen variants are in the current sample.dna ids (should return True)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">final_vars</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">())))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subsetting sample.dna (columns) based on reduced variant list. Keeping all cells that passed filtering</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">[:,</span> <span class="n">final_vars</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check the shape of the final filtered DNA object, i.e. (number of barcodes/cells, number of ids/variants)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotation-addition">
<h2>Annotation Addition<a class="headerlink" href="#annotation-addition" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_annotations</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Fetch annotations using varsome</span>
<span class="c1"># Note: run this on a filtered DNA sample - too many variants (e.g., 100+) are not handled correctly by the method</span>
<span class="n">ann</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_annotations</span><span class="p">()</span>  
</pre></div>
</div>
</div>
</div>
</section>
<section id="variant-selection-and-subclone-identification">
<h2>Variant selection and Subclone identification<a class="headerlink" href="#variant-selection-and-subclone-identification" title="Permalink to this heading">#</a></h2>
<p>In this section of the notebook, all variants remaining in the data will populate in a variant table. This table is interaction, variants can be selected, and rows can be sorted by ascending/descending values. The variant name can be clicked on and will navigate to the variants varsome url in your default browser.</p>
<ol class="arabic simple">
<li><p>Variants selected in this table will populate in a subclone table below.</p></li>
<li><p>The variants in the subclone table can be highlighted and assessed for Read Depth, Genotype Quality and Variant Allele Frequency.</p></li>
<li><p>Subclones can be renamed by clicking on the pencil icon.</p></li>
<li><p><b>ADO score:</b> The ADO score can be adjusted, but by default is set to 1. Any clones with an ADO score higher this will be moved into the ADO subclones column. We recommend moving any clones with a score of .8 or higher into this column.</p></li>
<li><p><b> Min clone size: </b> The Min Clone Size can also be adjusted, but by default is set to 1. Any clones that represent less than 1% of the sample will be moved into the Small Subclones column.</p></li>
<li><p>Cells with missing genotype information across any of the selected variants will be moved into the missing GT column.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Run the variant table workflow to select variants and begin clone identification</span>
<span class="n">wfv</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">VariantSubcloneTable</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
<span class="n">wfv</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subsample the Variants to only the variants selected from the workflow</span>
<span class="n">variants</span> <span class="o">=</span> <span class="n">wfv</span><span class="o">.</span><span class="n">selected_variants</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="p">[:,</span><span class="n">variants</span><span class="p">]</span>

<span class="c1"># Save the full set of cells to a new variable that you can call on later</span>
<span class="c1"># Do this before renaming the variants</span>
<span class="n">full_sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[:]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rename your variants</span>
<span class="c1"># Any of these column values can be added to the id names</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add annotation to the id names</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">set_ids_from_cols</span><span class="p">([</span><span class="s1">&#39;annotated Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">])</span>

<span class="c1"># Another example:</span>
<span class="c1"># sample.dna.set_ids_from_cols([&#39;annotated Gene&#39;, &#39;CHROM&#39;, &#39;POS&#39;, &#39;REF&#39;, &#39;ALT&#39;])</span>

<span class="c1"># Annotations are now added to the variants</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>

<span class="c1"># Use sample.dna.reset_ids() to get the original ids</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot heatmap using NGT_FILTERED.</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clone removal</span>
<span class="c1"># Remove barcodes from the missing GT, small subclones, ADO subclones or FP labels</span>
<span class="n">clones</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;missing&#39;</span><span class="p">,</span> <span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="s1">&#39;ADO&#39;</span><span class="p">,</span> <span class="s1">&#39;FP&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">clones</span><span class="p">:</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">barcodes</span><span class="p">({</span><span class="n">c</span><span class="p">})</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cells</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">sample</span><span class="o">.</span><span class="n">dna</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">cells</span><span class="p">)</span>
<span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Redraw heatmap</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can use the following line of code to change the color for heatmaps in the DNA, CNV or protein sections</span>
<span class="c1"># And rotate the tick labels</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;NGT_FILTERED&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span><span class="n">title</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">update_xaxes</span><span class="p">(</span><span class="n">tickangle</span> <span class="o">=</span> <span class="o">-</span><span class="mi">45</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">coloraxis</span><span class="o">.</span><span class="n">colorscale</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span>
<span class="n">fig</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Evaluate new total number of cells after the above filtering</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="adjusting-subclone-colors-or-heatmap-colors">
<h2>Adjusting subclone colors or heatmap colors<a class="headerlink" href="#adjusting-subclone-colors-or-heatmap-colors" title="Permalink to this heading">#</a></h2>
<p>If you want to change the color palette used for the subclones, you can use <code class="docutils literal notranslate"><span class="pre">set_palette</span></code>. For this you will need to provide a list of ALL subclone names, pointing to the hexcode/color you want to change that subclone to. Note: this function also works with the protein and CNV assays. See below for an example of this.</p>
<p>If you want to change the color palette used for any of the assays/layers, you can use <code class="docutils literal notranslate"><span class="pre">ms.Config.Colorscale</span></code>. Then you can list the assay and layer, and assign the plotly colorscale you would like to use for those graphics. You can visualized all available colorscales by using <code class="docutils literal notranslate"><span class="pre">plotly.colors.sequential.swatches_continuous().show()</span></code>. If you want to reset all color palettes back to their defaults, you can use <code class="docutils literal notranslate"><span class="pre">ms.Config.Colorscle.reset()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example of changing the colors assigned to each clone</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">set_palette</span><span class="p">({</span><span class="s1">&#39;Clone 1&#39;</span><span class="p">:</span> <span class="s1">&#39;#800080&#39;</span><span class="p">,</span> <span class="s1">&#39;Clone 2&#39;</span><span class="p">:</span> <span class="s1">&#39;#FF69B4&#39;</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can change colorScale to change the heatmap colors</span>
<span class="c1"># This will change the color of the dna.ngt layer to be viridis</span>
<span class="n">ms</span><span class="o">.</span><span class="n">Config</span><span class="o">.</span><span class="n">Colorscale</span><span class="o">.</span><span class="n">Dna</span><span class="o">.</span><span class="n">NGT</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;NGT&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># If you want to reset this, you can run the following to reset just that one modification</span>
<span class="c1">#ms.Config.Colorscale.Dna.reset(&quot;NGT&quot;)</span>
<span class="c1"># Or you can run this to reset any/all modifications:</span>
<span class="c1">#ms.Config.Colorscale.reset()</span>
<span class="c1"># You can change the colorScale for Dna, Cns or Protein</span>
</pre></div>
</div>
</div>
</div>
<p>If you ever want to return back to the original population of cells, you can reset the data using: <code class="docutils literal notranslate"><span class="pre">sample.reset(&quot;dna&quot;)</span></code> This command ‘sample.reset’ works on all assays, including CNV and protein.</p>
</section>
</section>
<section id="cnv-analysis">
<h1>CNV Analysis<a class="headerlink" href="#cnv-analysis" title="Permalink to this heading">#</a></h1>
<p><b>Topics covered</b></p>
<ol class="arabic simple">
<li><p>Amplicon filtering and ploidy estimation.</p></li>
<li><p>Visualization of ploidy across subclones present</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get gene names for amplicons</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">get_gene_names</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="cnv-workflow">
<h2>CNV workflow<a class="headerlink" href="#cnv-workflow" title="Permalink to this heading">#</a></h2>
<p>This workflow will normalize all reads and filter amplicons/cells based on the settings set at the beginning of the workflow: <br></p>
<ol class="arabic simple">
<li><p><b> Amplicon completeness: </b> refers to the minimum percentage of barcodes for an amplicon that must have reads greater than or equal to the minimum read depth set. By default this is set to 50. <br></p></li>
<li><p><b> Amplicon read depth: </b> refers to the minimum read depth for each amplicon-barcode combination to not be considered missing. By default this is set to 10. <br></p></li>
<li><p><b>Mean cell read depth: </b> refers the minimum mean read depth for a cell to be included in the analysis, otherwise the cell will be removed. By default this is set to 40.<br></p></li>
<li><p><b>Diploid clone in DNA: </b> refers to which subclone you are setting as the true diploid population. All ploidy estimates will be calculated in relation to this diploid population. We recommend setting this to your ‘WT’ population or most parent clone present. <br></p></li>
</ol>
<p><b> Note: </b> These settings are pretty stringent. If you are expecting large copy number events, you may want to reduce Amplicon Completeness and Amplicon Read Depth, to recover these events.</p>
<p>Once the above filters are set, the visualizations can be changed. <br></p>
<ol class="arabic simple">
<li><p><b> Plot:</b> Can be changed from Heatmap positions, to Heatmap genes, Line-plot positions, Line-plot genes <br></p></li>
<li><p><b> Clone for line plot:</b> If one of the line-plot visualizations is selected, only one clone can be shown at a time. This determines which one is plotted. <br></p></li>
<li><p><b> X-axis features: </b> If you would prefer to only plot a subset of the data (chromosomes or genes), you can select which chromosomes/genes you would like plotted with this function. Chromosomes can be selected for ‘positions’ type plots, and Genes can be selected for ‘genes’ type plots.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># CNV workflow to filter, normalize and estimate ploidy</span>
<span class="n">wfc</span> <span class="o">=</span> <span class="n">ms</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">CopyNumber</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
<span class="n">wfc</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
<span class="c1"># Amplicon completeness refers to the min percentage of barcodes for an amplicon that must have reads &gt;= the read_depth</span>
<span class="c1"># Read depth is the min required depth for each amplicon_barcode to not be considered as missing</span>
<span class="c1"># Mean cell read depth refers to the min mean read depth for a cell to be included, otherwise it is removed</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Heatmap with the features ordered by the default amplicon order</span>
<span class="c1"># If you want to plot just a subset of chromosomes, you can put them in list format for features</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s1">&#39;positions&#39;</span><span class="p">)</span> <span class="c1">#features=[&#39;7&#39;, &#39;17&#39;, &#39;20&#39;]</span>

<span class="c1"># Optionally, restrict the range of ploidy values based on observed/expected CNV events (commented out)</span>
<span class="c1">#fig.layout.coloraxis.cmax = 4</span>
<span class="c1">#fig.layout.coloraxis.cmin = 0</span>

<span class="c1"># Optionally, change the size of the figure:</span>
<span class="c1">#fig.layout.width = 1600</span>
<span class="c1">#fig.layout.height = 1500</span>

<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Heatmap with the features grouped by the genes</span>
<span class="c1"># If you want to plot just a subset of genes, you can put them in list format for features</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="n">convolve</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#features=[&quot;ASXL1&quot;, &quot;EZH2&quot;,&#39;TP53&#39;]</span>

<span class="c1"># Optionally, update the separating lines to be black</span>
<span class="c1">#for shape in fig.layout.shapes:</span>
<span class="c1">#    shape.line.color = &#39;#000000&#39;</span>

<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show heatmap with convolve and subclustering turned off</span>
<span class="n">bars</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">clustered_barcodes</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">subcluster</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># This is useful to create &quot;convolved&quot; heatmaps which are easier to interpret</span>
<span class="c1"># With the subclustering off and convolve=20, the noise will be reduced and real signals will be easier to determine</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="n">bars_order</span><span class="o">=</span><span class="n">bars</span><span class="p">,</span> <span class="n">convolve</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">900</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Signature will create a dataframe based on the layer you want to look at and which statistical value you want to view</span>
<span class="c1"># You can see: mean, median, mode, or std with this function</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># signaturemap will plot a heatmap of the signature dataframe created above</span>
<span class="c1"># The labels list will control the order of the subclones along the y-axis</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">signaturemap</span><span class="p">(</span><span class="s1">&#39;ploidy&#39;</span><span class="p">)</span> <span class="c1">#labels=[]</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="protein-analysis">
<h1>Protein Analysis<a class="headerlink" href="#protein-analysis" title="Permalink to this heading">#</a></h1>
<p><b>Topics covered</b></p>
<ol class="arabic simple">
<li><p>Normalization and assessment of AOC abundance</p></li>
<li><p>Clustering and visualization</p></li>
<li><p>Statistical significance analysis</p></li>
</ol>
<section id="normalization-and-data-inspection">
<h2>Normalization And Data Inspection<a class="headerlink" href="#normalization-and-data-inspection" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Normalize reads</span>
<span class="c1"># Three different methods available: CLR = Centered Log Ratio, NSP = Noise corrected and Scaled, asinh = Inverse Hyperbolictrans formation</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">normalize_reads</span><span class="p">(</span><span class="s1">&#39;NSP&#39;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Ridge plot: shows distribution of AOC counts across all cells (bimodal = positive and negative cell populations)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ridgeplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
                         <span class="n">features</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plotting combinations of AOCs on a biaxial scatterplot, to mimic FACS data</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">feature_scatter</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD45&#39;</span><span class="p">])</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#List all AOCs</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Optionally, remove AOCs from the data set prior to clustering </span>
<span class="c1"># For example, those that don&#39;t display any signal in any cells</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;Mouse IgG1k&#39;</span><span class="p">,</span> <span class="s1">&#39;CD25&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="clustering-and-visualization">
<h2>Clustering And Visualization<a class="headerlink" href="#clustering-and-visualization" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_pca</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Number of components should equal the number of AOCs remaining in the dataset</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span><span class="n">show_plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rerun PCA with optimal number of components based on elbow plot analysis</span>
<span class="c1"># Typically, components = 10 is appropriate for 45-plex Biolegend panel</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">show_plot</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now run UMAP on the chosen PCs </span>
<span class="c1"># UMAPs rely on an initial randomization which leads to different projections every time</span>
<span class="c1"># To address this, pass random_state to the run_umap method</span>
<span class="c1"># See https://jlmelville.github.io/uwot/abparams.html for appropriate spread/min_dist values</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span> <span class="c1">#, spread=, min_dist= </span>
<span class="c1">###verbose=True</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster the AOCs. Several options are available, see help info for more options</span>
<span class="c1"># Graph-community clustering: the higher the k value, the smaller the number of clusters --&gt; adjust if necessary</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;graph-community&#39;</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot heatmap of normalized, clustered protein data</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot UMAP for visualization of clusters</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span><span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP colored by the expression of each AOC</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span>
                           <span class="n">colorby</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
                           <span class="n">features</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ids</span><span class="p">())</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Relabel clone names acccording to biology, combine clones by assigned identical names</span>
<span class="c1"># Just an example! Need to fill this in based on knowledge of the sample and markers</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">rename_labels</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s1">&#39;1&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type A&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;2&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type B&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;3&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type C&#39;</span><span class="p">,</span>
        <span class="s1">&#39;4&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type D&#39;</span><span class="p">,</span>
        <span class="s1">&#39;5&#39;</span><span class="p">:</span> <span class="s1">&#39;Cell type E&#39;</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run heatmap again with new labels</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Drop cells (barcodes) that need to be removed</span>
<span class="c1"># Instead of overwriting the sample.protein variable, we define new variable called sample.protein2</span>
<span class="c1"># Note: if you do not wish to drop any cells, this code will producs an error (ignore)</span>
<span class="n">fp_barcodes2</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">barcodes</span><span class="p">(</span><span class="s2">&quot;FP&quot;</span><span class="p">)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">fp_barcodes2</span><span class="p">)</span>
<span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">get_labels</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Additionally, option to remove AOCs from the data </span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;CD25&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Shape of cleaned up protein data (number of cells, number of AOCs)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Re-run heatmap with new filtered data</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show heatmap with convolve and subclustering turned off</span>
<span class="n">bars</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">clustered_barcodes</span><span class="p">(</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">subcluster</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># This is useful to create &quot;convolved&quot; heatmaps which are easier to interpret</span>
<span class="c1"># With the subclustering off and convolve=20, the noise will be reduced and real signals will be easier to determine</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">bars_order</span><span class="o">=</span><span class="n">bars</span><span class="p">,</span> <span class="n">convolve</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">900</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Signature will create a dataframe based on the layer you want to look at and which statistical value you want to view</span>
<span class="c1"># You can see: mean, median, mode, or std with this function</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># signaturemap will plot a heatmap of the signature dataframe created above</span>
<span class="c1"># The labels list will control the order of the clusters along the y-axis</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">signaturemap</span><span class="p">(</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">)</span> <span class="c1">#labels=[]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="statistics">
<h2>Statistics<a class="headerlink" href="#statistics" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_protein</span><span class="p">,</span> <span class="n">tstat_protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">test_signature</span><span class="p">(</span><span class="s2">&quot;normalized_counts&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_protein</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pval_protein</span> <span class="o">=</span> <span class="n">pval_protein</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">**</span> <span class="o">-</span><span class="mi">50</span>
<span class="n">pvals_protein</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pval_protein</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">tstat_protein</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Colored tiles indicate strong association of a protein with a particular cluster</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pvals_protein</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="combined-visualizations">
<h1>Combined Visualizations<a class="headerlink" href="#combined-visualizations" title="Permalink to this heading">#</a></h1>
<p>In this section you will first subset the data to only retain barcodes with remaining DNA, CNV and Protein data. Then this data can be plotted together using <code class="docutils literal notranslate"><span class="pre">sample.heatmap()</span></code>, <code class="docutils literal notranslate"><span class="pre">sample.signaturemap()</span></code>, <code class="docutils literal notranslate"><span class="pre">sample.clone_vs_analyte()</span></code> and other functions.</p>
<p>You will also be able to quantify and visualize the overlap between the different analytes, using <code class="docutils literal notranslate"><span class="pre">sample.crosstabmap()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will return the barcodes common to all assays in the sample.</span>
<span class="n">sample</span><span class="o">.</span><span class="n">common_barcodes</span><span class="p">()</span>

<span class="c1"># Use that to filter the sample so that only the common barcodes are present in all assays</span>
<span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[</span><span class="n">sample</span><span class="o">.</span><span class="n">common_barcodes</span><span class="p">()]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check dimensionality for each subclass; the number of cells (first number) should be the same in each data set</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
<span class="n">sample</span><span class="o">.</span><span class="n">cnv</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># DNA + CNV + Protein heatmap</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="n">clusterby</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;cnv&#39;</span><span class="p">,</span> <span class="s1">&#39;protein&#39;</span><span class="p">),</span>  <span class="c1"># The first assay is used for the labels</span>
    <span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="s1">&#39;normalized_counts&#39;</span><span class="p">],</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>  <span class="c1"># If None, then clustered_ids is used</span>
    <span class="n">bars_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>  <span class="c1"># The order of the barcodes</span>
    <span class="n">widths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">order</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;cnv&#39;</span><span class="p">,</span> <span class="s1">&#39;protein&#39;</span><span class="p">)</span>  <span class="c1"># The order in which the heatmaps should be drawn</span>
<span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">1500</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># DNA + CNV heatmap</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="n">clusterby</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;cnv&#39;</span><span class="p">),</span>  <span class="c1"># The first assay is used for the labels</span>
    <span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;ploidy&#39;</span><span class="p">],</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;genes&#39;</span><span class="p">],</span>  <span class="c1"># If None, then clustered_ids is used</span>
    <span class="n">bars_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>  <span class="c1"># The order of the barcodes</span>
    <span class="n">widths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">order</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;cnv&#39;</span><span class="p">)</span>  <span class="c1"># The order in which the heatmaps should be drawn</span>
<span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">1200</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># DNA + Protein heatmap</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="n">clusterby</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;protein&#39;</span><span class="p">),</span>  <span class="c1"># The first assay is used for the labels</span>
    <span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;normalized_counts&#39;</span><span class="p">],</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>  <span class="c1"># If None, then clustered_ids is used</span>
    <span class="n">bars_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>  <span class="c1"># The order of the barcodes</span>
    <span class="n">widths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">order</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;protein&#39;</span><span class="p">)</span>  <span class="c1"># The order in which the heatmaps should be drawn</span>
<span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="mi">1200</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">signaturemap</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot a combined signature heatmap, showing DNA, CNV and Protein signatures for all subclones</span>
<span class="c1"># If you want to only plot 2 of the analytes </span>
<span class="c1"># You can simply adjust the code below by removing the arguments for the analyte you no longer wish to see</span>
<span class="n">sample</span><span class="o">.</span><span class="n">signaturemap</span><span class="p">(</span>
    <span class="n">clusterby</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;cnv&#39;</span><span class="p">,</span> <span class="s1">&#39;protein&#39;</span><span class="p">),</span>
    <span class="n">attributes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;NGT&#39;</span><span class="p">,</span> <span class="s1">&#39;ploidy&#39;</span><span class="p">,</span> <span class="s1">&#39;normalized_counts&#39;</span><span class="p">),</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
    <span class="n">signature_kind</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">],</span>
    <span class="n">widths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">order</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="s1">&#39;cnv&#39;</span><span class="p">,</span> <span class="s1">&#39;protein&#39;</span><span class="p">)</span>  <span class="c1"># The order in which the heatmaps should be drawn</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clone vs analyte</span>
<span class="c1"># Visualize the CNV data stratified by clone</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">(</span><span class="s1">&#39;cnv&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">texts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="n">fig</span>
<span class="c1">#fig.savefig(&#39;genotype_cnv.png&#39;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the protein data (violin plots) stratified by clone</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">(</span><span class="s1">&#39;protein&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">texts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="n">fig</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To review a subset of features, first subset the assay itself </span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="p">[:,</span> <span class="p">[</span><span class="s1">&#39;CD11b&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD38&#39;</span><span class="p">,</span> <span class="s1">&#39;CD90&#39;</span><span class="p">]]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">clone_vs_analyte</span><span class="p">(</span><span class="s1">&#39;protein&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">texts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="n">fig</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the protein dataframe to include all AOCs again</span>
<span class="n">sample</span><span class="o">.</span><span class="n">reset</span><span class="p">(</span><span class="s1">&#39;protein&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the protein heatmap with the DNA clone labels</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span><span class="n">splitby</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the protein UMAP and color with the DNA clone labels</span>
<span class="n">protein_umap</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;umap&quot;</span><span class="p">]</span>
<span class="n">fig</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="n">protein_umap</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">)</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the protein UMAP and color with variant layers (e.g., VAFs) for specific variants</span>
<span class="n">protein_umap</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;umap&quot;</span><span class="p">]</span>
<span class="n">feats</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span> <span class="c1"># plot the first 3 variants</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="n">protein_umap</span><span class="p">,</span> <span class="n">colorby</span><span class="o">=</span><span class="s2">&quot;AF&quot;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">feats</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the AOC ridge plots split by DNA clone</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ridgeplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
                         <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD11b&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD38&#39;</span><span class="p">,</span> <span class="s1">&#39;CD90&#39;</span><span class="p">],</span>
                         <span class="n">splitby</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the AOC ridge plots split by protein cluster</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">ridgeplot</span><span class="p">(</span><span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span>
                         <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD11b&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD38&#39;</span><span class="p">,</span> <span class="s1">&#39;CD90&#39;</span><span class="p">],</span>
                         <span class="n">splitby</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the biaxial AOC plot split by DNA clone</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">feature_scatter</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">],</span> <span class="n">colorby</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
<span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the biaxial AOC plot colored by genotype (or AF_MISSING, etc.) for specific variants</span>
<span class="n">variants</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;WT1:11:32421651:T:C&#39;</span><span class="p">,</span> <span class="s1">&#39;FLT3:13:28597686:G:A&#39;</span><span class="p">]</span>
<span class="n">layer</span> <span class="o">=</span> <span class="s1">&#39;NGT_FILTERED&#39;</span>
<span class="n">AOCs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">]</span> <span class="c1"># only two!</span>

<span class="c1"># Set the layout and size of the resulting plots (rows, columns) - may need to adjust based on number of variants</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">variants</span><span class="p">),</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="c1"># Generate plots for each variant</span>
<span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
    <span class="n">color_variant</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layer</span><span class="p">][:,</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span> <span class="o">==</span> <span class="n">variant</span><span class="p">]</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">feature_scatter</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s1">&#39;normalized_counts&#39;</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="n">AOCs</span><span class="p">,</span> <span class="n">colorby</span> <span class="o">=</span> <span class="n">color_variant</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">variant</span><span class="p">))</span>
    <span class="n">mutils</span><span class="o">.</span><span class="n">static_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">counter</span><span class="p">])</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="n">counter</span> <span class="o">+</span> <span class="mi">1</span>

<span class="c1"># Show the plot</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">counter</span><span class="p">):</span>
    <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>    
</pre></div>
</div>
</div>
</div>
<section id="quantify-overlaps-between-dna-clones-and-protein-clusters">
<h2>Quantify overlaps between DNA clones and Protein clusters<a class="headerlink" href="#quantify-overlaps-between-dna-clones-and-protein-clusters" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use crosstab to calculate the overlap between DNA subclones and Protein Clusters</span>
<span class="c1"># This will generate the dataframe</span>
<span class="c1"># What percentage of each Protein cluster is defined by each Subclone</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">get_labels</span><span class="p">(),</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will generate the heatmap of the dataframe generated above</span>
<span class="c1"># What percentage of each Protein cluster is defined by each Subclone</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">crosstabmap</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">get_labels</span><span class="p">(),</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will generate the dataframe</span>
<span class="c1"># What percentage of each Subclone is defined by each Protein cluster</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">(),</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will generate the heatmap of the dataframe generated above</span>
<span class="c1"># What percentage of each Subclone is defined by each Protein cluster</span>
<span class="n">sample</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">crosstabmap</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">get_labels</span><span class="p">(),</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="export-and-save-data">
<h1>Export and Save Data<a class="headerlink" href="#export-and-save-data" title="Permalink to this heading">#</a></h1>
<p>In this section you can export a filtered .h5 file, which will contain all new labels/layers, and contain only the filtered barcodes/cells remaining. You can also export all data (row attributes, column attributes and layers) for every assay (DNA and CNV) into easily parsable .csv tables.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save new h5 file that includes only the final, cleaned dataset</span>
<span class="n">ms</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;FilteredData.h5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># With new code implementation:</span>
<span class="c1"># Export data into csv formats</span>
<span class="c1"># DNA, CNV and metadata will be included in the zip</span>
<span class="n">ms</span><span class="o">.</span><span class="n">to_zip</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;filename&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="appendix">
<h1>Appendix<a class="headerlink" href="#appendix" title="Permalink to this heading">#</a></h1>
<section id="dna-signature">
<h2>DNA signature<a class="headerlink" href="#dna-signature" title="Permalink to this heading">#</a></h2>
<p>Using <code class="docutils literal notranslate"><span class="pre">.signature()</span></code> and <code class="docutils literal notranslate"><span class="pre">.signaturemap()</span></code> you can visualize different statistical metrics of your data, including: mean, median, mode and std.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Signature will create a dataframe based on the layer you want to look at and which statistical value you want to view</span>
<span class="c1"># You can see: mean, median, mode, or std with this function</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="s1">&#39;AF&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># signaturemap will plot a heatmap of the signature dataframe created above</span>
<span class="c1"># The labels list will control the order of the subclones along the y-axis</span>
<span class="n">sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">signaturemap</span><span class="p">(</span><span class="s1">&#39;AF&#39;</span><span class="p">)</span> <span class="c1">#labels=[]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="compass-imputation">
<h2>Compass imputation<a class="headerlink" href="#compass-imputation" title="Permalink to this heading">#</a></h2>
<p><b> This section of the notebook is OPTIONAL and is still a work in progress</b></p>
<p>Compass can be used to impute the genotypes of cells with some missing data. It can also be used to infer the phylogeny of all subclones present in the sample. If a cells nature cannot be determined, Compass will label these cells as Mixed or Ambiguous.<b> Compass can take ~1-10 minutes to run depending on the size of the data.</b></p>
<p>The Compass publication can be found <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2022.01.06.475205v3">here</a><br></p>
<p><b>Note:</b> Compass can give different subclone composition than the variant subclone workflow.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use compass to assume the subclone architecture and phylogeny</span>
<span class="c1"># Depending on the size and complexity of the data, this step can take ~1-10 minutes</span>
<span class="c1"># Use full_sample instead of sample to include all previously removed cells and use the unannotated variant names</span>
<span class="n">compass</span> <span class="o">=</span> <span class="n">COMPASS</span><span class="p">(</span><span class="n">full_sample</span><span class="p">,</span> <span class="n">somatic_variants</span><span class="o">=</span><span class="n">variants</span><span class="p">)</span>
<span class="n">compass</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The phylogentic tree prediction by COMPASS</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">compass</span><span class="o">.</span><span class="n">plot_tree</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># dict of node names pointing to descriptions</span>
<span class="c1"># compass_labels_ is just the node names</span>
<span class="c1"># we want the compass_labels to be the node_descriptions</span>
<span class="n">labs</span> <span class="o">=</span> <span class="n">compass</span><span class="o">.</span><span class="n">labels_</span>  <span class="c1"># Stores the labels</span>
<span class="n">desc</span> <span class="o">=</span> <span class="n">compass</span><span class="o">.</span><span class="n">node_descriptions</span><span class="p">()</span>  <span class="c1"># Stores the dict mapping the label to the description</span>
<span class="n">compass_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">desc</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">lab</span><span class="p">,</span> <span class="n">lab</span><span class="p">)</span> <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">labs</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Store the compass label as a row_attr</span>
<span class="n">full_sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">add_row_attr</span><span class="p">(</span><span class="s1">&#39;compass_labels&#39;</span><span class="p">,</span> <span class="n">compass_labels</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Store the index positions of the COMPASS ids for the heatmap</span>
<span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">full_sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">(),</span> <span class="n">compass</span><span class="o">.</span><span class="n">somatic_variants</span><span class="p">)</span>

<span class="c1"># Visualize the assignment using the variants passed to COMPASS</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">full_sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="s1">&#39;NGT&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">full_sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">ids</span><span class="p">()[</span><span class="n">idx</span><span class="p">],</span> <span class="n">splitby</span><span class="o">=</span><span class="s1">&#39;compass_labels&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will return a dataframe </span>
<span class="c1"># Showing the overlap of variant workflow clones and compass identified clones</span>
<span class="n">full_sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">compass_labels</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will plot a heatmap of the crosstab dataframe</span>
<span class="n">full_sample</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">crosstabmap</span><span class="p">(</span><span class="n">compass_labels</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>


                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="dna-multisample.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Multisample DNA-only notebook</p>
      </div>
    </a>
    <a class="right-next"
       href="dna-protein-multisample.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Multisample DNA + Protein notebook</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Single Sample DNA + Protein notebook</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-structure">Data Structure</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#dna-analysis">DNA Analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-filtering">Basic filtering</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-addition">Annotation Addition</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variant-selection-and-subclone-identification">Variant selection and Subclone identification</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adjusting-subclone-colors-or-heatmap-colors">Adjusting subclone colors or heatmap colors</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv-analysis">CNV Analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv-workflow">CNV workflow</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#protein-analysis">Protein Analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#normalization-and-data-inspection">Normalization And Data Inspection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clustering-and-visualization">Clustering And Visualization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistics">Statistics</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#combined-visualizations">Combined Visualizations</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantify-overlaps-between-dna-clones-and-protein-clusters">Quantify overlaps between DNA clones and Protein clusters</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#export-and-save-data">Export and Save Data</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#appendix">Appendix</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dna-signature">DNA signature</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#compass-imputation">Compass imputation</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Saurabh Parikh
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021, Mission Bio.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>